{"cells":[{"cell_type":"markdown","source":["# GECKO"],"metadata":{}},{"cell_type":"markdown","source":["GECKO algorithm can be used to easily adjust the metabolic activity within the\n","cell to respect many known parameters, measured by proteomics and other\n","methods.\n","\n","The original description from GECKO is by: [Sánchez, et. al., \"Improving the\n","phenotype predictions of a yeast genome‐scale metabolic model by incorporating\n","enzymatic constraints.\", Molecular systems biology,\n","2017](https://doi.org/10.15252/msb.20167411).\n","\n","The analysis method and implementation in COBREXA is similar to\n","[sMOMENT](14_smoment.md), but GECKO is able to process and represent much\n","larger scale of the constraints -- mainly, it supports multiple isozymes for\n","each reaction, and the isozymes can be grouped into \"enzyme mass groups\" to\n","simplify interpretation of data from proteomics."],"metadata":{}},{"cell_type":"markdown","source":["For demonstration, we will generate artificial random data in a way similar\n","to the [sMOMENT example](14_smoment.md):"],"metadata":{}},{"outputs":[{"output_type":"execute_result","data":{"text/plain":"69-element Vector{String}:\n \"PFK\"\n \"PFL\"\n \"PGI\"\n \"PGK\"\n \"PGL\"\n \"ACALD\"\n \"AKGt2r\"\n \"PGM\"\n \"PIt2r\"\n \"ALCD2x\"\n ⋮\n \"MALt2_2\"\n \"MDH\"\n \"ME1\"\n \"ME2\"\n \"NADH16\"\n \"NADTRHD\"\n \"NH4t\"\n \"O2t\"\n \"PDH\""},"metadata":{},"execution_count":1}],"cell_type":"code","source":["!isfile(\"e_coli_core.json\") &&\n","    download(\"http://bigg.ucsd.edu/static/models/e_coli_core.json\", \"e_coli_core.json\")\n","\n","using COBREXA, GLPK\n","\n","model = load_model(\"e_coli_core.json\")\n","\n","import Random\n","Random.seed!(1) # repeatability\n","\n","gene_product_masses = Dict(genes(model) .=> randn(n_genes(model)) .* 10 .+ 60)\n","\n","rxns = filter(\n","    x ->\n","        !looks_like_biomass_reaction(x) &&\n","            !looks_like_exchange_reaction(x) &&\n","            !isnothing(reaction_gene_association(model, x)),\n","    reactions(model),\n",")"],"metadata":{},"execution_count":1},{"cell_type":"markdown","source":["The main difference from sMOMENT comes from allowing multiple isozymes per\n","reaction (reactions with missing isozyme informations will be ignored,\n","leaving them as-is):"],"metadata":{}},{"outputs":[{"output_type":"execute_result","data":{"text/plain":"Dict{String, Vector{Isozyme}} with 69 entries:\n  \"ACALD\"   => [Isozyme(Dict(\"b0351\"=>1), 462.073, 330.922), Isozyme(Dict(\"b124…\n  \"PTAr\"    => [Isozyme(Dict(\"b2297\"=>1), 471.533, 631.132), Isozyme(Dict(\"b245…\n  \"ALCD2x\"  => [Isozyme(Dict(\"b0356\"=>1), 617.367, 537.297), Isozyme(Dict(\"b147…\n  \"PDH\"     => [Isozyme(Dict(\"b0114\"=>1, \"b0115\"=>1, \"b0116\"=>1), 670.35, 423.7…\n  \"PYK\"     => [Isozyme(Dict(\"b1854\"=>1), 607.735, 454.975), Isozyme(Dict(\"b167…\n  \"CO2t\"    => [Isozyme(Dict(\"s0001\"=>1), 847.517, 338.625)]\n  \"MALt2_2\" => [Isozyme(Dict(\"b3528\"=>1), 383.908, 397.408)]\n  \"CS\"      => [Isozyme(Dict(\"b0720\"=>1), 750.383, 549.73)]\n  \"PGM\"     => [Isozyme(Dict(\"b3612\"=>1), 586.875, 439.579), Isozyme(Dict(\"b439…\n  \"TKT1\"    => [Isozyme(Dict(\"b2935\"=>1), 650.793, 528.251), Isozyme(Dict(\"b246…\n  \"ACONTa\"  => [Isozyme(Dict(\"b0118\"=>1), 621.458, 449.334), Isozyme(Dict(\"b127…\n  \"GLNS\"    => [Isozyme(Dict(\"b3870\"=>1), 411.219, 513.669), Isozyme(Dict(\"b129…\n  \"ICL\"     => [Isozyme(Dict(\"b4015\"=>1), 657.74, 627.894)]\n  \"FBA\"     => [Isozyme(Dict(\"b2097\"=>1), 575.634, 511.58), Isozyme(Dict(\"b1773…\n  \"FORt2\"   => [Isozyme(Dict(\"b0904\"=>1), 669.647, 627.895), Isozyme(Dict(\"b249…\n  \"G6PDH2r\" => [Isozyme(Dict(\"b1852\"=>1), 737.135, 408.509)]\n  \"AKGDH\"   => [Isozyme(Dict(\"b0727\"=>1, \"b0116\"=>1, \"b0726\"=>1), 537.049, 670.…\n  \"TKT2\"    => [Isozyme(Dict(\"b2935\"=>1), 676.15, 452.057), Isozyme(Dict(\"b2465…\n  \"FRD7\"    => [Isozyme(Dict(\"b4151\"=>1, \"b4153\"=>1, \"b4154\"=>1, \"b4152\"=>1), 5…\n  ⋮         => ⋮"},"metadata":{},"execution_count":2}],"cell_type":"code","source":["rxn_isozymes = Dict(\n","    rxn => [\n","        Isozyme(\n","            Dict(isozyme_genes .=> 1),\n","            randn() * 100 + 600, #forward kcat\n","            randn() * 100 + 500, #reverse kcat\n","        ) for isozyme_genes in reaction_gene_association(model, rxn)\n","    ] for rxn in rxns\n",")"],"metadata":{},"execution_count":2},{"cell_type":"markdown","source":["We also construct similar bounds for total gene product amounts:"],"metadata":{}},{"outputs":[{"output_type":"execute_result","data":{"text/plain":"Dict{String, Tuple{Float64, Float64}} with 137 entries:\n  \"b4301\" => (0.0, 10.0)\n  \"b1602\" => (0.0, 10.0)\n  \"b4154\" => (0.0, 10.0)\n  \"b3236\" => (0.0, 10.0)\n  \"b1621\" => (0.0, 10.0)\n  \"b1779\" => (0.0, 10.0)\n  \"b3951\" => (0.0, 10.0)\n  \"b1676\" => (0.0, 10.0)\n  \"b3114\" => (0.0, 10.0)\n  \"b1241\" => (0.0, 10.0)\n  \"b2276\" => (0.0, 10.0)\n  \"b1761\" => (0.0, 10.0)\n  \"b3925\" => (0.0, 10.0)\n  \"b3493\" => (0.0, 10.0)\n  \"b3733\" => (0.0, 10.0)\n  \"b2926\" => (0.0, 10.0)\n  \"b0979\" => (0.0, 10.0)\n  \"b4015\" => (0.0, 10.0)\n  \"b2296\" => (0.0, 10.0)\n  ⋮       => ⋮"},"metadata":{},"execution_count":3}],"cell_type":"code","source":["gene_product_bounds = Dict(genes(model) .=> Ref((0.0, 10.0)))"],"metadata":{},"execution_count":3},{"cell_type":"markdown","source":["With this, the construction of the model constrained by all enzymatic\n","information is straightforward:"],"metadata":{}},{"outputs":[{"output_type":"execute_result","data":{"text/plain":"Metabolic model of type GeckoModel\nsparse([9, 51, 55, 64, 65, 73, 9, 51, 55, 64  …  200, 201, 202, 203, 204, 205, 206, 207, 208, 209], [1, 1, 1, 1, 1, 1, 2, 2, 2, 2  …  325, 326, 327, 328, 329, 330, 331, 332, 333, 334], [1.0, 1.0, -1.0, -1.0, 1.0, -0.001546133025849219, 1.0, 1.0, -1.0, -1.0  …  1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0], 209, 334)\nNumber of reactions: 334\nNumber of metabolites: 209\n"},"metadata":{},"execution_count":4}],"cell_type":"code","source":["gecko_model =\n","    model |> with_gecko(;\n","        reaction_isozymes = rxn_isozymes,\n","        gene_product_bounds,\n","        gene_product_molar_mass = gene_product_masses,\n","        gene_product_mass_group = _ -> \"uncategorized\", # all products belong to the same \"uncategorized\" category\n","        gene_product_mass_group_bound = _ -> 100.0, # the total limit of mass in the single category\n","    )"],"metadata":{},"execution_count":4},{"cell_type":"markdown","source":["(Alternatively, you may use `make_gecko_model`, which does the same\n","without piping by `|>`.)"],"metadata":{}},{"cell_type":"markdown","source":["The stoichiometry and coupling in the gecko model is noticeably more complex;\n","you may notice new \"reactions\" added that simulate the gene product\n","utilization:"],"metadata":{}},{"outputs":[{"output_type":"execute_result","data":{"text/plain":"262×334 SparseArrays.SparseMatrixCSC{Float64, Int64} with 1386 stored entries:\n⠄⠦⠤⠤⠄⠶⠶⠦⠇⠶⡴⠠⠀⠀⢧⠀⠢⣲⠼⠿⠶⣶⡄⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀\n⠀⡎⢶⡉⣁⡄⣬⣠⡅⠔⠰⡈⠀⠀⠐⡄⡀⠄⣫⡉⡉⠩⠝⠇⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀\n⠭⠰⠀⠒⣆⣀⣊⣬⢗⣾⢠⣵⣆⠆⠀⡁⠨⠁⠈⠀⠂⠥⠁⠅⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀\n⣧⡣⠁⠤⠟⠒⠻⠿⡗⠆⠠⣧⣤⢵⠀⢠⣄⡀⠘⠋⠚⠠⠂⠆⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀\n⢶⣂⠀⠀⠀⠀⠀⠀⠃⠀⠀⠉⠉⠉⠁⠈⠉⠉⠓⠀⠀⠀⠀⠰⢄⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀\n⠀⠈⠙⠺⠥⣀⡀⠀⠤⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠤⠀⠠⠄⠀⠱⢄⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀\n⠀⠀⠀⠀⠀⠀⠘⣆⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠁⠀⠀⠀⠑⢆⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀\n⠀⠀⠀⠀⠀⠀⠀⠀⠉⠳⢄⡀⠀⠀⠀⠀⠀⠐⠀⠀⠀⠐⠀⠀⠀⠀⠀⠀⠀⠑⢆⠀⠀⠀⠀⠀⠀⠀⠀⠀\n⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠙⠲⠄⠀⠠⣄⠀⠀⠀⠀⠀⠐⠀⠀⠀⠀⠀⠀⠀⠀⠙⢦⠀⠀⠀⠀⠀⠀⠀\n⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⠧⣜⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠙⢦⡀⠀⠀⠀⠀\n⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⠳⣄⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠙⢦⡀⠀⠀\n⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⡇⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠙⢦⡀\n⠒⠰⠤⣄⣀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠉⠃⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠙\n⠀⠀⠀⠀⠀⠉⠁⠲⠤⢤⡀⣀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀\n⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⠙⠀⠀⠐⠲⠤⢠⣀⣀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀\n⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠉⠘⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒"},"metadata":{},"execution_count":5}],"cell_type":"code","source":["[stoichiometry(gecko_model); coupling(gecko_model)]"],"metadata":{},"execution_count":5},{"cell_type":"markdown","source":["Again, the resulting model can be used in any type of analysis. For example, flux balance analysis:"],"metadata":{}},{"outputs":[{"output_type":"execute_result","data":{"text/plain":"A JuMP Model\nMaximization problem with:\nVariables: 334\nObjective function type: JuMP.AffExpr\n`JuMP.AffExpr`-in-`MathOptInterface.EqualTo{Float64}`: 209 constraints\n`JuMP.AffExpr`-in-`MathOptInterface.LessThan{Float64}`: 774 constraints\nModel mode: AUTOMATIC\nCachingOptimizer state: ATTACHED_OPTIMIZER\nSolver name: GLPK\nNames registered in the model: c_lbs, c_ubs, lbs, mb, ubs, x"},"metadata":{},"execution_count":6}],"cell_type":"code","source":["opt_model = flux_balance_analysis(gecko_model, GLPK.Optimizer)"],"metadata":{},"execution_count":6},{"cell_type":"markdown","source":["Get the fluxes"],"metadata":{}},{"outputs":[{"output_type":"execute_result","data":{"text/plain":"Dict{String, Float64} with 95 entries:\n  \"ACALD\"       => 0.0\n  \"PTAr\"        => 6.34803\n  \"ALCD2x\"      => 0.0\n  \"PDH\"         => 10.4628\n  \"PYK\"         => 2.52634\n  \"CO2t\"        => -16.3044\n  \"EX_nh4_e\"    => -3.97209\n  \"MALt2_2\"     => 0.0\n  \"CS\"          => 1.38468\n  \"PGM\"         => -14.9919\n  \"TKT1\"        => 1.85156\n  \"EX_mal__L_e\" => 0.0\n  \"ACONTa\"      => 1.38468\n  \"EX_pi_e\"     => -2.67974\n  \"GLNS\"        => 0.186264\n  \"ICL\"         => 0.0\n  \"EX_o2_e\"     => -15.4623\n  \"FBA\"         => 7.29353\n  \"EX_gln__L_e\" => 0.0\n  ⋮             => ⋮"},"metadata":{},"execution_count":7}],"cell_type":"code","source":["flux_sol = flux_dict(gecko_model, opt_model)"],"metadata":{},"execution_count":7},{"cell_type":"markdown","source":["Get the gene product concentrations"],"metadata":{}},{"outputs":[{"output_type":"execute_result","data":{"text/plain":"Dict{String, Float64} with 137 entries:\n  \"b4301\" => -3.40857e-19\n  \"b1602\" => 0.0\n  \"b4154\" => 0.0\n  \"b3236\" => 0.000804895\n  \"b1621\" => 0.0144236\n  \"b3951\" => 0.0\n  \"b1779\" => 0.0252316\n  \"b1676\" => 0.0\n  \"b3114\" => 0.0\n  \"b1241\" => 0.0\n  \"b2276\" => 0.0499511\n  \"b1761\" => 0.00565514\n  \"b3925\" => 0.0\n  \"b3493\" => 0.00349559\n  \"b3733\" => 0.0550335\n  \"b2926\" => 0.0332496\n  \"b0979\" => 0.0581609\n  \"b4015\" => 0.0\n  \"b2296\" => 0.0170712\n  ⋮       => ⋮"},"metadata":{},"execution_count":8}],"cell_type":"code","source":["gp_concs = gene_product_dict(gecko_model, opt_model)"],"metadata":{},"execution_count":8},{"cell_type":"markdown","source":["Get the total masses assigned to each mass group"],"metadata":{}},{"outputs":[{"output_type":"execute_result","data":{"text/plain":"Dict{String, Float64} with 1 entry:\n  \"uncategorized\" => 100.0"},"metadata":{},"execution_count":9}],"cell_type":"code","source":["gene_product_mass_group_dict(gecko_model, opt_model)"],"metadata":{},"execution_count":9},{"cell_type":"markdown","source":["Variability:"],"metadata":{}},{"outputs":[{"output_type":"execute_result","data":{"text/plain":"95×2 Matrix{Float64}:\n   5.94319       11.2165\n   0.0           14.5575\n  -0.0321351      9.85813\n -18.1604       -14.1881\n   0.0            9.89027\n  -3.45412        0.0\n  -1.18575        0.0\n -17.1251       -13.1528\n   2.54576        2.67974\n  -3.45412        0.0\n   ⋮            \n   0.0            0.0\n  -2.77471        8.56368\n   0.0            3.01162\n   0.0            4.05986\n  27.1611        31.3545\n   0.0            7.91762\n   3.77348        4.79778\n  14.092         17.1644\n   1.60011e-15   15.6326"},"metadata":{},"execution_count":10}],"cell_type":"code","source":["flux_variability_analysis(gecko_model, GLPK.Optimizer, bounds = gamma_bounds(0.95))"],"metadata":{},"execution_count":10},{"cell_type":"markdown","source":["...and sampling:"],"metadata":{}},{"outputs":[{"output_type":"execute_result","data":{"text/plain":"3340×95 Matrix{Float64}:\n 5.79321  7.53434  4.341     -9.82259  …  2.58742  0.178106  4.65941  1.68849\n 5.97172  7.69418  4.47547  -10.2705      2.82836  0.191991  5.18346  1.86908\n 5.88919  7.08271  4.19867   -9.8486      2.55031  0.2609    4.80288  1.79194\n 5.78822  7.59651  4.33514   -9.84917     2.56176  0.188864  4.66782  1.61288\n 5.38645  7.6815   4.22388   -9.25145     2.24728  0.14908   4.24294  1.36614\n 5.79294  7.4851   4.27372   -9.70463  …  2.58035  0.180579  4.68954  1.63211\n 5.85428  7.62246  4.38253   -9.86583     2.55356  0.171362  4.67094  1.72747\n 5.38753  7.2377   4.08927   -9.7917      2.9556   0.206229  5.38608  1.98014\n 5.70746  7.3966   4.24649   -9.70319     2.63334  0.178416  4.66491  1.69637\n 5.69817  7.69786  4.2997    -9.98324     3.02154  0.186499  4.66858  1.4801\n ⋮                                     ⋱                              \n 6.41219  7.55886  4.77719  -10.7465      2.51659  0.190647  4.86579  2.17292\n 6.09411  7.09449  4.54019   -9.93358     2.30911  0.221519  4.17799  1.88773\n 5.98383  7.43657  4.61554  -10.2747      2.38729  0.175689  4.70194  2.12283\n 6.06742  7.24054  4.54556   -9.99035     2.35708  0.190526  4.38236  1.95298\n 6.09424  7.25184  4.57651  -10.0623   …  2.35591  0.186185  4.36823  1.95021\n 6.21735  8.24101  4.63718  -10.4816      2.35836  0.174095  5.30491  1.95086\n 5.92004  7.28773  4.60061   -9.98184     2.19296  0.176998  4.47597  2.137\n 6.17176  7.33389  4.65558  -10.1599      2.34695  0.174037  4.59866  2.17223\n 6.14954  7.02426  4.42488   -9.97557     2.96939  0.16056   4.262    1.86895"},"metadata":{},"execution_count":11}],"cell_type":"code","source":["affine_hit_and_run(gecko_model, warmup_from_variability(gecko_model, GLPK.Optimizer))' *\n","reaction_flux(gecko_model)"],"metadata":{},"execution_count":11},{"cell_type":"markdown","source":["---\n","\n","*This notebook was generated using [Literate.jl](https://github.com/fredrikekre/Literate.jl).*"],"metadata":{}}],"nbformat_minor":3,"metadata":{"language_info":{"file_extension":".jl","mimetype":"application/julia","name":"julia","version":"1.1.0"},"kernelspec":{"name":"julia-1.1","display_name":"Julia 1.1.0","language":"julia"}},"nbformat":4}