{"cells":[{"cell_type":"markdown","source":["# sMOMENT"],"metadata":{}},{"cell_type":"markdown","source":["sMOMENT algorithm can be used to easily adjust the metabolic activity within\n","the cell to respect known enzymatic parameters and enzyme mass constraints\n","measured by proteomics and other methods.\n","\n","The original description from sMOMENT is by [Bekiaris, and Klamt, \"Automatic\n","construction of metabolic models with enzyme constraints.\", BMC\n","bioinformatics, 2020](https://doi.org/10.1186/s12859-019-3329-9)\n","\n","Let's load some packages:"],"metadata":{}},{"outputs":[{"output_type":"execute_result","data":{"text/plain":"Metabolic model of type JSONModel\nsparse([9, 51, 55, 64, 65, 34, 44, 59, 66, 64  …  20, 22, 23, 25, 16, 17, 34, 44, 57, 59], [1, 1, 1, 1, 1, 2, 2, 2, 2, 3  …  93, 93, 94, 94, 95, 95, 95, 95, 95, 95], [1.0, 1.0, -1.0, -1.0, 1.0, -1.0, 1.0, -1.0, 1.0, 1.0  …  1.0, -1.0, 1.0, -1.0, -1.0, 1.0, -1.0, 1.0, 1.0, -1.0], 72, 95)\nNumber of reactions: 95\nNumber of metabolites: 72\n"},"metadata":{},"execution_count":1}],"cell_type":"code","source":["!isfile(\"e_coli_core.json\") &&\n","    download(\"http://bigg.ucsd.edu/static/models/e_coli_core.json\", \"e_coli_core.json\")\n","\n","using COBREXA, GLPK\n","\n","model = load_model(\"e_coli_core.json\")"],"metadata":{},"execution_count":1},{"cell_type":"markdown","source":["We will necessarily need the enzyme turnover numbers (aka \"kcats\") and masses\n","of the required gene products.  You do not necessarily need to know all data\n","for the given model, but the more you have, the better the approximation will\n","be.\n","\n","For the demonstration purpose, we will generate the data randomly. In a\n","realistic setting, you would input experimental or database-originating data\n","here:"],"metadata":{}},{"outputs":[{"output_type":"execute_result","data":{"text/plain":"Dict{String, Float64} with 137 entries:\n  \"b4301\" => 69.9065\n  \"b1602\" => 70.0268\n  \"b4154\" => 55.4119\n  \"b3236\" => 50.4756\n  \"b1621\" => 59.9869\n  \"b1779\" => 56.4588\n  \"b3951\" => 64.7369\n  \"b1676\" => 54.0343\n  \"b3114\" => 56.9763\n  \"b1241\" => 60.6193\n  \"b2276\" => 55.0068\n  \"b1761\" => 63.8676\n  \"b3925\" => 55.3065\n  \"b3493\" => 47.598\n  \"b3733\" => 52.1707\n  \"b2926\" => 68.4462\n  \"b0979\" => 45.6388\n  \"b4015\" => 56.497\n  \"b2296\" => 44.2344\n  ⋮       => ⋮"},"metadata":{},"execution_count":2}],"cell_type":"code","source":["import Random\n","Random.seed!(1) # repeatability\n","\n","gene_product_masses = Dict(genes(model) .=> randn(n_genes(model)) .* 10 .+ 60)"],"metadata":{},"execution_count":2},{"cell_type":"markdown","source":["We only take the reactions that have gene products (i.e., enzymes) associated with them):"],"metadata":{}},{"outputs":[{"output_type":"execute_result","data":{"text/plain":"69-element Vector{String}:\n \"PFK\"\n \"PFL\"\n \"PGI\"\n \"PGK\"\n \"PGL\"\n \"ACALD\"\n \"AKGt2r\"\n \"PGM\"\n \"PIt2r\"\n \"ALCD2x\"\n ⋮\n \"MALt2_2\"\n \"MDH\"\n \"ME1\"\n \"ME2\"\n \"NADH16\"\n \"NADTRHD\"\n \"NH4t\"\n \"O2t\"\n \"PDH\""},"metadata":{},"execution_count":3}],"cell_type":"code","source":["rxns = filter(\n","    x ->\n","        !looks_like_biomass_reaction(x) &&\n","            !looks_like_exchange_reaction(x) &&\n","            !isnothing(reaction_gene_association(model, x)),\n","    reactions(model),\n",")"],"metadata":{},"execution_count":3},{"cell_type":"markdown","source":["The information about each enzyme and its capabilities is stored in an\n","`Isozyme` structure. For simplicity, sMOMENT ignores much of the\n","information about the multiplicity of required gene products and\n","other."],"metadata":{}},{"outputs":[{"output_type":"execute_result","data":{"text/plain":"Dict{String, Isozyme} with 69 entries:\n  \"NH4t\"    => Isozyme(Dict(\"b0451\"=>1, \"s0001\"=>1), 669.647, 627.895)\n  \"ACALD\"   => Isozyme(Dict(\"b1241\"=>1, \"b0351\"=>1), 592.595, 607.625)\n  \"PTAr\"    => Isozyme(Dict(\"b2297\"=>1, \"b2458\"=>1), 584.264, 557.713)\n  \"PGL\"     => Isozyme(Dict(\"b0767\"=>1), 540.211, 485.094)\n  \"NADTRHD\" => Isozyme(Dict(\"b1603\"=>1, \"b1602\"=>1, \"b3962\"=>1), 612.8, 608.149)\n  \"ALCD2x\"  => Isozyme(Dict(\"b1241\"=>1, \"b0356\"=>1, \"b1478\"=>1), 462.073, 330.9…\n  \"PGK\"     => Isozyme(Dict(\"b2926\"=>1), 720.862, 424.687)\n  \"PDH\"     => Isozyme(Dict(\"b0114\"=>1, \"b0115\"=>1, \"b0116\"=>1), 669.985, 380.3…\n  \"LDH_D\"   => Isozyme(Dict(\"b1380\"=>1, \"b2133\"=>1), 676.15, 452.057)\n  \"PYK\"     => Isozyme(Dict(\"b1854\"=>1, \"b1676\"=>1), 646.569, 533.939)\n  \"CO2t\"    => Isozyme(Dict(\"s0001\"=>1), 599.569, 371.856)\n  \"PIt2r\"   => Isozyme(Dict(\"b2987\"=>1, \"b3493\"=>1), 460.427, 476.199)\n  \"ME1\"     => Isozyme(Dict(\"b1479\"=>1), 871.006, 523.277)\n  \"MALt2_2\" => Isozyme(Dict(\"b3528\"=>1), 524.693, 464.676)\n  \"CS\"      => Isozyme(Dict(\"b0720\"=>1), 634.028, 438.943)\n  \"PGM\"     => Isozyme(Dict(\"b4395\"=>1, \"b3612\"=>1, \"b0755\"=>1), 433.345, 483.6…\n  \"TKT1\"    => Isozyme(Dict(\"b2935\"=>1, \"b2465\"=>1), 615.688, 426.719)\n  \"ATPS4r\"  => Isozyme(Dict(\"b3737\"=>1, \"b3734\"=>1, \"b3732\"=>1, \"b3736\"=>1, \"b3…\n  \"ACONTa\"  => Isozyme(Dict(\"b0118\"=>1, \"b1276\"=>1), 379.032, 588.815)\n  ⋮         => ⋮"},"metadata":{},"execution_count":4}],"cell_type":"code","source":["rxn_isozymes = Dict(\n","    rxn => Isozyme(\n","        Dict(vcat(reaction_gene_association(model, rxn)...) .=> 1),\n","        randn() * 100 + 600, #forward kcat\n","        randn() * 100 + 500, #reverse kcat\n","    ) for rxn in rxns\n",")"],"metadata":{},"execution_count":4},{"cell_type":"markdown","source":["In case some of the reactions are missing in `rxn_isozymes`, sMOMENT simply\n","ignores them.\n","\n","Once the data is gathered, we create a model that wraps the original model\n","with additional sMOMENT structure:"],"metadata":{}},{"outputs":[{"output_type":"execute_result","data":{"text/plain":"Metabolic model of type SMomentModel\nsparse([9, 51, 55, 64, 65, 34, 44, 59, 66, 64  …  23, 25, 23, 25, 16, 17, 34, 44, 57, 59], [1, 1, 1, 1, 1, 2, 2, 2, 2, 3  …  129, 129, 130, 130, 131, 131, 131, 131, 131, 131], [1.0, 1.0, -1.0, -1.0, 1.0, -1.0, 1.0, -1.0, 1.0, -1.0  …  -1.0, 1.0, 1.0, -1.0, -1.0, 1.0, -1.0, 1.0, 1.0, -1.0], 72, 131)\nNumber of reactions: 131\nNumber of metabolites: 72\n"},"metadata":{},"execution_count":5}],"cell_type":"code","source":["smoment_model =\n","    model |> with_smoment(\n","        reaction_isozyme = rxn_isozymes,\n","        gene_product_molar_mass = gene_product_masses,\n","        total_enzyme_capacity = 50.0,\n","    )"],"metadata":{},"execution_count":5},{"cell_type":"markdown","source":["(You could alternatively use the `make_smoment_model` to create the\n","structure more manually; but `with_smoment` is easier to use e.g.\n","with `screen`.)"],"metadata":{}},{"cell_type":"markdown","source":["In turn, you should have a complete model with unidirectional reactions and\n","additional coupling, as specified by the sMOMENT method:"],"metadata":{}},{"outputs":[{"output_type":"execute_result","data":{"text/plain":"73×131 SparseArrays.SparseMatrixCSC{Float64, Int64} with 599 stored entries:\n⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠰⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠣⠀⠀⠀⠀⠀⠀⠈⠶⠶⠄⠀⡀⡀⠀⠀⠀\n⠂⠀⠓⠂⠒⠂⠘⠩⠓⠃⠛⠚⠂⠘⠘⡊⠓⠀⠂⠀⠀⠀⠀⠐⢁⠀⠀⠑⠈⠓⠒⠚⠓⠊⠡⣖⠓⠒⠀⠀\n⠀⠀⠶⠀⠰⠆⠀⠀⠀⠰⠀⢰⠀⠀⠀⠀⠀⠀⡆⠀⠀⠀⠀⠀⠈⠀⠀⠀⠈⢡⡶⣠⣤⢠⢠⠶⠹⢿⣆⠰\n⠀⣉⠀⢠⠀⠀⠀⠀⠀⠀⠀⢀⠀⠀⠐⠠⠄⠀⠀⠀⠀⠀⠀⠀⠀⠆⠀⠀⠀⠀⠉⠀⠀⠀⠀⠀⠀⠀⠐⠆\n⠀⠀⠆⠀⣒⠀⡘⠀⢃⠃⠒⠛⠀⠈⠀⠈⠁⠒⠀⠀⠀⠀⠀⠀⠀⢀⠀⠂⠈⠠⠞⠒⠀⠐⠀⠀⠈⠁⠀⠀\n⠈⠀⠀⠀⠀⠀⠀⠀⢀⠁⠀⠩⠁⣠⡔⠀⠐⠀⠠⠄⢀⠀⠀⠀⠀⠀⠁⠀⠊⠀⠈⠀⠀⢀⠀⠉⠀⠙⠀⠈\n⠠⠀⠭⠀⠈⠛⠀⠀⠀⢀⠠⠤⠀⠠⢠⠀⡤⣬⠉⠁⠐⠀⠀⠀⠀⠀⡀⠀⠄⠀⠀⠀⠀⠀⠀⠄⠄⠀⠀⠠\n⠄⠤⢠⡄⠀⠀⠎⠉⠄⠤⠴⠦⠀⠂⠀⠀⠀⠤⠐⠒⠀⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠤⠤⠀⠠⠀⠀⠀⠀⠀\n⢁⠉⣀⠀⠀⠀⠑⠂⠃⢙⢉⣉⠶⢐⠀⠀⠀⣉⠀⢀⡀⠠⠀⠀⠀⠀⢀⠀⠀⠀⠀⠉⠀⠐⠐⠀⡀⠒⠀⢐\n⢤⠄⠀⠀⠈⠁⠀⠀⠀⠀⠀⠨⠀⠀⠀⠀⠘⠨⠅⠭⠀⠐⡀⠀⠀⠀⠠⢄⠠⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀\n⣠⣄⣀⣀⣀⣀⣀⣀⡀⣀⣀⡠⣀⣀⣀⣀⣐⣠⣤⣤⡄⠀⠑⠀⠀⠀⢠⣀⣚⣢⣤⣀⣀⣀⣀⣀⣀⣀⣀⣀"},"metadata":{},"execution_count":6}],"cell_type":"code","source":["[stoichiometry(smoment_model); coupling(smoment_model)]"],"metadata":{},"execution_count":6},{"cell_type":"markdown","source":["the type (SMomentModel) is a model wrapper -- it is a thin additional layer\n","that just adds the necessary sMOMENT-relevant information atop the original\n","model, which is unmodified. That makes the process very efficient and\n","suitable for large-scale data processing. You can still access the original\n","\"base\" model hidden in the SMomentModel using `unwrap_model`."],"metadata":{}},{"cell_type":"markdown","source":["Other than that, the `SMomentModel` is a model type like any other,\n","and you can run any analysis you want on it, such as FBA:"],"metadata":{}},{"outputs":[{"output_type":"execute_result","data":{"text/plain":"Dict{String, Float64} with 95 entries:\n  \"ACALD\"       => 7.21514e-15\n  \"PTAr\"        => 4.6195\n  \"ALCD2x\"      => 7.21514e-15\n  \"PDH\"         => 5.64771\n  \"PYK\"         => 2.26508\n  \"CO2t\"        => -5.61213\n  \"EX_nh4_e\"    => -1.16158\n  \"MALt2_2\"     => 0.0\n  \"CS\"          => 0.229832\n  \"PGM\"         => -6.97218\n  \"TKT1\"        => 0.0769017\n  \"EX_mal__L_e\" => 0.0\n  \"ACONTa\"      => 0.229832\n  \"EX_pi_e\"     => -0.783652\n  \"GLNS\"        => 0.0544703\n  \"ICL\"         => -7.26379e-16\n  \"EX_o2_e\"     => -5.36585\n  \"FBA\"         => 3.65917\n  \"EX_gln__L_e\" => 7.27645e-17\n  ⋮             => ⋮"},"metadata":{},"execution_count":7}],"cell_type":"code","source":["flux_balance_analysis_dict(smoment_model, GLPK.Optimizer)"],"metadata":{},"execution_count":7},{"cell_type":"markdown","source":["(Notice that the total reaction fluxes are reported despite the fact that\n","reactions are indeed split in the model! The underlying mechanism is provided\n","by `reaction_flux` accessor.)"],"metadata":{}},{"cell_type":"markdown","source":["[Variability](06_fva.md) of the sMOMENT model can be explored as such:"],"metadata":{}},{"outputs":[{"output_type":"execute_result","data":{"text/plain":"95×2 Matrix{Float64}:\n   2.45267       5.27497\n   0.0           8.65379\n   1.02326       5.40892\n -10.4873       -5.50558\n   5.32907e-15   2.38384\n  -1.94964       0.0\n  -0.364104      0.0\n -10.1846       -5.20283\n   0.74447       0.783652\n  -1.94964       0.0\n   ⋮            \n   0.0           0.0\n  -1.3876        0.949426\n   0.0           0.82907\n   0.0           1.3876\n   7.28026      11.5838\n   0.0           1.43239\n   1.1035        1.41386\n   3.64013       6.0218\n  -3.55271e-15   6.98001"},"metadata":{},"execution_count":8}],"cell_type":"code","source":["flux_variability_analysis(smoment_model, GLPK.Optimizer, bounds = gamma_bounds(0.95))"],"metadata":{},"execution_count":8},{"cell_type":"markdown","source":["...and a sMOMENT model sample can be obtained [as usual with\n","sampling](16_hit_and_run.md):"],"metadata":{}},{"outputs":[{"output_type":"execute_result","data":{"text/plain":"1310×95 Matrix{Float64}:\n 3.10627  1.10496   2.59295  -5.80686  …  0.21826   2.28749  2.494\n 2.69496  0.739709  2.27435  -4.99406     0.195278  2.32306  2.12976\n 3.04349  1.0918    2.50525  -5.66219     0.232397  2.44905  2.28156\n 3.32046  1.15351   2.84213  -6.07036     0.281722  2.19832  2.46613\n 2.97204  1.03287   2.41138  -5.26167     0.177656  2.50011  1.82459\n 2.98846  1.04878   2.55933  -5.56856  …  0.226775  2.38195  2.25295\n 3.10718  1.03907   2.60585  -5.70943     0.218163  2.34915  2.37647\n 3.06437  1.11192   2.51873  -5.66579     0.229319  2.47645  2.25147\n 2.96738  0.91041   2.42985  -5.58605     0.247736  2.34892  2.34282\n 2.90515  0.979711  2.3791   -5.39361     0.212013  2.4115   2.10713\n ⋮                                     ⋱                     \n 2.98631  0.840977  2.39294  -5.37079     0.297915  2.44947  2.07908\n 2.96247  0.74846   2.54009  -5.51737     0.197808  2.17741  2.50925\n 2.79659  0.689854  2.29403  -5.11059     0.277939  2.46768  2.14363\n 2.73231  0.485262  2.20988  -5.08981     0.260519  2.48995  2.28067\n 2.98948  0.664077  2.47531  -5.44751  …  0.253846  2.32322  2.51151\n 2.90486  0.735478  2.38507  -5.35785     0.253099  2.37717  2.37272\n 2.9292   0.75605   2.39598  -5.38655     0.283592  2.39599  2.17941\n 2.88497  0.749269  2.37233  -5.33546     0.250493  2.38008  2.3561\n 2.9382   0.227246  2.48772  -5.36324     0.248709  2.13937  2.6633"},"metadata":{},"execution_count":9}],"cell_type":"code","source":["(\n","    affine_hit_and_run(\n","        smoment_model,\n","        warmup_from_variability(smoment_model, GLPK.Optimizer),\n","    )' * reaction_flux(smoment_model)\n",")"],"metadata":{},"execution_count":9},{"cell_type":"markdown","source":["---\n","\n","*This notebook was generated using [Literate.jl](https://github.com/fredrikekre/Literate.jl).*"],"metadata":{}}],"nbformat_minor":3,"metadata":{"language_info":{"file_extension":".jl","mimetype":"application/julia","name":"julia","version":"1.1.0"},"kernelspec":{"name":"julia-1.1","display_name":"Julia 1.1.0","language":"julia"}},"nbformat":4}