<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Function reference · COBREXA.jl</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script><link href="../assets/favicon.ico" rel="icon" type="image/x-icon"/></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../"><img class="docs-light-only" src="../assets/logo.svg" alt="COBREXA.jl logo"/><img class="docs-dark-only" src="../assets/logo-dark.svg" alt="COBREXA.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit">COBREXA.jl</span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><a class="tocitem" href="../tutorials/">Quickstart tutorials</a></li><li><a class="tocitem" href="../advanced/">Advanced tutorials</a></li><li><a class="tocitem" href="../notebooks/">Examples and notebooks</a></li><li class="is-active"><a class="tocitem" href>Function reference</a><ul class="internal"><li><a class="tocitem" href="#Base-Types"><span>Base Types</span></a></li><li><a class="tocitem" href="#Model-types-and-contents"><span>Model types and contents</span></a></li><li><a class="tocitem" href="#Base-functions"><span>Base functions</span></a></li><li><a class="tocitem" href="#File-I/O-and-serialization"><span>File I/O and serialization</span></a></li><li><a class="tocitem" href="#Model-reconstruction"><span>Model reconstruction</span></a></li><li><a class="tocitem" href="#Analysis-functions"><span>Analysis functions</span></a></li><li><a class="tocitem" href="#Miscellaneous-utilities"><span>Miscellaneous utilities</span></a></li></ul></li><li><a class="tocitem" href="../howToContribute/">How to contribute</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Function reference</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Function reference</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/master/docs/src/functions.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Functions"><a class="docs-heading-anchor" href="#Functions">Functions</a><a id="Functions-1"></a><a class="docs-heading-anchor-permalink" href="#Functions" title="Permalink"></a></h1><h2 id="Base-Types"><a class="docs-heading-anchor" href="#Base-Types">Base Types</a><a id="Base-Types-1"></a><a class="docs-heading-anchor-permalink" href="#Base-Types" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="COBREXA.Maybe" href="#COBREXA.Maybe"><code>COBREXA.Maybe</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">Maybe{T} = Union{Nothing, T}</code></pre><p>A nice name for &quot;nullable&quot; type.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/abstract/Maybe.jl#L2-L6">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA._default-Union{Tuple{T}, Tuple{T, Union{Nothing, T}}} where T" href="#COBREXA._default-Union{Tuple{T}, Tuple{T, Union{Nothing, T}}} where T"><code>COBREXA._default</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">_default(d::T, x::Maybe{T})::T where {T}</code></pre><p>Fold the <code>Maybe{T}</code> down to <code>T</code> by defaulting.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/abstract/Maybe.jl#L9-L13">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA._maybemap-Tuple{Any, Union{Nothing, T} where T}" href="#COBREXA._maybemap-Tuple{Any, Union{Nothing, T} where T}"><code>COBREXA._maybemap</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">_maybemap(f, x::Maybe)::Maybe</code></pre><p>Apply a function to <code>x</code> only if it is not <code>nothing</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/abstract/Maybe.jl#L18-L22">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.Annotations" href="#COBREXA.Annotations"><code>COBREXA.Annotations</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">Annotations = Dict{String,Vector{String}}</code></pre><p>Dictionary used to store (possible multiple) standardized annotations of something, such as a <a href="#COBREXA.Metabolite"><code>Metabolite</code></a> and a <a href="#COBREXA.Reaction"><code>Reaction</code></a>.</p><p><strong>Example</strong></p><pre><code class="language-none">Annotations(&quot;PubChem&quot; =&gt; [&quot;CID12345&quot;, &quot;CID54321&quot;])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/abstract/MetabolicModel.jl#L39-L49">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.GeneAssociation" href="#COBREXA.GeneAssociation"><code>COBREXA.GeneAssociation</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">GeneAssociation = Vector{Vector{String}}</code></pre><p>An association to genes, represented as a logical formula in a positive disjunctive normal form (DNF). (The 2nd-level vectors of strings are connected by &quot;and&quot; to form conjunctions, and the 1st-level vectors of these conjunctions are connected by &quot;or&quot; to form the DNF.)</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/abstract/MetabolicModel.jl#L22-L29">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.MetabolicModel" href="#COBREXA.MetabolicModel"><code>COBREXA.MetabolicModel</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">abstract type MetabolicModel end</code></pre><p>A helper supertype that wraps everything usable as a linear-like model for COBREXA functions.</p><p>If you want your model type to work with COBREXA, add the <code>MetabolicModel</code> as its supertype, and implement the accessor functions. Accessors <a href="#COBREXA.reactions-Tuple{MetabolicModel}"><code>reactions</code></a>, <a href="#COBREXA.metabolites-Tuple{MetabolicModel}"><code>metabolites</code></a>, <a href="#COBREXA.stoichiometry-Tuple{MetabolicModel}"><code>stoichiometry</code></a>, <a href="#COBREXA.bounds-Tuple{MetabolicModel}"><code>bounds</code></a> and <a href="#COBREXA.objective-Tuple{MetabolicModel}"><code>objective</code></a> must be implemented; others are not mandatory and default to safe &quot;empty&quot; values.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/abstract/MetabolicModel.jl#L2-L13">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.MetaboliteFormula" href="#COBREXA.MetaboliteFormula"><code>COBREXA.MetaboliteFormula</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">MetaboliteFormula = Dict{String,Int}</code></pre><p>Dictionary of atoms and their abundances in a molecule.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/abstract/MetabolicModel.jl#L32-L36">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.Notes" href="#COBREXA.Notes"><code>COBREXA.Notes</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">Notes = Dict{String,Vector{String}}</code></pre><p>Free-form notes about something (e.g. a <a href="#COBREXA.Gene"><code>Gene</code></a>), categorized by &quot;topic&quot;.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/abstract/MetabolicModel.jl#L52-L57">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.balance-Tuple{MetabolicModel}" href="#COBREXA.balance-Tuple{MetabolicModel}"><code>COBREXA.balance</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">balance(a::MetabolicModel)::SparseVec</code></pre><p>Get the sparse balance vector of a model (ie. the <code>b</code> from <code>S x = b</code>).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/abstract/MetabolicModel.jl#L116-L120">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.bounds-Tuple{MetabolicModel}" href="#COBREXA.bounds-Tuple{MetabolicModel}"><code>COBREXA.bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">bounds(a::MetabolicModel)::Tuple{SparseVec,SparseVec}</code></pre><p>Get the lower and upper flux bounds of a model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/abstract/MetabolicModel.jl#L107-L111">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.coupling-Tuple{MetabolicModel}" href="#COBREXA.coupling-Tuple{MetabolicModel}"><code>COBREXA.coupling</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">coupling(a::MetabolicModel)::SparseMat</code></pre><p>Get a matrix of coupling constraint definitions of a model. By default, there is no coupling in the models.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/abstract/MetabolicModel.jl#L134-L139">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.coupling_bounds-Tuple{MetabolicModel}" href="#COBREXA.coupling_bounds-Tuple{MetabolicModel}"><code>COBREXA.coupling_bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">coupling_bounds(a::MetabolicModel)::Tuple{SparseVec,SparseVec}</code></pre><p>Get the lower and upper bounds for each coupling bound in a model, as specified by <code>coupling</code>. By default, the model does not have any coupling bounds.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/abstract/MetabolicModel.jl#L153-L158">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.gene_annotations-Tuple{MetabolicModel, String}" href="#COBREXA.gene_annotations-Tuple{MetabolicModel, String}"><code>COBREXA.gene_annotations</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">gene_annotations(a::MetabolicModel, gene_id::String)::Annotations</code></pre><p>Return standardized names that identify the corresponding gene or product. The dictionary assigns vectors of possible identifiers to identifier system names, e.g. <code>&quot;PDB&quot; =&gt; [&quot;PROT01&quot;]</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/abstract/MetabolicModel.jl#L251-L257">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.gene_notes-Tuple{MetabolicModel, String}" href="#COBREXA.gene_notes-Tuple{MetabolicModel, String}"><code>COBREXA.gene_notes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">gene_notes(model::MetabolicModel, gene_id::String)::Notes</code></pre><p>Return the notes associated with the gene <code>gene_id</code> in <code>model</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/abstract/MetabolicModel.jl#L280-L284">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.genes-Tuple{MetabolicModel}" href="#COBREXA.genes-Tuple{MetabolicModel}"><code>COBREXA.genes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">genes(a::MetabolicModel)::Vector{String}</code></pre><p>Return identifiers of all genes contained in the model. By default, there are no genes.</p><p>In SBML, these are usually called &quot;gene products&quot; but we write <code>genes</code> for simplicity.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/abstract/MetabolicModel.jl#L163-L171">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_annotations-Tuple{MetabolicModel, String}" href="#COBREXA.metabolite_annotations-Tuple{MetabolicModel, String}"><code>COBREXA.metabolite_annotations</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_annotations(a::MetabolicModel, metabolite_id::String)::Annotations</code></pre><p>Return standardized names that may help to reliably identify the metabolite. The dictionary assigns vectors of possible identifiers to identifier system names, e.g. <code>&quot;ChEMBL&quot; =&gt; [&quot;123&quot;]</code> or <code>&quot;PubChem&quot; =&gt; [&quot;CID123&quot;, &quot;CID654645645&quot;]</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/abstract/MetabolicModel.jl#L240-L246">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_charge-Tuple{MetabolicModel, String}" href="#COBREXA.metabolite_charge-Tuple{MetabolicModel, String}"><code>COBREXA.metabolite_charge</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_charge(model::MetabolicModel, metabolite_id::String)::Maybe{Int}</code></pre><p>Return the charge associated with metabolite <code>metabolite_id</code> in <code>model</code>. Returns <code>nothing</code> if charge not present.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/abstract/MetabolicModel.jl#L219-L224">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_compartment-Tuple{MetabolicModel, String}" href="#COBREXA.metabolite_compartment-Tuple{MetabolicModel, String}"><code>COBREXA.metabolite_compartment</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_compartment(model::MetabolicModel, metabolite_id::String)::Maybe{String}</code></pre><p>Return the compartment of metabolite <code>metabolite_id</code> in <code>model</code> if it is assigned. If not, return <code>nothing</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/abstract/MetabolicModel.jl#L289-L294">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_formula-Tuple{MetabolicModel, String}" href="#COBREXA.metabolite_formula-Tuple{MetabolicModel, String}"><code>COBREXA.metabolite_formula</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_formula(
    a::MetabolicModel,
    metabolite_id::String,
)::Maybe{MetaboliteFormula}</code></pre><p>Return the formula of metabolite <code>metabolite_id</code> in <code>model</code>. Return <code>nothing</code> in case the formula is not known or irrelevant.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/abstract/MetabolicModel.jl#L203-L211">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_notes-Tuple{MetabolicModel, String}" href="#COBREXA.metabolite_notes-Tuple{MetabolicModel, String}"><code>COBREXA.metabolite_notes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_notes(model::MetabolicModel, metabolite_id::String)::Notes</code></pre><p>Return the notes associated with metabolite <code>reaction_id</code> in <code>model</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/abstract/MetabolicModel.jl#L271-L275">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolites-Tuple{MetabolicModel}" href="#COBREXA.metabolites-Tuple{MetabolicModel}"><code>COBREXA.metabolites</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolites(a::MetabolicModel)::Vector{String}</code></pre><p>Return a vector of metabolite identifiers in a model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/abstract/MetabolicModel.jl#L71-L75">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.n_coupling_constraints-Tuple{MetabolicModel}" href="#COBREXA.n_coupling_constraints-Tuple{MetabolicModel}"><code>COBREXA.n_coupling_constraints</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">n_coupling_constraints(a::MetabolicModel)::Int</code></pre><p>Get the number of coupling constraints in a model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/abstract/MetabolicModel.jl#L144-L148">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.n_genes-Tuple{MetabolicModel}" href="#COBREXA.n_genes-Tuple{MetabolicModel}"><code>COBREXA.n_genes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">n_genes(a::MetabolicModel)::Int</code></pre><p>Return the number of genes in the model (as returned by <a href="#COBREXA.genes-Tuple{MetabolicModel}"><code>genes</code></a>). If you just need the number of the genes, this may be much more efficient than calling <a href="#COBREXA.genes-Tuple{MetabolicModel}"><code>genes</code></a> and measuring the array.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/abstract/MetabolicModel.jl#L176-L182">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.n_metabolites-Tuple{MetabolicModel}" href="#COBREXA.n_metabolites-Tuple{MetabolicModel}"><code>COBREXA.n_metabolites</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">n_metabolites(a::MetabolicModel)::Int</code></pre><p>Get the number of metabolites in a model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/abstract/MetabolicModel.jl#L89-L93">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.n_reactions-Tuple{MetabolicModel}" href="#COBREXA.n_reactions-Tuple{MetabolicModel}"><code>COBREXA.n_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">n_reactions(a::MetabolicModel)::Int</code></pre><p>Get the number of reactions in a model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/abstract/MetabolicModel.jl#L80-L84">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.objective-Tuple{MetabolicModel}" href="#COBREXA.objective-Tuple{MetabolicModel}"><code>COBREXA.objective</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">objective(a::MetabolicModel)::SparseVec</code></pre><p>Get the objective vector of a model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/abstract/MetabolicModel.jl#L125-L129">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.precache!-Tuple{MetabolicModel}" href="#COBREXA.precache!-Tuple{MetabolicModel}"><code>COBREXA.precache!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">precache!(a::MetabolicModel)::Nothing</code></pre><p>Do whatever is feasible to get the model into a state that can be read from as-quickly-as-possible. This may include e.g. generating helper index structures and loading delayed parts of the model from disk. The model should be modified &quot;transparently&quot; in-place. Analysis functions call this right before applying modifications or converting the model to the optimization model using <a href="#COBREXA.make_optimization_model-Tuple{MetabolicModel, Any}"><code>make_optimization_model</code></a>; usually on the same machine where the optimizers (and, generally, the core analysis algorithms) will run. The calls are done in a good hope that the performance will be improved.</p><p>By default, it should be safe to do nothing.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/abstract/MetabolicModel.jl#L323-L336">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_annotations-Tuple{MetabolicModel, String}" href="#COBREXA.reaction_annotations-Tuple{MetabolicModel, String}"><code>COBREXA.reaction_annotations</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_annotations(a::MetabolicModel, reaction_id::String)::Annotations</code></pre><p>Return standardized names that may help identifying the reaction. The dictionary assigns vectors of possible identifiers to identifier system names, e.g. <code>&quot;Reactome&quot; =&gt; [&quot;reactomeID123&quot;]</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/abstract/MetabolicModel.jl#L229-L235">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_gene_association-Tuple{MetabolicModel, String}" href="#COBREXA.reaction_gene_association-Tuple{MetabolicModel, String}"><code>COBREXA.reaction_gene_association</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_gene_association(a::MetabolicModel, gene_id::String)::Maybe{GeneAssociation}</code></pre><p>Returns the sets of genes that need to be present so that the reaction can work (technically, a DNF on gene availability, with positive atoms only).</p><p>For simplicity, <code>nothing</code> may be returned, meaning that the reaction always takes place. (in DNF, that would be equivalent to returning <code>[[]]</code>.)</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/abstract/MetabolicModel.jl#L187-L195">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_notes-Tuple{MetabolicModel, String}" href="#COBREXA.reaction_notes-Tuple{MetabolicModel, String}"><code>COBREXA.reaction_notes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_notes(model::MetabolicModel, reaction_id::String)::Notes</code></pre><p>Return the notes associated with reaction <code>reaction_id</code> in <code>model</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/abstract/MetabolicModel.jl#L262-L266">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_stoichiometry-Tuple{MetabolicModel, String}" href="#COBREXA.reaction_stoichiometry-Tuple{MetabolicModel, String}"><code>COBREXA.reaction_stoichiometry</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_stoichiometry(model::MetaboliteModel, rid::String)::Dict{String, Float64}</code></pre><p>Return the stoichiometry of reaction with ID <code>rid</code> in the model. The dictionary maps the metabolite IDs to their stoichiometric coefficients.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/abstract/MetabolicModel.jl#L309-L314">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_subsystem-Tuple{MetabolicModel, String}" href="#COBREXA.reaction_subsystem-Tuple{MetabolicModel, String}"><code>COBREXA.reaction_subsystem</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_subsystem(model::MetabolicModel, reaction_id::String)::Maybe{String}</code></pre><p>Return the subsystem of reaction <code>reaction_id</code> in <code>model</code> if it is assigned. If not, return <code>nothing</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/abstract/MetabolicModel.jl#L299-L304">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reactions-Tuple{MetabolicModel}" href="#COBREXA.reactions-Tuple{MetabolicModel}"><code>COBREXA.reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reactions(a::MetabolicModel)::Vector{String}</code></pre><p>Return a vector of reaction identifiers in a model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/abstract/MetabolicModel.jl#L62-L66">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.stoichiometry-Tuple{MetabolicModel}" href="#COBREXA.stoichiometry-Tuple{MetabolicModel}"><code>COBREXA.stoichiometry</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">stoichiometry(a::MetabolicModel)::SparseMat</code></pre><p>Get the sparse stoichiometry matrix of a model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/abstract/MetabolicModel.jl#L98-L102">source</a></section></article><h2 id="Model-types-and-contents"><a class="docs-heading-anchor" href="#Model-types-and-contents">Model types and contents</a><a id="Model-types-and-contents-1"></a><a class="docs-heading-anchor-permalink" href="#Model-types-and-contents" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="COBREXA.CoreModel" href="#COBREXA.CoreModel"><code>COBREXA.CoreModel</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">struct CoreModel &lt;: MetabolicModel</code></pre><p>A &quot;bare bones&quot; core linear optimization problem of the form, with reaction and metabolite names.</p><pre><code class="language-none">min c^T x
s.t. S x = b
      xₗ ≤ x ≤ xᵤ</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/CoreModel.jl#L2-L12">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Base.convert-Union{Tuple{M}, Tuple{Type{CoreModel}, M}} where M&lt;:MetabolicModel" href="#Base.convert-Union{Tuple{M}, Tuple{Type{CoreModel}, M}} where M&lt;:MetabolicModel"><code>Base.convert</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">Base.convert(::Type{CoreModel}, m::M) where {M &lt;: MetabolicModel}</code></pre><p>Make a <code>CoreModel</code> out of any compatible model type.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/CoreModel.jl#L100-L104">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.balance-Tuple{CoreModel}" href="#COBREXA.balance-Tuple{CoreModel}"><code>COBREXA.balance</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">balance(a::CoreModel)::SparseVec</code></pre><p><code>CoreModel</code> target flux balance.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/CoreModel.jl#L70-L74">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.bounds-Tuple{CoreModel}" href="#COBREXA.bounds-Tuple{CoreModel}"><code>COBREXA.bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">bounds(a::CoreModel)::Tuple{SparseVec,SparseVec}</code></pre><p><code>CoreModel</code> flux bounds.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/CoreModel.jl#L63-L67">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolites-Tuple{CoreModel}" href="#COBREXA.metabolites-Tuple{CoreModel}"><code>COBREXA.metabolites</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolites(a::CoreModel)::Vector{String}</code></pre><p>Metabolites in a <code>CoreModel</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/CoreModel.jl#L49-L53">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.objective-Tuple{CoreModel}" href="#COBREXA.objective-Tuple{CoreModel}"><code>COBREXA.objective</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">objective(a::CoreModel)::SparseVec</code></pre><p><code>CoreModel</code> objective vector.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/CoreModel.jl#L77-L81">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_stoichiometry-Tuple{CoreModel, Any}" href="#COBREXA.reaction_stoichiometry-Tuple{CoreModel, Any}"><code>COBREXA.reaction_stoichiometry</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_stoichiometry(model::CoreModel, ridx)::Dict{String, Float64}</code></pre><p>Return the stoichiometry of reaction at index <code>ridx</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/CoreModel.jl#L92-L96">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_stoichiometry-Tuple{CoreModel, String}" href="#COBREXA.reaction_stoichiometry-Tuple{CoreModel, String}"><code>COBREXA.reaction_stoichiometry</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_stoichiometry(model::CoreModel, rid::String)::Dict{String, Float64}</code></pre><p>Return the stoichiometry of reaction with ID <code>rid</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/CoreModel.jl#L84-L88">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reactions-Tuple{CoreModel}" href="#COBREXA.reactions-Tuple{CoreModel}"><code>COBREXA.reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reactions(a::CoreModel)::Vector{String}</code></pre><p>Get the reactions in a <code>CoreModel</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/CoreModel.jl#L42-L46">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.stoichiometry-Tuple{CoreModel}" href="#COBREXA.stoichiometry-Tuple{CoreModel}"><code>COBREXA.stoichiometry</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">stoichiometry(a::CoreModel)::SparseMat</code></pre><p><code>CoreModel</code> stoichiometry matrix.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/CoreModel.jl#L56-L60">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.CoreModelCoupled" href="#COBREXA.CoreModelCoupled"><code>COBREXA.CoreModelCoupled</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">struct CoreModelCoupled &lt;: MetabolicModel</code></pre><p>The linear model with additional coupling constraints in the form</p><pre><code class="language-none">    cₗ ≤ C x ≤ cᵤ</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/CoreModelCoupled.jl#L2-L9">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Base.convert-Tuple{Type{CoreModelCoupled}, MetabolicModel}" href="#Base.convert-Tuple{Type{CoreModelCoupled}, MetabolicModel}"><code>Base.convert</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">Base.convert(::Type{CoreModelCoupled}, mm::MetabolicModel)</code></pre><p>Make a <code>CoreModelCoupled</code> out of any compatible model type.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/CoreModelCoupled.jl#L117-L121">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.balance-Tuple{CoreModelCoupled}" href="#COBREXA.balance-Tuple{CoreModelCoupled}"><code>COBREXA.balance</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">balance(a::CoreModelCoupled)</code></pre><p>Extract balance from <a href="#COBREXA.CoreModelCoupled"><code>CoreModelCoupled</code></a> (uses the internal <a href="#COBREXA.CoreModel"><code>CoreModel</code></a>).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/CoreModelCoupled.jl#L64-L69">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.bounds-Tuple{CoreModelCoupled}" href="#COBREXA.bounds-Tuple{CoreModelCoupled}"><code>COBREXA.bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">bounds(a::CoreModelCoupled)</code></pre><p>Extract bounds from <a href="#COBREXA.CoreModelCoupled"><code>CoreModelCoupled</code></a> (uses the internal <a href="#COBREXA.CoreModel"><code>CoreModel</code></a>).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/CoreModelCoupled.jl#L56-L61">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.coupling-Tuple{CoreModelCoupled}" href="#COBREXA.coupling-Tuple{CoreModelCoupled}"><code>COBREXA.coupling</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">coupling(a::CoreModelCoupled)::SparseMat</code></pre><p>Coupling constraint matrix for a <code>CoreModelCoupled</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/CoreModelCoupled.jl#L80-L84">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.coupling_bounds-Tuple{CoreModelCoupled}" href="#COBREXA.coupling_bounds-Tuple{CoreModelCoupled}"><code>COBREXA.coupling_bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">coupling_bounds(a::CoreModelCoupled)::Tuple{SparseVec,SparseVec}</code></pre><p>Coupling bounds for a <code>CoreModelCoupled</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/CoreModelCoupled.jl#L94-L98">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolites-Tuple{CoreModelCoupled}" href="#COBREXA.metabolites-Tuple{CoreModelCoupled}"><code>COBREXA.metabolites</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolites(a::CoreModelCoupled)</code></pre><p>Extract metabolites from <a href="#COBREXA.CoreModelCoupled"><code>CoreModelCoupled</code></a> (uses the internal <a href="#COBREXA.CoreModel"><code>CoreModel</code></a>).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/CoreModelCoupled.jl#L40-L45">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.n_coupling_constraints-Tuple{CoreModelCoupled}" href="#COBREXA.n_coupling_constraints-Tuple{CoreModelCoupled}"><code>COBREXA.n_coupling_constraints</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">n_coupling_constraints(a::CoreModelCoupled)::Int</code></pre><p>The number of coupling constraints in a <code>CoreModelCoupled</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/CoreModelCoupled.jl#L87-L91">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.objective-Tuple{CoreModelCoupled}" href="#COBREXA.objective-Tuple{CoreModelCoupled}"><code>COBREXA.objective</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">objective(a::CoreModelCoupled)</code></pre><p>Extract objective from <a href="#COBREXA.CoreModelCoupled"><code>CoreModelCoupled</code></a> (uses the internal <a href="#COBREXA.CoreModel"><code>CoreModel</code></a>).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/CoreModelCoupled.jl#L72-L77">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_stoichiometry-Tuple{CoreModelCoupled, Any}" href="#COBREXA.reaction_stoichiometry-Tuple{CoreModelCoupled, Any}"><code>COBREXA.reaction_stoichiometry</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_stoichiometry(model::CoreModelCoupled, ridx)::Dict{String, Float64}</code></pre><p>Return the stoichiometry of reaction at index <code>ridx</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/CoreModelCoupled.jl#L108-L112">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_stoichiometry-Tuple{CoreModelCoupled, String}" href="#COBREXA.reaction_stoichiometry-Tuple{CoreModelCoupled, String}"><code>COBREXA.reaction_stoichiometry</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_stoichiometry(model::CoreModelCoupled, rid::String)::Dict{String, Float64}</code></pre><p>Return the stoichiometry of reaction with ID <code>rid</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/CoreModelCoupled.jl#L101-L105">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reactions-Tuple{CoreModelCoupled}" href="#COBREXA.reactions-Tuple{CoreModelCoupled}"><code>COBREXA.reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reactions(a::CoreModelCoupled)</code></pre><p>Extract reactions from <a href="#COBREXA.CoreModelCoupled"><code>CoreModelCoupled</code></a> (uses the internal <a href="#COBREXA.CoreModel"><code>CoreModel</code></a>).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/CoreModelCoupled.jl#L32-L37">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.stoichiometry-Tuple{CoreModelCoupled}" href="#COBREXA.stoichiometry-Tuple{CoreModelCoupled}"><code>COBREXA.stoichiometry</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">stoichiometry(a::CoreModelCoupled)</code></pre><p>Extract stoichiometry from <a href="#COBREXA.CoreModelCoupled"><code>CoreModelCoupled</code></a> (uses the internal <a href="#COBREXA.CoreModel"><code>CoreModel</code></a>).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/CoreModelCoupled.jl#L48-L53">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.FluxSummary" href="#COBREXA.FluxSummary"><code>COBREXA.FluxSummary</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">FluxSummary</code></pre><p>A struct used to store summary information about the solution of a constraint based analysis result.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/FluxSummary.jl#L1-L6">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.FluxSummary-Tuple{}" href="#COBREXA.FluxSummary-Tuple{}"><code>COBREXA.FluxSummary</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">FluxSummary()</code></pre><p>A default empty constructor for <code>FluxSummary</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/FluxSummary.jl#L14-L18">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.flux_summary-Tuple{Union{Nothing, Dict{String, Float64}}}" href="#COBREXA.flux_summary-Tuple{Union{Nothing, Dict{String, Float64}}}"><code>COBREXA.flux_summary</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">flux_summary(flux_result::Dict{String, Float64};
            exclude_exchanges = false,
            exchange_prefixes = _constants.exchange_prefixes,
            biomass_strings = _constants.biomass_strings,
            exclude_biomass = false,
            small_flux_bound = 1.0/_constants.default_reaction_bound^2,
            large_flux_bound = _constants.default_reaction_bound,
            keep_unbounded = false,
            )::FluxSummary</code></pre><p>Summarize a dictionary of fluxes into small, useful representation of the most important information contained. Useful for pretty-printing and quickly exploring the results. Internally this function uses <a href="#COBREXA.looks_like_biomass_reaction-Tuple{String}"><code>looks_like_biomass_reaction</code></a> and <a href="#COBREXA.looks_like_exchange_reaction-Tuple{String}"><code>looks_like_exchange_reaction</code></a>. The corresponding keyword arguments passed to these functions. Use this if your model has non-standard ids for reactions. Fluxes smaller than <code>small_flux_bound</code> are not stored, while fluxes larger than <code>large_flux_bound</code> are only stored if <code>keep_unbounded</code> is <code>true</code>.</p><p><strong>Example</strong></p><pre><code class="language-none">julia&gt; sol = flux_balance_analysis_dict(model, Tulip.Optimizer)
julia&gt; fr = flux_summary(sol)
Biomass:
  BIOMASS_Ecoli_core_w_GAM: 0.8739
Import:
  EX_o2_e:     -21.7995
  EX_glc__D_e: -10.0
  EX_nh4_e:    -4.7653
  EX_pi_e:     -3.2149
Export:
  EX_h_e:      17.5309
  EX_co2_e:    22.8098
  EX_h2o_e:    29.1758</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/FluxSummary.jl#L28-L64">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.FluxVariabilitySummary" href="#COBREXA.FluxVariabilitySummary"><code>COBREXA.FluxVariabilitySummary</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">FluxVariabilitySummary</code></pre><p>A struct used to store summary information about the solution of a flux variability analysis result.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/FluxVariabilitySummary.jl#L1-L6">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.FluxVariabilitySummary-Tuple{}" href="#COBREXA.FluxVariabilitySummary-Tuple{}"><code>COBREXA.FluxVariabilitySummary</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">FluxVariabilitySummary()</code></pre><p>A default empty constructor for <code>FluxVariabilitySummary</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/FluxVariabilitySummary.jl#L12-L16">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.flux_variability_summary-Tuple{Tuple{Dict{String, Dict{String, Float64}}, Dict{String, Dict{String, Float64}}}}" href="#COBREXA.flux_variability_summary-Tuple{Tuple{Dict{String, Dict{String, Float64}}, Dict{String, Dict{String, Float64}}}}"><code>COBREXA.flux_variability_summary</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">flux_variability_summary(flux_result::Tuple{Dict{String, Dict{String, Float64}}, Dict{String, Dict{String, Float64}}};
    exclude_exchanges = false,
    exchange_prefixes = _constants.exchange_prefixes,
    biomass_strings = _constants.biomass_strings,
    exclude_biomass = false,
    )::FluxVariabilitySummary</code></pre><p>Summarize a dictionary of flux dictionaries obtained eg. from flux<em>variability</em>analysis<em>dict. The simplified summary representation is useful for pretty-printing and easily showing the most important results. Internally this function uses [`looks</em>like<em>biomass</em>reaction<code>](@ref) and [</code>looks<em>like</em>exchange_reaction`](@ref). The corresponding keyword arguments passed to these functions. Use this if your model has non-standard ids for reactions.</p><p><strong>Example</strong></p><pre><code class="language-none">julia&gt; sol = flux_variability_analysis_dict(model, Gurobi.Optimizer; bounds = objective_bounds(0.99))
julia&gt; flux_res = flux_variability_summary(sol)
Biomass                     Lower bound   Upper bound
  BIOMASS_Ecoli_core_w_GAM: 0.8652        0.8652
Exchange
  EX_h2o_e:                 28.34         28.34
  EX_co2_e:                 22.0377       22.0377
  EX_o2_e:                  -22.1815      -22.1815
  EX_h_e:                   17.3556       17.3556
  EX_glc__D_e:              -10.0         -10.0
  EX_nh4_e:                 -4.8448       -4.8448
  EX_pi_e:                  -3.2149       -3.2149
  EX_for_e:                 0.0           0.0
  ...                       ...           ...</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/FluxVariabilitySummary.jl#L24-L57">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.Gene" href="#COBREXA.Gene"><code>COBREXA.Gene</code></a> — <span class="docstring-category">Type</span></header><section><div><p>Gene struct.</p><p><strong>Fields</strong></p><pre><code class="language-none">id :: String
notes :: Dict{String, Vector{String}}
annotation :: Dict{String, Union{Vector{String}, String}}</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/Gene.jl#L1-L10">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.JSONModel" href="#COBREXA.JSONModel"><code>COBREXA.JSONModel</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">struct JSONModel &lt;: MetabolicModel
    json::Dict{String,Any}
    rxn_index::Dict{String,Int}
    rxns::Vector{Any}
    met_index::Dict{String,Int}
    mets::Vector{Any}
    gene_index::Dict{String,Int}
    genes::Vector{Any}
end</code></pre><p>A struct used to store the contents of a JSON model, i.e. a model read from a file ending with <code>.json</code>. These model files typically store all the model data in arrays of JSON objects (represented in Julia as vectors of dictionaries).</p><p>Usually, not all of the fields of the input JSON can be easily represented when converting to other models, care should be taken to avoid losing information.</p><p>Direct work with the <code>json</code> structure is not very efficient; the model structure therefore caches some of the internal structure in the extra fields. The single-parameter <a href="#COBREXA.JSONModel"><code>JSONModel</code></a> constructor creates these caches correctly from the <code>json</code>. The model structure is designed as read-only, and changes in <code>json</code> invalidate the cache.</p><p><strong>Example</strong></p><pre><code class="language-none">model = load_json_model(&quot;some_model.json&quot;)
model.json # see the actual underlying JSON
reactions(model) # see the list of reactions</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/JSONModel.jl#L1-L31">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Base.convert-Tuple{Type{JSONModel}, MetabolicModel}" href="#Base.convert-Tuple{Type{JSONModel}, MetabolicModel}"><code>Base.convert</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">Base.convert(::Type{JSONModel}, mm::MetabolicModel)</code></pre><p>Convert any <a href="#COBREXA.MetabolicModel"><code>MetabolicModel</code></a> to <a href="#COBREXA.JSONModel"><code>JSONModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/JSONModel.jl#L269-L273">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.bounds-Tuple{JSONModel}" href="#COBREXA.bounds-Tuple{JSONModel}"><code>COBREXA.bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">bounds(model::JSONModel)</code></pre><p>Get the bounds for reactions, assuming the information is stored in <code>.lower_bound</code> and <code>.upper_bound</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/JSONModel.jl#L142-L147">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.gene_annotations-Tuple{JSONModel, String}" href="#COBREXA.gene_annotations-Tuple{JSONModel, String}"><code>COBREXA.gene_annotations</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">gene_annotations(model::JSONModel, gid::String)::Annotations</code></pre><p>Gene annotations from the <a href="#COBREXA.JSONModel"><code>JSONModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/JSONModel.jl#L207-L211">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.gene_notes-Tuple{JSONModel, String}" href="#COBREXA.gene_notes-Tuple{JSONModel, String}"><code>COBREXA.gene_notes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">gene_notes(model::JSONModel, gid::String)::Notes</code></pre><p>Gene notes from the <a href="#COBREXA.JSONModel"><code>JSONModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/JSONModel.jl#L217-L221">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.genes-Tuple{JSONModel}" href="#COBREXA.genes-Tuple{JSONModel}"><code>COBREXA.genes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">genes(model::JSONModel)</code></pre><p>Extract gene names from a JSON model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/JSONModel.jl#L93-L97">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_annotations-Tuple{JSONModel, String}" href="#COBREXA.metabolite_annotations-Tuple{JSONModel, String}"><code>COBREXA.metabolite_annotations</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_annotations(model::JSONModel, mid::String)::Annotations</code></pre><p>Metabolite annotations from the <a href="#COBREXA.JSONModel"><code>JSONModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/JSONModel.jl#L243-L247">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_charge-Tuple{JSONModel, String}" href="#COBREXA.metabolite_charge-Tuple{JSONModel, String}"><code>COBREXA.metabolite_charge</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_charge(model::JSONModel, mid::String)</code></pre><p>Return the metabolite <code>.charge</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/JSONModel.jl#L191-L195">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_compartment-Tuple{JSONModel, String}" href="#COBREXA.metabolite_compartment-Tuple{JSONModel, String}"><code>COBREXA.metabolite_compartment</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_compartment(model::JSONModel, mid::String)</code></pre><p>Return the metabolite <code>.compartment</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/JSONModel.jl#L199-L203">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_formula-Tuple{JSONModel, String}" href="#COBREXA.metabolite_formula-Tuple{JSONModel, String}"><code>COBREXA.metabolite_formula</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_formula(model::JSONModel, mid::String)</code></pre><p>Parse and return the metabolite <code>.formula</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/JSONModel.jl#L183-L187">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_notes-Tuple{JSONModel, String}" href="#COBREXA.metabolite_notes-Tuple{JSONModel, String}"><code>COBREXA.metabolite_notes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_notes(model::JSONModel, mid::String)::Notes</code></pre><p>Metabolite notes from the <a href="#COBREXA.JSONModel"><code>JSONModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/JSONModel.jl#L253-L257">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolites-Tuple{JSONModel}" href="#COBREXA.metabolites-Tuple{JSONModel}"><code>COBREXA.metabolites</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolites(model::JSONModel)</code></pre><p>Extract metabolite names (stored as <code>.id</code>) from JSON model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/JSONModel.jl#L86-L90">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.objective-Tuple{JSONModel}" href="#COBREXA.objective-Tuple{JSONModel}"><code>COBREXA.objective</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">objective(model::JSONModel)</code></pre><p>Collect <code>.objective_coefficient</code> keys from model reactions.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/JSONModel.jl#L157-L161">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_annotations-Tuple{JSONModel, String}" href="#COBREXA.reaction_annotations-Tuple{JSONModel, String}"><code>COBREXA.reaction_annotations</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_annotations(model::JSONModel, rid::String)::Annotations</code></pre><p>Reaction annotations from the <a href="#COBREXA.JSONModel"><code>JSONModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/JSONModel.jl#L225-L229">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_gene_association-Tuple{JSONModel, String}" href="#COBREXA.reaction_gene_association-Tuple{JSONModel, String}"><code>COBREXA.reaction_gene_association</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_gene_associaton(model::JSONModel, rid::String)</code></pre><p>Parses the <code>.gene_reaction_rule</code> from reactions.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/JSONModel.jl#L165-L169">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_notes-Tuple{JSONModel, String}" href="#COBREXA.reaction_notes-Tuple{JSONModel, String}"><code>COBREXA.reaction_notes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_notes(model::JSONModel, rid::String)::Notes</code></pre><p>Reaction notes from the <a href="#COBREXA.JSONModel"><code>JSONModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/JSONModel.jl#L235-L239">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_stoichiometry-Tuple{JSONModel, String}" href="#COBREXA.reaction_stoichiometry-Tuple{JSONModel, String}"><code>COBREXA.reaction_stoichiometry</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_stoichiometry(model::JSONModel, rid::String)::Dict{String, Float64}</code></pre><p>Return the stoichiometry of reaction with ID <code>rid</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/JSONModel.jl#L261-L265">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_subsystem-Tuple{JSONModel, String}" href="#COBREXA.reaction_subsystem-Tuple{JSONModel, String}"><code>COBREXA.reaction_subsystem</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_subsystem(model::JSONModel, rid::String)</code></pre><p>Parses the <code>.subsystem</code> out from reactions.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/JSONModel.jl#L175-L179">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reactions-Tuple{JSONModel}" href="#COBREXA.reactions-Tuple{JSONModel}"><code>COBREXA.reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reactions(model::JSONModel)</code></pre><p>Extract reaction names (stored as <code>.id</code>) from JSON model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/JSONModel.jl#L79-L83">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.stoichiometry-Tuple{JSONModel}" href="#COBREXA.stoichiometry-Tuple{JSONModel}"><code>COBREXA.stoichiometry</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">stoichiometry(model::JSONModel)</code></pre><p>Get the stoichiometry. Assuming the information is stored in reaction object under key <code>.metabolites</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/JSONModel.jl#L100-L105">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.MATModel" href="#COBREXA.MATModel"><code>COBREXA.MATModel</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">struct MATModel</code></pre><p>Wrapper around the models loaded in dictionaries from the MATLAB representation.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/MATModel.jl#L1-L5">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Base.convert-Tuple{Type{MATModel}, MetabolicModel}" href="#Base.convert-Tuple{Type{MATModel}, MetabolicModel}"><code>Base.convert</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">Base.convert(::Type{MATModel}, m::MetabolicModel)</code></pre><p>Convert any metabolic model to <code>MATModel</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/MATModel.jl#L198-L202">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA._mat_has_squashed_coupling-Tuple{Any}" href="#COBREXA._mat_has_squashed_coupling-Tuple{Any}"><code>COBREXA._mat_has_squashed_coupling</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">_mat_has_squashed_coupling(mat)</code></pre><p>Guesses whether C in the MAT file is stored in A=[S;C].</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/MATModel.jl#L26-L30">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.balance-Tuple{MATModel}" href="#COBREXA.balance-Tuple{MATModel}"><code>COBREXA.balance</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">balance(m::MATModel)</code></pre><p>Extracts balance from the MAT model, defaulting to zeroes if not present.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/MATModel.jl#L66-L70">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.bounds-Tuple{MATModel}" href="#COBREXA.bounds-Tuple{MATModel}"><code>COBREXA.bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">bounds(m::MATModel)</code></pre><p>Extracts bounds from the MAT file, saved under <code>lb</code> and <code>ub</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/MATModel.jl#L56-L60">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.coupling-Tuple{MATModel}" href="#COBREXA.coupling-Tuple{MATModel}"><code>COBREXA.coupling</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">coupling(m::MATModel)</code></pre><p>Extract coupling matrix stored, in <code>C</code> key.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/MATModel.jl#L87-L91">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.coupling_bounds-Tuple{MATModel}" href="#COBREXA.coupling_bounds-Tuple{MATModel}"><code>COBREXA.coupling_bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">coupling_bounds(m::MATModel)</code></pre><p>Extracts the coupling constraints. Currently, there are several accepted ways to store these in MATLAB models; this takes the constraints from vectors <code>cl</code> and <code>cu</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/MATModel.jl#L96-L100">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.genes-Tuple{MATModel}" href="#COBREXA.genes-Tuple{MATModel}"><code>COBREXA.genes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">genes(m::MATModel)</code></pre><p>Extracts the possible gene list from <code>genes</code> key.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/MATModel.jl#L116-L120">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_charge-Tuple{MATModel, String}" href="#COBREXA.metabolite_charge-Tuple{MATModel, String}"><code>COBREXA.metabolite_charge</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_charge(m::MATModel, mid::String)</code></pre><p>Extract metabolite charge from <code>metCharge</code> or <code>metCharges</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/MATModel.jl#L150-L154">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_compartment-Tuple{MATModel, String}" href="#COBREXA.metabolite_compartment-Tuple{MATModel, String}"><code>COBREXA.metabolite_compartment</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_compartment(m::MATModel, mid::String)</code></pre><p>Extract metabolite compartment from <code>metCompartment</code> or <code>metCompartments</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/MATModel.jl#L160-L164">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_formula-Tuple{MATModel, String}" href="#COBREXA.metabolite_formula-Tuple{MATModel, String}"><code>COBREXA.metabolite_formula</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_formula(m::MATModel, mid::String)</code></pre><p>Extract metabolite formula from key <code>metFormula</code> or <code>metFormulas</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/MATModel.jl#L140-L144">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolites-Tuple{MATModel}" href="#COBREXA.metabolites-Tuple{MATModel}"><code>COBREXA.metabolites</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolites(m::MATModel)::Vector{String}</code></pre><p>Extracts metabolite names from <code>mets</code> key in the MAT file.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/MATModel.jl#L35-L39">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.objective-Tuple{MATModel}" href="#COBREXA.objective-Tuple{MATModel}"><code>COBREXA.objective</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">objective(m::MATModel)</code></pre><p>Extracts the objective from the MAT model (defaults to zeroes).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/MATModel.jl#L79-L83">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_gene_association-Tuple{MATModel, String}" href="#COBREXA.reaction_gene_association-Tuple{MATModel, String}"><code>COBREXA.reaction_gene_association</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_gene_association(m::MATModel, rid::String)</code></pre><p>Extracts the associations from <code>grRules</code> key, if present.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/MATModel.jl#L126-L130">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_stoichiometry-Tuple{MATModel, Any}" href="#COBREXA.reaction_stoichiometry-Tuple{MATModel, Any}"><code>COBREXA.reaction_stoichiometry</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_stoichiometry(model::MATModel, ridx)::Dict{String, Float64}</code></pre><p>Return the stoichiometry of reaction at index <code>ridx</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/MATModel.jl#L181-L185">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_stoichiometry-Tuple{MATModel, String}" href="#COBREXA.reaction_stoichiometry-Tuple{MATModel, String}"><code>COBREXA.reaction_stoichiometry</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_stoichiometry(model::MATModel, rid::String)::Dict{String, Float64}</code></pre><p>Return the stoichiometry of reaction with ID <code>rid</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/MATModel.jl#L171-L175">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reactions-Tuple{MATModel}" href="#COBREXA.reactions-Tuple{MATModel}"><code>COBREXA.reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reactions(m::MATModel)::Vector{String}</code></pre><p>Extracts reaction names from <code>rxns</code> key in the MAT file.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/MATModel.jl#L13-L17">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.stoichiometry-Tuple{MATModel}" href="#COBREXA.stoichiometry-Tuple{MATModel}"><code>COBREXA.stoichiometry</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">stoichiometry(m::MATModel)</code></pre><p>Extract the stoichiometry matrix, stored under key <code>S</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/MATModel.jl#L49-L53">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.Metabolite" href="#COBREXA.Metabolite"><code>COBREXA.Metabolite</code></a> — <span class="docstring-category">Type</span></header><section><div><p>Metabolite structure.</p><p><strong>Fields</strong></p><pre><code class="language-none">id :: String
formula :: String
charge :: Int
compartment :: String
notes :: Dict{String, Vector{String}}
annotation :: Dict{String, Union{Vector{String}, String}}</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/Metabolite.jl#L1-L13">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.Reaction" href="#COBREXA.Reaction"><code>COBREXA.Reaction</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">mutable struct Reaction
    id::String
    metabolites::Dict{String,Float64}
    lb::Float64
    ub::Float64
    grr::Maybe{GeneAssociation}
    subsystem::Maybe{String}
    notes::Notes
    annotations::Annotations
    objective_coefficient::Float64
end</code></pre><p>A structure for representing a single reaction in a <a href="#COBREXA.StandardModel"><code>StandardModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/Reaction.jl#L1-L15">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.Reaction" href="#COBREXA.Reaction"><code>COBREXA.Reaction</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">Reaction(
    id = &quot;&quot;;
    metabolites = Dict{String,Float64}(),
    lb = -_constants.default_reaction_bound,
    ub = _constants.default_reaction_bound,
    grr = nothing,
    subsystem = nothing,
    notes = Notes(),
    annotations = Annotations(),
    objective_coefficient = 0.0,
)</code></pre><p>A constructor for Reaction that only takes a reaction <code>id</code> and assigns default/uninformative values to all the fields that are not explicitely assigned.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/Reaction.jl#L28-L44">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.Reaction" href="#COBREXA.Reaction"><code>COBREXA.Reaction</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">Reaction(
    id::String,
    metabolites::Dict{String,Union{Int, Float64}},
    dir = :bidirectional;
    default_bound = _constants.default_reaction_bound,
)</code></pre><p>Convenience constructor for <code>Reaction</code>. The reaction equation is specified using <code>metabolites</code>, which is a dictionary mapping metabolite ids to stoichiometric coefficients. The direcion of the reaction is set through <code>dir</code> which can take <code>:bidirectional</code>, <code>:forward</code>, and <code>:reverse</code> as values. Finally, the <code>default_bound</code> is the value taken to mean infinity in the context of constraint based models, often this is set to a very high flux value like 1000.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/Reaction.jl#L70-L84">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.ReactionStatus" href="#COBREXA.ReactionStatus"><code>COBREXA.ReactionStatus</code></a> — <span class="docstring-category">Type</span></header><section><div><p>Used for concise reporting of modeling results.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/ReactionStatus.jl#L1-L3">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.SBMLModel" href="#COBREXA.SBMLModel"><code>COBREXA.SBMLModel</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">struct SBMLModel</code></pre><p>Thin wrapper around the model from SBML.jl library. Allows easy conversion from SBML to any other model format.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/SBMLModel.jl#L1-L6">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Base.convert-Tuple{Type{SBMLModel}, MetabolicModel}" href="#Base.convert-Tuple{Type{SBMLModel}, MetabolicModel}"><code>Base.convert</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">Base.convert(::Type{SBMLModel}, mm::MetabolicModel)</code></pre><p>Convert any metabolic model to <a href="#COBREXA.SBMLModel"><code>SBMLModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/SBMLModel.jl#L154-L158">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.balance-Tuple{SBMLModel}" href="#COBREXA.balance-Tuple{SBMLModel}"><code>COBREXA.balance</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">balance(model::SBMLModel)::SparseVec</code></pre><p>Balance vector of a <a href="#COBREXA.SBMLModel"><code>SBMLModel</code></a>. This is always zero.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/SBMLModel.jl#L73-L77">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.bounds-Tuple{SBMLModel}" href="#COBREXA.bounds-Tuple{SBMLModel}"><code>COBREXA.bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">bounds(model::SBMLModel)::Tuple{SparseVec,SparseVec}</code></pre><p>Get the lower and upper flux bounds of model <a href="#COBREXA.SBMLModel"><code>SBMLModel</code></a>. Throws <code>DomainError</code> in case if the SBML contains mismatching units.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/SBMLModel.jl#L49-L54">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.genes-Tuple{SBMLModel}" href="#COBREXA.genes-Tuple{SBMLModel}"><code>COBREXA.genes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">genes(model::SBMLModel)::Vector{String}</code></pre><p>Get genes of a <a href="#COBREXA.SBMLModel"><code>SBMLModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/SBMLModel.jl#L87-L91">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_charge-Tuple{SBMLModel, String}" href="#COBREXA.metabolite_charge-Tuple{SBMLModel, String}"><code>COBREXA.metabolite_charge</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_charge(model::SBMLModel, mid::String)::Maybe{Int}</code></pre><p>Get charge of a chosen metabolite from <a href="#COBREXA.SBMLModel"><code>SBMLModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/SBMLModel.jl#L117-L121">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_formula-Tuple{SBMLModel, String}" href="#COBREXA.metabolite_formula-Tuple{SBMLModel, String}"><code>COBREXA.metabolite_formula</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_formula(model::SBMLModel, mid::String)::Maybe{MetaboliteFormula}</code></pre><p>Get <a href="#COBREXA.MetaboliteFormula"><code>MetaboliteFormula</code></a> from a chosen metabolite from <a href="#COBREXA.SBMLModel"><code>SBMLModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/SBMLModel.jl#L109-L113">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolites-Tuple{SBMLModel}" href="#COBREXA.metabolites-Tuple{SBMLModel}"><code>COBREXA.metabolites</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolites(model::SBMLModel)::Vector{String}</code></pre><p>Get metabolites from a <a href="#COBREXA.SBMLModel"><code>SBMLModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/SBMLModel.jl#L18-L22">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.n_genes-Tuple{SBMLModel}" href="#COBREXA.n_genes-Tuple{SBMLModel}"><code>COBREXA.n_genes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">n_genes(model::SBMLModel)::Int</code></pre><p>Get number of genes in <a href="#COBREXA.SBMLModel"><code>SBMLModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/SBMLModel.jl#L94-L98">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.n_metabolites-Tuple{SBMLModel}" href="#COBREXA.n_metabolites-Tuple{SBMLModel}"><code>COBREXA.n_metabolites</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">n_metabolites(model::SBMLModel)::Int</code></pre><p>Efficient counting of metabolites in <a href="#COBREXA.SBMLModel"><code>SBMLModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/SBMLModel.jl#L32-L36">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.n_reactions-Tuple{SBMLModel}" href="#COBREXA.n_reactions-Tuple{SBMLModel}"><code>COBREXA.n_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">n_reactions(model::SBMLModel)::Int</code></pre><p>Efficient counting of reactions in <a href="#COBREXA.SBMLModel"><code>SBMLModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/SBMLModel.jl#L25-L29">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.objective-Tuple{SBMLModel}" href="#COBREXA.objective-Tuple{SBMLModel}"><code>COBREXA.objective</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">objective(model::SBMLModel)::SparseVec</code></pre><p>Objective of the <a href="#COBREXA.SBMLModel"><code>SBMLModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/SBMLModel.jl#L80-L84">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_gene_association-Tuple{SBMLModel, String}" href="#COBREXA.reaction_gene_association-Tuple{SBMLModel, String}"><code>COBREXA.reaction_gene_association</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_gene_association(model::SBMLModel, rid::String)::Maybe{GeneAssociation}</code></pre><p>Retrieve the <a href="#COBREXA.GeneAssociation"><code>GeneAssociation</code></a> from <a href="#COBREXA.SBMLModel"><code>SBMLModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/SBMLModel.jl#L101-L105">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_stoichiometry-Tuple{SBMLModel, String}" href="#COBREXA.reaction_stoichiometry-Tuple{SBMLModel, String}"><code>COBREXA.reaction_stoichiometry</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_stoichiometry(model::SBMLModel, rid::String)::Dict{String, Float64}</code></pre><p>Return the stoichiometry of reaction with ID <code>rid</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/SBMLModel.jl#L138-L142">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reactions-Tuple{SBMLModel}" href="#COBREXA.reactions-Tuple{SBMLModel}"><code>COBREXA.reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reactions(model::SBMLModel)::Vector{String}</code></pre><p>Get reactions from a <a href="#COBREXA.SBMLModel"><code>SBMLModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/SBMLModel.jl#L11-L15">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.stoichiometry-Tuple{SBMLModel}" href="#COBREXA.stoichiometry-Tuple{SBMLModel}"><code>COBREXA.stoichiometry</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">stoichiometry(model::SBMLModel)::SparseMat</code></pre><p>Recreate the stoichiometry matrix from the <a href="#COBREXA.SBMLModel"><code>SBMLModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/SBMLModel.jl#L39-L43">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.Serialized" href="#COBREXA.Serialized"><code>COBREXA.Serialized</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">mutable struct Serialized{M &lt;: MetabolicModel}
    m::Maybe{M}
    filename::String
end</code></pre><p>A meta-model that represents a model that is serialized on the disk. The internal model will be loaded on-demand by using any accessor, or by calling <a href="#COBREXA.precache!-Tuple{MetabolicModel}"><code>precache!</code></a> directly.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/Serialized.jl#L2-L11">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.precache!-Tuple{Serialized}" href="#COBREXA.precache!-Tuple{Serialized}"><code>COBREXA.precache!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">precache!(model::Serialized{MetabolicModel})::Nothing</code></pre><p>Load the <code>Serialized</code> model from disk in case it&#39;s not alreadly loaded.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/Serialized.jl#L52-L56">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.StandardModel" href="#COBREXA.StandardModel"><code>COBREXA.StandardModel</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">mutable struct StandardModel</code></pre><p><code>StandardModel</code> is used to store a constraint based metabolic model with meta-information.  Meta-information is defined as annotation details, which include gene-reaction-rules, formulas, etc.</p><p>This model type seeks to keep as much meta-information as possible, as opposed to <code>CoreModel</code> and <code>CoreModelCoupled</code>, which keep the bare neccessities only. When merging models and keeping meta-information is important, use this as the model type.  If meta-information is not important, use the more efficient core model types.  See <a href="#COBREXA.CoreModel"><code>CoreModel</code></a> and <a href="#COBREXA.CoreModelCoupled"><code>CoreModelCoupled</code></a> for comparison.</p><p>In this model, reactions, metabolites, and genes are stored in ordered dictionaries indexed by each struct&#39;s <code>id</code> field.  For example, <code>model.reactions[&quot;rxn1_id&quot;]</code> returns a <code>Reaction</code> where the field <code>id</code> equals <code>&quot;rxn1_id&quot;</code>.  This makes adding and removing reactions efficient.</p><p>Note that the stoichiometric matrix (or any other core data, e.g. flux bounds) is not stored directly as in <code>CoreModel</code>.  When this model type is used in analysis functions, these core data structures are built from scratch each time an analysis function is called.  This can cause performance issues if you run many small analysis functions sequentially.  Consider using the core model types if performance is critical.</p><p>See also: <a href="#COBREXA.Reaction"><code>Reaction</code></a>, <a href="#COBREXA.Metabolite"><code>Metabolite</code></a>, <a href="#COBREXA.Gene"><code>Gene</code></a></p><p><strong>Fields</strong></p><pre><code class="language-none">id :: String
reactions :: OrderedDict{String, Reaction}
metabolites :: OrderedDict{String, Metabolite}
genes :: OrderedDict{String, Gene}</code></pre><p><strong>Example</strong></p><pre><code class="language-none">model = load_model(StandardModel, &quot;my_model.json&quot;)
keys(model.reactions)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/StandardModel.jl#L1-L42">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Base.convert-Tuple{Type{StandardModel}, MetabolicModel}" href="#Base.convert-Tuple{Type{StandardModel}, MetabolicModel}"><code>Base.convert</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Base.convert(::Type{StandardModel}, model::MetabolicModel)</p><p>Convert any <code>MetabolicModel</code> into a <code>StandardModel</code>. Note, some data loss may occur since only the generic interface is used during the conversion process.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/StandardModel.jl#L293-L299">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.balance-Tuple{StandardModel}" href="#COBREXA.balance-Tuple{StandardModel}"><code>COBREXA.balance</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">balance(model::StandardModel)</code></pre><p>Return the balance of the linear problem, i.e. b in Sv = 0 where S is the stoichiometric matrix and v is the flux vector.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/StandardModel.jl#L172-L177">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.bounds-Tuple{StandardModel}" href="#COBREXA.bounds-Tuple{StandardModel}"><code>COBREXA.bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">bounds(model::StandardModel)</code></pre><p>Return the lower and upper bounds, respectively, for reactions in <code>model</code>. Order matches that of the reaction ids returned in <code>reactions()</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/StandardModel.jl#L163-L168">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.gene_annotations-Tuple{StandardModel, String}" href="#COBREXA.gene_annotations-Tuple{StandardModel, String}"><code>COBREXA.gene_annotations</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">gene_annotations(model::StandardModel, id::String)::Annotations</code></pre><p>Return the annotation associated with gene <code>id</code> in <code>model</code>. Return an empty Dict if not present.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/StandardModel.jl#L259-L264">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.gene_notes-Tuple{StandardModel, String}" href="#COBREXA.gene_notes-Tuple{StandardModel, String}"><code>COBREXA.gene_notes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">gene_notes(model::StandardModel, id::String)::Notes</code></pre><p>Return the notes associated with gene <code>id</code> in <code>model</code>. Return an empty Dict if not present.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/StandardModel.jl#L251-L256">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.genes-Tuple{StandardModel}" href="#COBREXA.genes-Tuple{StandardModel}"><code>COBREXA.genes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">genes(model::StandardModel)</code></pre><p>Return a vector of gene id strings in <code>model</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/StandardModel.jl#L91-L95">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.lower_bounds-Tuple{StandardModel}" href="#COBREXA.lower_bounds-Tuple{StandardModel}"><code>COBREXA.lower_bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">lower_bounds(model::StandardModel)</code></pre><p>Return the lower bounds for all reactions in <code>model</code> in sparse format.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/StandardModel.jl#L146-L150">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_annotations-Tuple{StandardModel, String}" href="#COBREXA.metabolite_annotations-Tuple{StandardModel, String}"><code>COBREXA.metabolite_annotations</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_annotations(model::StandardModel, id::String)::Annotations</code></pre><p>Return the annotation associated with metabolite <code>id</code> in <code>model</code>. Return an empty Dict if not present.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/StandardModel.jl#L242-L247">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_charge-Tuple{StandardModel, String}" href="#COBREXA.metabolite_charge-Tuple{StandardModel, String}"><code>COBREXA.metabolite_charge</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_charge(model::StandardModel, id::String)</code></pre><p>Return the charge associated with metabolite <code>id</code> in <code>model</code>. Return nothing if not present.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/StandardModel.jl#L206-L211">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_compartment-Tuple{StandardModel, String}" href="#COBREXA.metabolite_compartment-Tuple{StandardModel, String}"><code>COBREXA.metabolite_compartment</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_compartment(model::StandardModel, id::String)</code></pre><p>Return compartment associated with metabolite <code>id</code> in <code>model</code>. Return <code>nothing</code> if not present.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/StandardModel.jl#L215-L220">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_formula-Tuple{StandardModel, String}" href="#COBREXA.metabolite_formula-Tuple{StandardModel, String}"><code>COBREXA.metabolite_formula</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_formula(model::StandardModel, id::String)</code></pre><p>Return the formula of reaction <code>id</code> in <code>model</code>. Return <code>nothing</code> if not present.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/StandardModel.jl#L197-L202">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_notes-Tuple{StandardModel, String}" href="#COBREXA.metabolite_notes-Tuple{StandardModel, String}"><code>COBREXA.metabolite_notes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_notes(model::StandardModel, id::String)::Notes</code></pre><p>Return the notes associated with metabolite <code>id</code> in <code>model</code>. Return an empty Dict if not present.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/StandardModel.jl#L233-L238">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolites-Tuple{StandardModel}" href="#COBREXA.metabolites-Tuple{StandardModel}"><code>COBREXA.metabolites</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolites(model::StandardModel)</code></pre><p>Return a vector of metabolite id strings contained in <code>model</code>. The order of metabolite strings returned here matches the order used to construct the stoichiometric matrix.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/StandardModel.jl#L75-L81">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.n_genes-Tuple{StandardModel}" href="#COBREXA.n_genes-Tuple{StandardModel}"><code>COBREXA.n_genes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">n_genes(model::StandardModel)</code></pre><p>Return the number of genes in <code>model</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/StandardModel.jl#L98-L102">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.n_metabolites-Tuple{StandardModel}" href="#COBREXA.n_metabolites-Tuple{StandardModel}"><code>COBREXA.n_metabolites</code></a> — <span class="docstring-category">Method</span></header><section><div><p>n_metabolites(model::StandardModel)</p><p>Return the number of metabolites in <code>model</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/StandardModel.jl#L84-L88">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.n_reactions-Tuple{StandardModel}" href="#COBREXA.n_reactions-Tuple{StandardModel}"><code>COBREXA.n_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">n_reactions(model::StandardModel)</code></pre><p>Return the number of reactions contained in <code>model</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/StandardModel.jl#L67-L71">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.objective-Tuple{StandardModel}" href="#COBREXA.objective-Tuple{StandardModel}"><code>COBREXA.objective</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">objective(model::StandardModel)</code></pre><p>Return sparse objective vector for <code>model</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/StandardModel.jl#L180-L184">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_annotations-Tuple{StandardModel, String}" href="#COBREXA.reaction_annotations-Tuple{StandardModel, String}"><code>COBREXA.reaction_annotations</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_annotations(model::StandardModel, id::String)::Annotations</code></pre><p>Return the annotation associated with reaction <code>id</code> in <code>model</code>. Return an empty Dict if not present.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/StandardModel.jl#L276-L281">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_gene_association-Tuple{StandardModel, String}" href="#COBREXA.reaction_gene_association-Tuple{StandardModel, String}"><code>COBREXA.reaction_gene_association</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_gene_association(model::StandardModel, id::String)</code></pre><p>Return the gene reaction rule in string format for reaction with <code>id</code> in <code>model</code>. Return <code>nothing</code> if not available.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/StandardModel.jl#L188-L193">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_notes-Tuple{StandardModel, String}" href="#COBREXA.reaction_notes-Tuple{StandardModel, String}"><code>COBREXA.reaction_notes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_notes(model::StandardModel, id::String)::Notes</code></pre><p>Return the notes associated with reaction <code>id</code> in <code>model</code>. Return an empty Dict if not present.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/StandardModel.jl#L268-L273">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_stoichiometry-Tuple{StandardModel, String}" href="#COBREXA.reaction_stoichiometry-Tuple{StandardModel, String}"><code>COBREXA.reaction_stoichiometry</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_stoichiometry(model::StandardModel, rid::String)::Dict{String, Float64}</code></pre><p>Return the stoichiometry of reaction with ID <code>rid</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/StandardModel.jl#L285-L289">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_subsystem-Tuple{StandardModel, String}" href="#COBREXA.reaction_subsystem-Tuple{StandardModel, String}"><code>COBREXA.reaction_subsystem</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_subsystem(id::String, model::StandardModel)</code></pre><p>Return the subsystem associated with reaction <code>id</code> in <code>model</code>. Return <code>nothing</code> if not present.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/StandardModel.jl#L224-L229">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reactions-Tuple{StandardModel}" href="#COBREXA.reactions-Tuple{StandardModel}"><code>COBREXA.reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reactions(model::StandardModel)</code></pre><p>Return a vector of reaction id strings contained in <code>model</code>. The order of reaction ids returned here matches the order used to construct the stoichiometric matrix.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/StandardModel.jl#L58-L64">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.stoichiometry-Tuple{StandardModel}" href="#COBREXA.stoichiometry-Tuple{StandardModel}"><code>COBREXA.stoichiometry</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">stoichiometry(model::StandardModel)</code></pre><p>Return the stoichiometric matrix associated with <code>model</code> in sparse format.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/StandardModel.jl#L105-L109">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.upper_bounds-Tuple{StandardModel}" href="#COBREXA.upper_bounds-Tuple{StandardModel}"><code>COBREXA.upper_bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">upper_bounds(model::StandardModel)</code></pre><p>Return the upper bounds for all reactions in <code>model</code> in sparse format. Order matches that of the reaction ids returned in <code>reactions()</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/types/StandardModel.jl#L154-L159">source</a></section></article><h2 id="Base-functions"><a class="docs-heading-anchor" href="#Base-functions">Base functions</a><a id="Base-functions-1"></a><a class="docs-heading-anchor-permalink" href="#Base-functions" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="COBREXA._constants" href="#COBREXA._constants"><code>COBREXA._constants</code></a> — <span class="docstring-category">Constant</span></header><section><div><p>A named tuple that contains the magic values that are used globally for whatever purposes.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/constants.jl#L2-L5">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.get_optmodel_bounds-Tuple{Any}" href="#COBREXA.get_optmodel_bounds-Tuple{Any}"><code>COBREXA.get_optmodel_bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">get_optmodel_bounds(opt_model)</code></pre><p>Returns vectors of the lower and upper bounds of <code>opt_model</code> constraints, where <code>opt_model</code> is a JuMP model constructed by e.g. <a href="#COBREXA.make_optimization_model-Tuple{MetabolicModel, Any}"><code>make_optimization_model</code></a> or <a href="#COBREXA.flux_balance_analysis-Union{Tuple{M}, Tuple{M, Any}} where M&lt;:MetabolicModel"><code>flux_balance_analysis</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/solver.jl#L60-L66">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.is_solved-Tuple{Any}" href="#COBREXA.is_solved-Tuple{Any}"><code>COBREXA.is_solved</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">is_solved(optmodel)</code></pre><p>Return <code>true</code> if <code>optmodel</code> solved successfully (solution is optimal or locally optimal).  Return <code>false</code> if any other termination status is reached. Termination status is defined in the documentation of <code>JuMP</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/solver.jl#L34-L40">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.make_optimization_model-Tuple{MetabolicModel, Any}" href="#COBREXA.make_optimization_model-Tuple{MetabolicModel, Any}"><code>COBREXA.make_optimization_model</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">make_optimization_model(
    model::MetabolicModel,
    optimizer;
    sense = MOI.MAX_SENSE,
)</code></pre><p>Convert <code>MetabolicModel</code>s to a JuMP model, place objectives and the equality constraint.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/solver.jl#L2-L11">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.optimize_objective-Tuple{Any}" href="#COBREXA.optimize_objective-Tuple{Any}"><code>COBREXA.optimize_objective</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">optimize_objective(optmodel)::Union{Float64,Nothing}</code></pre><p>Shortcut for running JuMP <code>optimize!</code> on a model and returning the objective value, if solved.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/solver.jl#L45-L50">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.set_optmodel_bound!-Tuple{Any, Any}" href="#COBREXA.set_optmodel_bound!-Tuple{Any, Any}"><code>COBREXA.set_optmodel_bound!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">set_optmodel_bound!(vidx, opt_model;
    ub::Maybe{Real} = nothing,
    lb::Maybe{Real} = nothing,
)</code></pre><p>Helper function to set the bounds of a variable in the model. Internally calls <code>set_normalized_rhs</code> from JuMP. If the bounds are set to <code>nothing</code>, they will not be changed.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/solver.jl#L72-L81">source</a></section></article><h2 id="File-I/O-and-serialization"><a class="docs-heading-anchor" href="#File-I/O-and-serialization">File I/O and serialization</a><a id="File-I/O-and-serialization-1"></a><a class="docs-heading-anchor-permalink" href="#File-I/O-and-serialization" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="COBREXA.load_model-Tuple{String}" href="#COBREXA.load_model-Tuple{String}"><code>COBREXA.load_model</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">load_model(file_name::String)::MetabolicModel</code></pre><p>Generic function for loading models that chooses a specific loader function from the <code>file_name</code> extension, or throws an error.</p><p>Currently, these model types are supported:</p><ul><li>SBML models (<code>*.xml</code>, loaded with <a href="#COBREXA.load_sbml_model-Tuple{String}"><code>load_sbml_model</code></a>)</li><li>JSON models (<code>*.json</code>, loaded with <a href="#COBREXA.load_json_model-Tuple{String}"><code>load_json_model</code></a>)</li><li>MATLAB models (<code>*.mat</code>, loaded with <a href="#COBREXA.load_mat_model-Tuple{String}"><code>load_mat_model</code></a>)</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/io/io.jl#L2-L13">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.load_model-Union{Tuple{T}, Tuple{Type{T}, String}} where T&lt;:MetabolicModel" href="#COBREXA.load_model-Union{Tuple{T}, Tuple{Type{T}, String}} where T&lt;:MetabolicModel"><code>COBREXA.load_model</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">load_model(type::Type{T}, file_name::String)::T where T</code></pre><p>Helper function tht loads the model using <a href="#COBREXA.load_model-Tuple{String}"><code>load_model</code></a> and return it converted to <code>type</code>.</p><p><strong>Example:</strong></p><pre><code class="language-none">load_model(CoreModel, &quot;mySBMLModel.xml&quot;)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/io/io.jl#L27-L36">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.save_model-Tuple{MetabolicModel, String}" href="#COBREXA.save_model-Tuple{MetabolicModel, String}"><code>COBREXA.save_model</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">save_model(model::MetabolicModel, file_name::String)</code></pre><p>Generic function for saving models that chooses a specific writer function from the <code>file_name</code> extension, or throws an error.</p><p>Currently, these model types are supported:</p><ul><li>JSON models (<code>*.json</code>, loaded with <a href="#COBREXA.save_json_model-Tuple{MetabolicModel, String}"><code>save_json_model</code></a>)</li><li>MATLAB models (<code>*.mat</code>, loaded with <a href="#COBREXA.save_mat_model-Tuple{MetabolicModel, String}"><code>save_mat_model</code></a>)</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/io/io.jl#L41-L51">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.load_json_model-Tuple{String}" href="#COBREXA.load_json_model-Tuple{String}"><code>COBREXA.load_json_model</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">load_json_model(filename::String)::JSONModel</code></pre><p>Load and return a JSON-formatted model that is stored in <code>file_name</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/io/json.jl#L1-L5">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.save_json_model-Tuple{MetabolicModel, String}" href="#COBREXA.save_json_model-Tuple{MetabolicModel, String}"><code>COBREXA.save_json_model</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">save_json_model(model::MetabolicModel, file_name::String)</code></pre><p>Save a <a href="#COBREXA.JSONModel"><code>JSONModel</code></a> in <code>model</code> to a JSON file <code>file_name</code>.</p><p>In case the <code>model</code> is not <code>JSONModel</code>, it will be converted automatically.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/io/json.jl#L10-L16">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.load_mat_model-Tuple{String}" href="#COBREXA.load_mat_model-Tuple{String}"><code>COBREXA.load_mat_model</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">load_mat_model(file_name::String)</code></pre><p>Load and return a MATLAB file <code>file_name</code> that contains a COBRA-compatible model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/io/mat.jl#L2-L7">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.save_mat_model-Tuple{MetabolicModel, String}" href="#COBREXA.save_mat_model-Tuple{MetabolicModel, String}"><code>COBREXA.save_mat_model</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">save_mat_model(model::MetabolicModel, file_name::String; model_name::String=&quot;model&quot;)</code></pre><p>Save a <a href="#COBREXA.MATModel"><code>MATModel</code></a> in <code>model</code> to a MATLAB file <code>file_name</code> in a format compatible with other MATLAB-based COBRA software.</p><p>In case the <code>model</code> is not <code>MATModel</code>, it will be converted automatically.</p><p><code>model_name</code> is the identifier name for the whole model written to the MATLAB file; defaults to just &quot;model&quot;.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/io/mat.jl#L14-L24">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.load_sbml_model-Tuple{String}" href="#COBREXA.load_sbml_model-Tuple{String}"><code>COBREXA.load_sbml_model</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">load_sbml_model(file_name::String)::SBMLModel</code></pre><p>Load and return a SBML XML model in <code>file_name</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/io/sbml.jl#L2-L6">source</a></section></article><h3 id="Pretty-printing"><a class="docs-heading-anchor" href="#Pretty-printing">Pretty printing</a><a id="Pretty-printing-1"></a><a class="docs-heading-anchor-permalink" href="#Pretty-printing" title="Permalink"></a></h3><article class="docstring"><header><a class="docstring-binding" id="Base.show-Tuple{IO, MIME{Symbol(&quot;text/plain&quot;)}, MetabolicModel}" href="#Base.show-Tuple{IO, MIME{Symbol(&quot;text/plain&quot;)}, MetabolicModel}"><code>Base.show</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Pretty printing of everything metabolic-modelish.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/io/show/MetabolicModel.jl#L1-L3">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA._pretty_substances-Tuple{Vector{String}}" href="#COBREXA._pretty_substances-Tuple{Vector{String}}"><code>COBREXA._pretty_substances</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">_pretty_substances(ss::Vector{String})::String</code></pre><p>Nicely format a substance list.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/io/show/Reaction.jl#L2-L6">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Base.show-Union{Tuple{M}, Tuple{IO, MIME{Symbol(&quot;text/plain&quot;)}, Serialized{M}}} where M" href="#Base.show-Union{Tuple{M}, Tuple{IO, MIME{Symbol(&quot;text/plain&quot;)}, Serialized{M}}} where M"><code>Base.show</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">Base.show(io::IO, ::MIME&quot;text/plain&quot;, m::Serialized{M}) where {M}</code></pre><p>Show the <a href="#COBREXA.Serialized"><code>Serialized</code></a> model without unnecessarily loading it.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/io/show/Serialized.jl#L2-L6">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA._pretty_print_keyvals-Tuple{Any, String, Any}" href="#COBREXA._pretty_print_keyvals-Tuple{Any, String, Any}"><code>COBREXA._pretty_print_keyvals</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">_pretty_print_keyvals(io, def::String, payload; kwargs...)</code></pre><p>Nicely prints keys and values.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/io/show/pretty_printing.jl#L1-L5">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA._pretty_print_keyvals-Tuple{Any, String, Dict}" href="#COBREXA._pretty_print_keyvals-Tuple{Any, String, Dict}"><code>COBREXA._pretty_print_keyvals</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">_pretty_print_keyvals(
    io,
    def::String,
    payload::Dict
)</code></pre><p>Specialization of <code>_pretty_print_keyvals</code> for dictionaries.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/io/show/pretty_printing.jl#L27-L35">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA._pretty_print_keyvals-Tuple{Any, String, String}" href="#COBREXA._pretty_print_keyvals-Tuple{Any, String, String}"><code>COBREXA._pretty_print_keyvals</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">_pretty_print_keyvals(
    io,
    def::String,
    payload::String
)</code></pre><p>Specialization of <code>_pretty_print_keyvals</code> for plain strings.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/io/show/pretty_printing.jl#L9-L17">source</a></section></article><h2 id="Model-reconstruction"><a class="docs-heading-anchor" href="#Model-reconstruction">Model reconstruction</a><a id="Model-reconstruction-1"></a><a class="docs-heading-anchor-permalink" href="#Model-reconstruction" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_reaction!-Tuple{CoreModel, Reaction}" href="#COBREXA.add_reaction!-Tuple{CoreModel, Reaction}"><code>COBREXA.add_reaction!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_reaction!(model::CoreModel, rxn::Reaction)</code></pre><p>Add <code>rxn</code> to <code>model</code>. The model must already contain the metabolites used by <code>rxn</code> in the model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/CoreModel.jl#L34-L39">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_reactions!-Tuple{CoreModel, Vector{Reaction}}" href="#COBREXA.add_reactions!-Tuple{CoreModel, Vector{Reaction}}"><code>COBREXA.add_reactions!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_reactions!(model::CoreModel, rxns::Vector{Reaction})</code></pre><p>Add <code>rxns</code> to <code>model</code> efficiently. The model must already contain the metabolites used by <code>rxns</code> in the model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/CoreModel.jl#L1-L6">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_reactions-Tuple{CoreModel, CoreModel}" href="#COBREXA.add_reactions-Tuple{CoreModel, CoreModel}"><code>COBREXA.add_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_reactions(m1::CoreModel, m2::CoreModel; check_consistency = false)</code></pre><p>Add all reactions from <code>m2</code> to <code>m1</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/CoreModel.jl#L147-L151">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_reactions-Union{Tuple{K}, Tuple{V2}, Tuple{V1}, Tuple{CoreModel, V1, V2, AbstractFloat, AbstractFloat, AbstractFloat, String, K}} where {V1&lt;:AbstractVector{Float64}, V2&lt;:AbstractVector{Float64}, K&lt;:AbstractVector{String}}" href="#COBREXA.add_reactions-Union{Tuple{K}, Tuple{V2}, Tuple{V1}, Tuple{CoreModel, V1, V2, AbstractFloat, AbstractFloat, AbstractFloat, String, K}} where {V1&lt;:AbstractVector{Float64}, V2&lt;:AbstractVector{Float64}, K&lt;:AbstractVector{String}}"><code>COBREXA.add_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_reactions(
    m::CoreModel,
    s::V1,
    b::V2,
    c::AbstractFloat,
    xl::AbstractFloat,
    xu::AbstractFloat,
    rxn::String,
    mets::K;
    check_consistency = false,
) where {V1&lt;:VecType,V2&lt;:VecType,K&lt;:StringVecType}</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/CoreModel.jl#L75-L87">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_reactions-Union{Tuple{K}, Tuple{V}, Tuple{M}, Tuple{CoreModel, M, V, V, V, V, K, K}} where {M&lt;:AbstractMatrix{Float64}, V&lt;:AbstractVector{Float64}, K&lt;:AbstractVector{String}}" href="#COBREXA.add_reactions-Union{Tuple{K}, Tuple{V}, Tuple{M}, Tuple{CoreModel, M, V, V, V, V, K, K}} where {M&lt;:AbstractMatrix{Float64}, V&lt;:AbstractVector{Float64}, K&lt;:AbstractVector{String}}"><code>COBREXA.add_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_reactions(
    m::CoreModel,
    Sp::M,
    b::V,
    c::V,
    xl::V,
    xu::V,
    rxns::K,
    mets::K;
    check_consistency = false,
) where {M&lt;:MatType,V&lt;:VecType,K&lt;:StringVecType}</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/CoreModel.jl#L166-L178">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_reactions-Union{Tuple{V2}, Tuple{V1}, Tuple{CoreModel, V1, V2, AbstractFloat, AbstractFloat, AbstractFloat}} where {V1&lt;:AbstractVector{Float64}, V2&lt;:AbstractVector{Float64}}" href="#COBREXA.add_reactions-Union{Tuple{V2}, Tuple{V1}, Tuple{CoreModel, V1, V2, AbstractFloat, AbstractFloat, AbstractFloat}} where {V1&lt;:AbstractVector{Float64}, V2&lt;:AbstractVector{Float64}}"><code>COBREXA.add_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_reactions(
    m::CoreModel,
    s::V1,
    b::V2,
    c::AbstractFloat,
    xl::AbstractFloat,
    xu::AbstractFloat;
    check_consistency = false,
) where {V1&lt;:VecType,V2&lt;:VecType}</code></pre><p>Add reaction(s) to a <code>CoreModel</code> model <code>m</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/CoreModel.jl#L42-L54">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_reactions-Union{Tuple{V}, Tuple{M}, Tuple{CoreModel, M, V, V, V, V}} where {M&lt;:AbstractMatrix{Float64}, V&lt;:AbstractVector{Float64}}" href="#COBREXA.add_reactions-Union{Tuple{V}, Tuple{M}, Tuple{CoreModel, M, V, V, V, V}} where {M&lt;:AbstractMatrix{Float64}, V&lt;:AbstractVector{Float64}}"><code>COBREXA.add_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_reactions(
    m::CoreModel,
    Sp::M,
    b::V,
    c::V,
    xl::V,
    xu::V;
    check_consistency = false,
) where {M&lt;:MatType,V&lt;:VecType}</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/CoreModel.jl#L112-L122">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.verify_consistency-Union{Tuple{K}, Tuple{V}, Tuple{M}, Tuple{CoreModel, M, V, V, V, V, K, K, Any, Any}} where {M&lt;:AbstractMatrix{Float64}, V&lt;:AbstractVector{Float64}, K&lt;:AbstractVector{String}}" href="#COBREXA.verify_consistency-Union{Tuple{K}, Tuple{V}, Tuple{M}, Tuple{CoreModel, M, V, V, V, V, K, K, Any, Any}} where {M&lt;:AbstractMatrix{Float64}, V&lt;:AbstractVector{Float64}, K&lt;:AbstractVector{String}}"><code>COBREXA.verify_consistency</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">verify_consistency(
    m::CoreModel,
    Sp::M,
    b::V,
    c::V,
    xl::V,
    xu::V,
    names::K,
    mets::K,
    new_reactions,
    new_metabolites,
) where {M&lt;:MatType,V&lt;:VecType,K&lt;:StringVecType}</code></pre><p>Check the consistency of given reactions with existing reactions in <code>m</code>.</p><p>TODO: work in progress, doesn&#39;t return consistency status.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/CoreModel.jl#L248-L265">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_coupling_constraints!-Tuple{CoreModelCoupled, AbstractVector{Float64}, AbstractFloat, AbstractFloat}" href="#COBREXA.add_coupling_constraints!-Tuple{CoreModelCoupled, AbstractVector{Float64}, AbstractFloat, AbstractFloat}"><code>COBREXA.add_coupling_constraints!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_coupling_constraints!(
    m::CoreModelCoupled,
    c::VecType,
    cl::AbstractFloat,
    cu::AbstractFloat,
)</code></pre><p>Overload for adding a single coupling constraint.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/CoreModelCoupled.jl#L189-L198">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_coupling_constraints!-Union{Tuple{V}, Tuple{CoreModelCoupled, AbstractMatrix{Float64}, V, V}} where V&lt;:AbstractVector{Float64}" href="#COBREXA.add_coupling_constraints!-Union{Tuple{V}, Tuple{CoreModelCoupled, AbstractMatrix{Float64}, V, V}} where V&lt;:AbstractVector{Float64}"><code>COBREXA.add_coupling_constraints!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_coupling_constraints!(
    m::CoreModelCoupled,
    C::MatType,
    cl::V,
    cu::V,
) where {V&lt;:VecType}</code></pre><p>In-place add a single coupling constraint in form</p><pre><code class="language-none">    cₗ ≤ C x ≤ cᵤ</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/CoreModelCoupled.jl#L214-L226">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_coupling_constraints-Tuple{CoreModel, Vararg{Any, N} where N}" href="#COBREXA.add_coupling_constraints-Tuple{CoreModel, Vararg{Any, N} where N}"><code>COBREXA.add_coupling_constraints</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_coupling_constraints(m::CoreModel, args...)</code></pre><p>Add coupling constraints to a plain <a href="#COBREXA.CoreModel"><code>CoreModel</code></a> (returns a <a href="#COBREXA.CoreModelCoupled"><code>CoreModelCoupled</code></a>).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/CoreModelCoupled.jl#L180-L185">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_coupling_constraints-Tuple{CoreModelCoupled, Vararg{Any, N} where N}" href="#COBREXA.add_coupling_constraints-Tuple{CoreModelCoupled, Vararg{Any, N} where N}"><code>COBREXA.add_coupling_constraints</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Add constraints of the following form to a CoreModelCoupled and return a modified one.</p><p>Add constraints to a <a href="#COBREXA.CoreModelCoupled"><code>CoreModelCoupled</code></a> and return a modified one.</p><p>The arguments are same as for in-place <a href="#COBREXA.add_coupling_constraints!-Tuple{CoreModelCoupled, AbstractVector{Float64}, AbstractFloat, AbstractFloat}"><code>add_coupling_constraints!</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/CoreModelCoupled.jl#L167-L173">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_reactions-Tuple{CoreModelCoupled, CoreModel}" href="#COBREXA.add_reactions-Tuple{CoreModelCoupled, CoreModel}"><code>COBREXA.add_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_reactions(m1::CoreModelCoupled, m2::CoreModel; check_consistency = false)</code></pre><p>Add all reactions from <code>m2</code> to <code>m1</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/CoreModelCoupled.jl#L107-L112">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_reactions-Union{Tuple{K}, Tuple{V2}, Tuple{V1}, Tuple{CoreModelCoupled, V1, V2, AbstractFloat, AbstractFloat, AbstractFloat, String, K}} where {V1&lt;:AbstractVector{Float64}, V2&lt;:AbstractVector{Float64}, K&lt;:AbstractVector{String}}" href="#COBREXA.add_reactions-Union{Tuple{K}, Tuple{V2}, Tuple{V1}, Tuple{CoreModelCoupled, V1, V2, AbstractFloat, AbstractFloat, AbstractFloat, String, K}} where {V1&lt;:AbstractVector{Float64}, V2&lt;:AbstractVector{Float64}, K&lt;:AbstractVector{String}}"><code>COBREXA.add_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_reactions(
    m::CoreModelCoupled,
    s::V1,
    b::V2,
    c::AbstractFloat,
    xl::AbstractFloat,
    xu::AbstractFloat,
    rxn::String,
    mets::K;
    check_consistency = false,
) where {V1&lt;:VecType,V2&lt;:VecType,K&lt;:StringVecType}</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/CoreModelCoupled.jl#L33-L46">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_reactions-Union{Tuple{K}, Tuple{V}, Tuple{M}, Tuple{CoreModelCoupled, M, V, V, V, V, K, K}} where {M&lt;:AbstractMatrix{Float64}, V&lt;:AbstractVector{Float64}, K&lt;:AbstractVector{String}}" href="#COBREXA.add_reactions-Union{Tuple{K}, Tuple{V}, Tuple{M}, Tuple{CoreModelCoupled, M, V, V, V, V, K, K}} where {M&lt;:AbstractMatrix{Float64}, V&lt;:AbstractVector{Float64}, K&lt;:AbstractVector{String}}"><code>COBREXA.add_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_reactions(
    m::CoreModelCoupled,
    Sp::M,
    b::V,
    c::V,
    xl::V,
    xu::V,
    rxns::K,
    mets::K;
    check_consistency = false,
) where {M&lt;:MatType,V&lt;:VecType,K&lt;:StringVecType}</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/CoreModelCoupled.jl#L123-L136">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_reactions-Union{Tuple{V2}, Tuple{V1}, Tuple{CoreModelCoupled, V1, V2, AbstractFloat, AbstractFloat, AbstractFloat}} where {V1&lt;:AbstractVector{Float64}, V2&lt;:AbstractVector{Float64}}" href="#COBREXA.add_reactions-Union{Tuple{V2}, Tuple{V1}, Tuple{CoreModelCoupled, V1, V2, AbstractFloat, AbstractFloat, AbstractFloat}} where {V1&lt;:AbstractVector{Float64}, V2&lt;:AbstractVector{Float64}}"><code>COBREXA.add_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_reactions(
    m::CoreModelCoupled,
    s::V1,
    b::V2,
    c::AbstractFloat,
    xl::AbstractFloat,
    xu::AbstractFloat;
    check_consistency = false,
) where {V1&lt;:VecType,V2&lt;:VecType}</code></pre><p>Add reaction(s) to a <code>CoreModelCoupled</code> model <code>m</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/CoreModelCoupled.jl#L1-L14">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_reactions-Union{Tuple{V}, Tuple{M}, Tuple{CoreModelCoupled, M, V, V, V, V}} where {M&lt;:AbstractMatrix{Float64}, V&lt;:AbstractVector{Float64}}" href="#COBREXA.add_reactions-Union{Tuple{V}, Tuple{M}, Tuple{CoreModelCoupled, M, V, V, V, V}} where {M&lt;:AbstractMatrix{Float64}, V&lt;:AbstractVector{Float64}}"><code>COBREXA.add_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_reactions(
    m::CoreModelCoupled,
    Sp::M,
    b::V,
    c::V,
    xl::V,
    xu::V;
    check_consistency = false,
) where {M&lt;:MatType,V&lt;:VecType}</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/CoreModelCoupled.jl#L77-L88">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_coupling_bounds!-Union{Tuple{V}, Tuple{CoreModelCoupled, Vector{Int64}}} where V&lt;:AbstractVector{Float64}" href="#COBREXA.change_coupling_bounds!-Union{Tuple{V}, Tuple{CoreModelCoupled, Vector{Int64}}} where V&lt;:AbstractVector{Float64}"><code>COBREXA.change_coupling_bounds!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_coupling_bounds!(
    model::CoreModelCoupled,
    constraints::Vector{Int};
    cl::V = Float64[],
    cu::V = Float64[],
) where {V&lt;:VecType}</code></pre><p>Change the lower and/or upper bounds (<code>cl</code> and <code>cu</code>) for the given list of coupling constraints.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/CoreModelCoupled.jl#L283-L293">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_coupling_constraints!-Tuple{CoreModelCoupled, Int64}" href="#COBREXA.remove_coupling_constraints!-Tuple{CoreModelCoupled, Int64}"><code>COBREXA.remove_coupling_constraints!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_coupling_constraints!(m::CoreModelCoupled, constraint::Int)</code></pre><p>Removes a single coupling constraints from a <a href="#COBREXA.CoreModelCoupled"><code>CoreModelCoupled</code></a> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/CoreModelCoupled.jl#L259-L264">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_coupling_constraints!-Tuple{CoreModelCoupled, Vector{Int64}}" href="#COBREXA.remove_coupling_constraints!-Tuple{CoreModelCoupled, Vector{Int64}}"><code>COBREXA.remove_coupling_constraints!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_coupling_constraints!(m::CoreModelCoupled, constraints::Vector{Int})</code></pre><p>Removes a set of coupling constraints from a <a href="#COBREXA.CoreModelCoupled"><code>CoreModelCoupled</code></a> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/CoreModelCoupled.jl#L269-L274">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_coupling_constraints-Tuple{CoreModelCoupled, Vararg{Any, N} where N}" href="#COBREXA.remove_coupling_constraints-Tuple{CoreModelCoupled, Vararg{Any, N} where N}"><code>COBREXA.remove_coupling_constraints</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_coupling_constraints(m::CoreModelCoupled, args...)</code></pre><p>Remove coupling constraints from the linear model, and return the modified model. Arguments are the same as for in-place version <a href="#COBREXA.remove_coupling_constraints!-Tuple{CoreModelCoupled, Int64}"><code>remove_coupling_constraints!</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/CoreModelCoupled.jl#L245-L251">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.:←-Tuple{Union{Nothing, Metabolite, MetaboliteWithCoefficient, Vector{MetaboliteWithCoefficient}}, Union{Nothing, Metabolite, MetaboliteWithCoefficient, Vector{MetaboliteWithCoefficient}}}" href="#COBREXA.:←-Tuple{Union{Nothing, Metabolite, MetaboliteWithCoefficient, Vector{MetaboliteWithCoefficient}}, Union{Nothing, Metabolite, MetaboliteWithCoefficient, Vector{MetaboliteWithCoefficient}}}"><code>COBREXA.:←</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">←(
    substrates::Union{
        Nothing,
        Metabolite,
        MetaboliteWithCoefficient,
        Vector{MetaboliteWithCoefficient},
    },
    products::Union{
        Nothing,
        Metabolite,
        MetaboliteWithCoefficient,
        Vector{MetaboliteWithCoefficient}
    },
)</code></pre><p>Make a reverse-only <a href="#COBREXA.Reaction"><code>Reaction</code></a> from <code>substrates</code> and <code>products</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/Reaction.jl#L100-L117">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.:→-Tuple{Union{Nothing, Metabolite, MetaboliteWithCoefficient, Vector{MetaboliteWithCoefficient}}, Union{Nothing, Metabolite, MetaboliteWithCoefficient, Vector{MetaboliteWithCoefficient}}}" href="#COBREXA.:→-Tuple{Union{Nothing, Metabolite, MetaboliteWithCoefficient, Vector{MetaboliteWithCoefficient}}, Union{Nothing, Metabolite, MetaboliteWithCoefficient, Vector{MetaboliteWithCoefficient}}}"><code>COBREXA.:→</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">→(
    substrates::Union{
        Nothing,
        Metabolite,
        MetaboliteWithCoefficient,
        Vector{MetaboliteWithCoefficient},
    },
    products::Union{
        Nothing,
        Metabolite,
        MetaboliteWithCoefficient,
        Vector{MetaboliteWithCoefficient}
    },
)</code></pre><p>Make a forward-only <a href="#COBREXA.Reaction"><code>Reaction</code></a> from <code>substrates</code> and <code>products</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/Reaction.jl#L64-L81">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.:↔-Tuple{Union{Nothing, Metabolite, MetaboliteWithCoefficient, Vector{MetaboliteWithCoefficient}}, Union{Nothing, Metabolite, MetaboliteWithCoefficient, Vector{MetaboliteWithCoefficient}}}" href="#COBREXA.:↔-Tuple{Union{Nothing, Metabolite, MetaboliteWithCoefficient, Vector{MetaboliteWithCoefficient}}, Union{Nothing, Metabolite, MetaboliteWithCoefficient, Vector{MetaboliteWithCoefficient}}}"><code>COBREXA.:↔</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">↔(
    substrates::Union{
        Nothing,
        Metabolite,
        MetaboliteWithCoefficient,
        Vector{MetaboliteWithCoefficient},
    },
    products::Union{
        Nothing,
        Metabolite,
        MetaboliteWithCoefficient,
        Vector{MetaboliteWithCoefficient}
    },
)</code></pre><p>Make a bidirectional (reversible) <a href="#COBREXA.Reaction"><code>Reaction</code></a> from <code>substrates</code> and <code>products</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/Reaction.jl#L136-L154">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_gene!-Tuple{StandardModel, Gene}" href="#COBREXA.add_gene!-Tuple{StandardModel, Gene}"><code>COBREXA.add_gene!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_gene!(model::StandardModel, genes::Gene)</code></pre><p>Add <code>gene</code> to <code>model</code> based on gene <code>id</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/StandardModel.jl#L48-L52">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_genes!-Tuple{StandardModel, Vector{Gene}}" href="#COBREXA.add_genes!-Tuple{StandardModel, Vector{Gene}}"><code>COBREXA.add_genes!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_genes!(model::StandardModel, genes::Vector{Gene})</code></pre><p>Add <code>genes</code> to <code>model</code> based on gene <code>id</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/StandardModel.jl#L37-L41">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_metabolite!-Tuple{StandardModel, Metabolite}" href="#COBREXA.add_metabolite!-Tuple{StandardModel, Metabolite}"><code>COBREXA.add_metabolite!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_metabolite!(model::StandardModel, met::Metabolite)</code></pre><p>Add <code>met</code> to <code>model</code> based on metabolite <code>id</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/StandardModel.jl#L30-L34">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_metabolites!-Tuple{StandardModel, Vector{Metabolite}}" href="#COBREXA.add_metabolites!-Tuple{StandardModel, Vector{Metabolite}}"><code>COBREXA.add_metabolites!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_metabolites!(model::StandardModel, mets::Vector{Metabolite})</code></pre><p>Add <code>mets</code> to <code>model</code> based on metabolite <code>id</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/StandardModel.jl#L19-L23">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_reaction!-Tuple{StandardModel, Reaction}" href="#COBREXA.add_reaction!-Tuple{StandardModel, Reaction}"><code>COBREXA.add_reaction!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_reaction!(model::StandardModel, rxn::Reaction)</code></pre><p>Add <code>rxn</code> to <code>model</code> based on reaction <code>id</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/StandardModel.jl#L12-L16">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_reactions!-Tuple{StandardModel, Vector{Reaction}}" href="#COBREXA.add_reactions!-Tuple{StandardModel, Vector{Reaction}}"><code>COBREXA.add_reactions!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_reactions!(model::StandardModel, rxns::Vector{Reaction})</code></pre><p>Add <code>rxns</code> to <code>model</code> based on reaction <code>id</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/StandardModel.jl#L1-L5">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_gene!-Tuple{StandardModel, String}" href="#COBREXA.remove_gene!-Tuple{StandardModel, String}"><code>COBREXA.remove_gene!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_gene!(
    model::StandardModel,
    id::Vector{String};
    knockout_reactions::Bool = false,
)</code></pre><p>Remove gene with <code>id</code> from <code>model</code>. If <code>knockout_reactions</code> is true, then also constrain reactions that require the genes to function to carry zero flux.</p><p><strong>Example</strong></p><pre><code class="language-none">remove_gene!(model, &quot;g1&quot;)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/StandardModel.jl#L142-L156">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_genes!-Tuple{StandardModel, Vector{String}}" href="#COBREXA.remove_genes!-Tuple{StandardModel, Vector{String}}"><code>COBREXA.remove_genes!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_genes!(
    model::StandardModel,
    ids::Vector{String};
    knockout_reactions::Bool = false,
)</code></pre><p>Remove all genes with <code>ids</code> from <code>model</code>. If <code>knockout_reactions</code> is true, then also constrain reactions that require the genes to function to carry zero flux.</p><p><strong>Example</strong></p><pre><code class="language-none">remove_genes!(model, [&quot;g1&quot;, &quot;g2&quot;])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/StandardModel.jl#L109-L123">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.@add_reactions!-Tuple{Symbol, Expr}" href="#COBREXA.@add_reactions!-Tuple{Symbol, Expr}"><code>COBREXA.@add_reactions!</code></a> — <span class="docstring-category">Macro</span></header><section><div><pre><code class="language-julia">@add_reactions!(model::Symbol, ex::Expr)</code></pre><p>Shortcut to add multiple reactions and their lower and upper bounds</p><p><strong>Call variants</strong></p><pre><code class="language-none">@add_reactions! model begin
    reaction_name, reaction
end

@add_reactions! model begin
    reaction_name, reaction, lower_bound
end

@add_reactions! model begin
    reaction_name, reaction, lower_bound, upper_bound
end</code></pre><p><strong>Examples</strong></p><pre><code class="language-none">@add_reactions! model begin
    &quot;v1&quot;, nothing → A, 0, 500
    &quot;v2&quot;, A ↔ B + C, -500
    &quot;v3&quot;, B + C → nothing
end</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/StandardModel.jl#L55-L85">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_model_with_exchanges-Tuple{CoreModel, MetabolicModel, Dict{String, String}}" href="#COBREXA.add_model_with_exchanges-Tuple{CoreModel, MetabolicModel, Dict{String, String}}"><code>COBREXA.add_model_with_exchanges</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_model_with_exchanges(
    community::CoreModel,
    model::MetabolicModel,
    exchange_rxn_mets::Dict{String,String};
    model_name = &quot;unknown_species&quot;,
    biomass_id = nothing,
)::CoreModel</code></pre><p>Add <code>model</code> to <code>community</code>, which is a pre-existing community model with exchange reactions and metabolites in the dictionary <code>exchange_rxn_mets</code>. The <code>model_name</code> is appended to each reaction and metabolite, see <a href="#COBREXA.join_with_exchanges-Union{Tuple{M}, Tuple{Vector{M}, Dict{String, String}}} where M&lt;:MetabolicModel"><code>join_with_exchanges</code></a>. If <code>biomass_id</code> is specified then a biomass metabolite for <code>model</code> is also added to the resulting model. The column corresponding to the <code>biomass_id</code> reaction then produces this new biomass metabolite with unit coefficient. Note, <code>exchange_rxn_ids</code> and <code>exchange_met_ids</code> must already exist in the <code>community</code> model.</p><p><strong>Example</strong></p><pre><code class="language-none">community = add_model_with_exchanges(community,
    model,
    exchange_rxn_mets;
    model_name=&quot;species_2&quot;,
    biomass_id=&quot;BIOMASS_Ecoli_core_w_GAM&quot;)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/community.jl#L62-L87">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_objective!-Tuple{CoreModel, Vector{String}}" href="#COBREXA.add_objective!-Tuple{CoreModel, Vector{String}}"><code>COBREXA.add_objective!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_objective!(
    community::CoreModel,
    objective_mets::Vector{String};
    objective_weights = Float64[],
    objective_column_index = 0,
)</code></pre><p>Add an objective to the <code>community</code> model. Supply the string names of the objective metabolites in <code>objective_mets</code>. Optionally specify the weight to assign each metabolite in the objective function, if unassigned then equal weight is assumed. Also, optionally specify whether the objective already exists in the model by assigning <code>objective_column_index</code>. If unassigned then an objective column will be added, otherwise the column at <code>objective_column_index</code> will be updated.</p><p>Note, the weights are negated inside the function so that the objective metabolites are seen as reagents/substrates, not products in the reaction equation.</p><p><strong>Example</strong></p><pre><code class="language-none">add_objective!(model, [&quot;met1&quot;, &quot;met2&quot;]) # adds a new column with weights = [1,1]
add_objective!(model, [&quot;met1&quot;, &quot;met2&quot;]; objective_weights=[0.1, 0.9]) # adds a new column
add_objective!(model, [&quot;met1&quot;, &quot;met2&quot;]; objective_weights=[0.1, 0.9], objective_column_index=10) # updates column 10</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/community.jl#L1-L26">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.join_with_exchanges-Union{Tuple{M}, Tuple{Vector{M}, Dict{String, String}}} where M&lt;:MetabolicModel" href="#COBREXA.join_with_exchanges-Union{Tuple{M}, Tuple{Vector{M}, Dict{String, String}}} where M&lt;:MetabolicModel"><code>COBREXA.join_with_exchanges</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">join_with_exchanges(models::Vector{M},
    exchange_rxn_mets::Dict{String, String};
    biomass_ids::Vector{String},
    model_names=String[]
)</code></pre><p>Return a <code>CoreModel</code> representing the community model of <code>models</code> joined through their exchange reactions and metabolites in the dictionary <code>exchange_rxn_mets</code>, which maps exchange reactions to their associated metabolite. These exchange reactions and metabolites link model metabolites to environmental metabolites and reactions. Optionally specify <code>model_names</code> to append a specific name to each reaction and metabolite of an organism for easier reference (default is <code>species_i</code> for each model index i in <code>models</code>). Note, the bounds of the environmental variables are all set to zero. Thus, to run a simulation you need to constrain them appropriately. All the other bounds are inherited from the models used to construct the community model.</p><p>If <code>biomass_ids</code> is supplied, then a community model is returned that has an extra reaction added to the end of the stoichiometric matrix (last column) that can be assigned as the objective reaction. It also creates biomass &quot;metabolites&quot; that can be used in this objective reaction. Note, this reaction is unspecified, further action needs to be taken to specify it, e.g. assign weights to the last column of the stoichiometric matrix in the rows corresponding to the biomass metabolites.</p><p>To further clarify how this <code>join</code> works. Suppose you have 2 organisms with stoichiometric matrices S₁ and S₂ and you want to link them with <code>exchange_rxn_mets = Dict(er₁ =&gt; em₁, er₂ =&gt; em₂, er₃ =&gt; em₃, ...)</code>. Then a new community stoichiometric matrix is constructed that looks like this:</p><pre><code class="language-none">            _      er₁  er₂  er₃  ...  b_
           |S₁                           |
           |   S₂                        |
        em₁|                             |
S   =   em₂|                             |
        em₃|                             |
        ...|                             |
        bm₁|                             |
        bm₂|_                           _|
</code></pre><p>The exchange reactions in each model get linked to environmental metabolites, <code>emᵢ</code>, and these get linked to environmental exchanges, <code>erᵢ</code>. These <code>erᵢ</code> behave like normal single organism exchange reactions. When <code>biomass_ids</code> are supplied, each model&#39;s biomass becomes a pseudo-metabolite (<code>bmᵢ</code>). These can be weighted in column <code>b</code>, called the <code>community_biomass</code> reaction in the community model, if desired. Refer to the tutorial if this is unclear.</p><p><strong>Example</strong></p><pre><code class="language-none">m1 = load_model(core_model_path)
m2 = load_model(CoreModel, core_model_path)

# need to list ALL the exchanges that will form part of the entire model
exchange_rxn_mets = Dict(k =&gt; first(keys(reaction_stoichiometry(m1, ex_rxn)))
    for filter(looks_like_exchange_reaction, reactions(m1)))

biomass_ids = [&quot;BIOMASS_Ecoli_core_w_GAM&quot;, &quot;BIOMASS_Ecoli_core_w_GAM&quot;]

community = join_with_exchanges(
    [m1, m2],
    exchange_rxn_mets;
    biomass_ids = biomass_ids,
)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/community.jl#L175-L239">source</a></section></article><h3 id="Model-variant-specifications"><a class="docs-heading-anchor" href="#Model-variant-specifications">Model variant specifications</a><a id="Model-variant-specifications-1"></a><a class="docs-heading-anchor-permalink" href="#Model-variant-specifications" title="Permalink"></a></h3><article class="docstring"><header><a class="docstring-binding" id="COBREXA.with_added_reactions-Tuple" href="#COBREXA.with_added_reactions-Tuple"><code>COBREXA.with_added_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">with_added_reactions(args...; kwargs...)</code></pre><p>Specifies a model variant with reactions added. Forwards the arguments to <a href="#COBREXA.add_reactions-Tuple{CoreModel, CoreModel}"><code>add_reactions</code></a>. Intended to be used with <a href="#COBREXA.screen-Tuple"><code>screen</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/modifications/generic.jl#L34-L39">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.with_changed_bound-Tuple" href="#COBREXA.with_changed_bound-Tuple"><code>COBREXA.with_changed_bound</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">with_changed_bound(args...; kwargs...)</code></pre><p>Specifies a model variant that has a new bound set. Forwards arguments to <a href="#COBREXA.change_bound-Tuple{CoreModel, Int64}"><code>change_bound</code></a>. Intended for usage with <a href="#COBREXA.screen-Tuple"><code>screen</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/modifications/generic.jl#L1-L6">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.with_changed_bounds-Tuple" href="#COBREXA.with_changed_bounds-Tuple"><code>COBREXA.with_changed_bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">with_changed_bounds(args...; kwargs...)</code></pre><p>Specifies a model variant that has new bounds set. Forwards arguments to <a href="#COBREXA.change_bounds-Tuple{CoreModel, AbstractVector{Int64}}"><code>change_bounds</code></a>. Intended for usage with <a href="#COBREXA.screen-Tuple"><code>screen</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/modifications/generic.jl#L9-L14">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.with_removed_metabolite-Tuple" href="#COBREXA.with_removed_metabolite-Tuple"><code>COBREXA.with_removed_metabolite</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">with_removed_metabolite(args...; kwargs...)</code></pre><p>Specifies a model variant without a certain metabolite. Forwards arguments to <a href="#COBREXA.remove_metabolite-Tuple{CoreModel, Int64}"><code>remove_metabolite</code></a>. Intended to be used with <a href="#COBREXA.screen-Tuple"><code>screen</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/modifications/generic.jl#L17-L22">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.with_removed_metabolites-Tuple" href="#COBREXA.with_removed_metabolites-Tuple"><code>COBREXA.with_removed_metabolites</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">with_removed_metabolites(args...; kwargs...)</code></pre><p>Plural version of <a href="#COBREXA.with_removed_metabolite-Tuple"><code>with_removed_metabolite</code></a>, calls <a href="#COBREXA.remove_metabolites-Tuple{CoreModel, AbstractVector{Int64}}"><code>remove_metabolites</code></a> internally.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/modifications/generic.jl#L25-L30">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.with_removed_reaction-Tuple" href="#COBREXA.with_removed_reaction-Tuple"><code>COBREXA.with_removed_reaction</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">with_removed_reaction(args...; kwargs...)</code></pre><p>Specifies a model variant without a certain reaction. Forwards arguments to <a href="#COBREXA.remove_reaction-Tuple{CoreModel, Int64}"><code>remove_reaction</code></a>. Intended to be used with <a href="#COBREXA.screen-Tuple"><code>screen</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/modifications/generic.jl#L42-L47">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.with_removed_reactions-Tuple" href="#COBREXA.with_removed_reactions-Tuple"><code>COBREXA.with_removed_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">with_removed_reactions(args...; kwargs...)</code></pre><p>Plural version of <a href="#COBREXA.with_removed_reaction-Tuple"><code>with_removed_reaction</code></a>, calls <a href="#COBREXA.remove_reactions-Tuple{CoreModel, AbstractVector{Int64}}"><code>remove_reactions</code></a> internally.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/reconstruction/modifications/generic.jl#L50-L55">source</a></section></article><h2 id="Analysis-functions"><a class="docs-heading-anchor" href="#Analysis-functions">Analysis functions</a><a id="Analysis-functions-1"></a><a class="docs-heading-anchor-permalink" href="#Analysis-functions" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="COBREXA.envelope_lattice-Tuple{MetabolicModel, Vector{Int64}}" href="#COBREXA.envelope_lattice-Tuple{MetabolicModel, Vector{Int64}}"><code>COBREXA.envelope_lattice</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">envelope_lattice(
    model::MetabolicModel,
    ridxs::Vector{Int};
    samples = 10,
    ranges = collect(zip(bounds(model)...))[ridxs],
    reaction_samples = fill(samples, length(ridxs)),
)</code></pre><p>Create a lattice (list of &quot;tick&quot; vectors) for reactions at indexes <code>ridxs</code> in a model. Arguments <code>samples</code>, <code>ranges</code>, and <code>reaction_samples</code> may be optionally specified to customize the lattice creation process.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/envelopes.jl#L11-L23">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.envelope_lattice-Tuple{MetabolicModel, Vector{String}}" href="#COBREXA.envelope_lattice-Tuple{MetabolicModel, Vector{String}}"><code>COBREXA.envelope_lattice</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">envelope_lattice(model::MetabolicModel, rids::Vector{String}; kwargs...)</code></pre><p>Version of <a href="#COBREXA.envelope_lattice-Tuple{MetabolicModel, Vector{Int64}}"><code>envelope_lattice</code></a> that works on string reaction IDs instead of integer indexes.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/envelopes.jl#L2-L7">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.objective_envelope-Tuple{MetabolicModel, Vector{Int64}, Any}" href="#COBREXA.objective_envelope-Tuple{MetabolicModel, Vector{Int64}, Any}"><code>COBREXA.objective_envelope</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">objective_envelope(
    model::MetabolicModel,
    ridxs::Vector{Int},
    optimizer;
    modifications = [],
    lattice_args = (),
    lattice = envelope_lattice(model, ridxs; lattice_args...),
    analysis = screen_optimize_objective,
    kwargs...,
)</code></pre><p>Compute an array of objective values for the <code>model</code> for rates of reactions specified <code>ridxs</code> fixed to a regular range of values between their respective lower and upper bounds.</p><p>This can be used to compute a &quot;production envelope&quot; of a metabolite; but generalizes to any specifiable objective and to multiple dimensions of the examined space. For example, to retrieve a production envelope of any metabolite, set the objective coefficient vector of the <code>model</code> to a vector that contains a single <code>1</code> for the exchange reaction that &quot;outputs&quot; this metabolite, and run <a href="#COBREXA.objective_envelope-Tuple{MetabolicModel, Vector{Int64}, Any}"><code>objective_envelope</code></a> with the exchange reaction of the &quot;parameter&quot; metabolite specified in <code>ridxs</code>.</p><p>Returns a named tuple that contains <code>lattice</code> with reference values of the metabolites, and an N-dimensional array <code>values</code> with the computed objective values, where N is the number of specified reactions.  Because of the increasing dimensionality, the computation gets very voluminous with increasing length of <code>ridxs</code>. The <code>lattice</code> for computing the optima can be supplied in the argument; by default it is created by <a href="#COBREXA.envelope_lattice-Tuple{MetabolicModel, Vector{Int64}}"><code>envelope_lattice</code></a> called on the model and reaction indexes. Additional arguments for the call to <a href="#COBREXA.envelope_lattice-Tuple{MetabolicModel, Vector{Int64}}"><code>envelope_lattice</code></a> can be optionally specified in <code>lattice_args</code>.</p><p><code>kwargs</code> are internally forwarded to <a href="#COBREXA.screen_optmodel_modifications-Tuple"><code>screen_optmodel_modifications</code></a>. <code>modifications</code> are appended to the list of modifications after the lattice bounds are set. By default, this returns the objective values for all points in the lattice; alternate outputs can be implemented via the <code>analysis</code> argument.</p><p><strong>Example</strong></p><pre><code class="language-none">julia&gt; m = load_model(&quot;test/downloaded/e_coli_core.xml&quot;);

julia&gt; envelope = objective_envelope(m, [&quot;R_EX_gln__L_e&quot;, &quot;R_EX_fum_e&quot;],
                                     Tulip.Optimizer;
                                     lattice_args=(samples=6,));

julia&gt; envelope.lattice   # the reaction rates for which the optima were computed
2-element Vector{Vector{Float64}}:
 [0.0, 200.0, 400.0, 600.0, 800.0, 1000.0]
 [0.0, 200.0, 400.0, 600.0, 800.0, 1000.0]

julia&gt; envelope.values   # the computed flux objective values for each reaction rate combination
6×6 Matrix{Float64}:
  0.873922   9.25815  17.4538  19.56   20.4121  20.4121
 13.0354    17.508    19.9369  21.894  22.6825  22.6825
 16.6666    18.6097   20.2847  21.894  22.6825  22.6825
 16.6666    18.6097   20.2847  21.894  22.6825  22.6825
 16.6666    18.6097   20.2847  21.894  22.6825  22.6825
 16.6666    18.6097   20.2847  21.894  22.6825  22.6825</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/envelopes.jl#L49-L109">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.objective_envelope-Tuple{MetabolicModel, Vector{String}, Vararg{Any, N} where N}" href="#COBREXA.objective_envelope-Tuple{MetabolicModel, Vector{String}, Vararg{Any, N} where N}"><code>COBREXA.objective_envelope</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">objective_envelope(model::MetabolicModel, rids::Vector{String}, args...; kwargs...)</code></pre><p>Version of <a href="#COBREXA.objective_envelope-Tuple{MetabolicModel, Vector{Int64}, Any}"><code>objective_envelope</code></a> that works on string reaction IDs instead of integer indexes.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/envelopes.jl#L35-L40">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.flux_balance_analysis-Union{Tuple{M}, Tuple{M, Any}} where M&lt;:MetabolicModel" href="#COBREXA.flux_balance_analysis-Union{Tuple{M}, Tuple{M, Any}} where M&lt;:MetabolicModel"><code>COBREXA.flux_balance_analysis</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">flux_balance_analysis(
    model::M,
    optimizer;
    modifications = [],
) where {M&lt;:MetabolicModel}</code></pre><p>Run flux balance analysis (FBA) on the <code>model</code> optionally specifying <code>modifications</code> to the problem.  Basically, FBA solves this optimization problem:</p><pre><code class="language-none">max cᵀx
s.t. S x = b
     xₗ ≤ x ≤ xᵤ</code></pre><p>See &quot;Orth, J., Thiele, I. &amp; Palsson, B. What is flux balance analysis?. Nat Biotechnol 28, 245-248 (2010). https://doi.org/10.1038/nbt.1614&quot; for more information.</p><p>The <code>optimizer</code> must be set to a <code>JuMP</code>-compatible optimizer, such as <code>GLPK.Optimizer</code> or <code>Tulip.Optimizer</code></p><p>Optionally, you may specify one or more modifications to be applied to the model before the analysis, such as <a href="#COBREXA.change_optimizer_attribute-Tuple{Any, Any}"><code>change_optimizer_attribute</code></a>, <a href="#COBREXA.change_objective-Tuple{Union{String, Vector{String}}}"><code>change_objective</code></a>, and <a href="#COBREXA.change_sense-Tuple{Any}"><code>change_sense</code></a>.</p><p>Returns an optimized <code>JuMP</code> model.</p><p><strong>Example</strong></p><pre><code class="language-none">model = load_model(&quot;e_coli_core.json&quot;)
solution = flux_balance_analysis(model, GLPK.optimizer)
value.(solution[:x])  # extract flux steady state from the optimizer

biomass_reaction_id = findfirst(model.reactions, &quot;BIOMASS_Ecoli_core_w_GAM&quot;)

modified_solution = flux_balance_analysis(model, GLPK.optimizer;
    modifications=[change_objective(biomass_reaction_id)])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/flux_balance_analysis.jl#L31-L70">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.flux_balance_analysis_dict-Tuple{MetabolicModel, Vararg{Any, N} where N}" href="#COBREXA.flux_balance_analysis_dict-Tuple{MetabolicModel, Vararg{Any, N} where N}"><code>COBREXA.flux_balance_analysis_dict</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">flux_balance_analysis_dict(model::MetabolicModel, args...)::Union{Dict{String, Float64},Nothing}</code></pre><p>A variant of FBA that returns a dictionary assigning fluxes to reactions, if the solution is found. Arguments are passed to <a href="#COBREXA.flux_balance_analysis-Union{Tuple{M}, Tuple{M, Any}} where M&lt;:MetabolicModel"><code>flux_balance_analysis</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/flux_balance_analysis.jl#L15-L20">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.flux_balance_analysis_vec-Tuple" href="#COBREXA.flux_balance_analysis_vec-Tuple"><code>COBREXA.flux_balance_analysis_vec</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">flux_balance_analysis_vec(args...)::Union{Vector{Float64},Nothing}</code></pre><p>A variant of FBA that returns a vector of fluxes in the same order as reactions of the model, if the solution is found.</p><p>Arguments are passed to <a href="#COBREXA.flux_balance_analysis-Union{Tuple{M}, Tuple{M, Any}} where M&lt;:MetabolicModel"><code>flux_balance_analysis</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/flux_balance_analysis.jl#L1-L8">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA._max_variability_flux-Tuple{Any, Any, Any}" href="#COBREXA._max_variability_flux-Tuple{Any, Any, Any}"><code>COBREXA._max_variability_flux</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">_max_variability_flux(opt_model, rid, ret)</code></pre><p>Internal helper for maximizing reactions in optimization model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/flux_variability_analysis.jl#L157-L161">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.flux_variability_analysis-Tuple{MetabolicModel, Any}" href="#COBREXA.flux_variability_analysis-Tuple{MetabolicModel, Any}"><code>COBREXA.flux_variability_analysis</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">flux_variability_analysis(
    model::MetabolicModel,
    optimizer;
    kwargs...
)</code></pre><p>A simpler version of <a href="#COBREXA.flux_variability_analysis-Tuple{MetabolicModel, Any}"><code>flux_variability_analysis</code></a> that maximizes and minimizes all reactions in the model. Arguments are forwarded.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/flux_variability_analysis.jl#L102-L111">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.flux_variability_analysis-Tuple{MetabolicModel, Vector{Int64}, Any}" href="#COBREXA.flux_variability_analysis-Tuple{MetabolicModel, Vector{Int64}, Any}"><code>COBREXA.flux_variability_analysis</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">flux_variability_analysis(
    model::MetabolicModel,
    reactions::Vector{Int},
    optimizer;
    modifications = [],
    workers = [myid()],
    optimal_objective_value = nothing,
    bounds = z -&gt; (z, Inf),
    ret = objective_value,
)::Matrix{Float64}</code></pre><p>Flux variability analysis solves a pair of optimization problems in <code>model</code> for each flux listed in <code>reactions</code>:</p><pre><code class="language-none"> min,max xᵢ
s.t. S x = b
    xₗ ≤ x ≤ xᵤ
     cᵀx ≥ bounds(Z₀)[1]
     cᵀx ≤ bounds(Z₀)[2]</code></pre><p>where Z₀:= cᵀx₀ is the objective value of an optimal solution of the associated FBA problem (see <a href="#COBREXA.flux_balance_analysis-Union{Tuple{M}, Tuple{M, Any}} where M&lt;:MetabolicModel"><code>flux_balance_analysis</code></a> for a related analysis, also for explanation of the <code>modifications</code> argument).</p><p>The <code>bounds</code> is a user-supplied function that specifies the objective bounds for the variability optimizations, by default it restricts the flux objective value to the precise optimum reached in FBA. It can return <code>-Inf</code> and <code>Inf</code> in first and second pair to remove the limit. Use <a href="#COBREXA.gamma_bounds-Tuple{Any}"><code>gamma_bounds</code></a> and <a href="#COBREXA.objective_bounds-Tuple{Any}"><code>objective_bounds</code></a> for simple bounds.</p><p><code>optimizer</code> must be set to a <code>JuMP</code>-compatible optimizer. The computation of the individual optimization problems is transparently distributed to <code>workers</code> (see <code>Distributed.workers()</code>).  The value of Z₀ can be optionally supplied in argument <code>optimal_objective_value</code>, which prevents this function from calling the non-parallelizable FBA. Separating the single-threaded FBA and multithreaded variability computation can be used to improve resource allocation efficiency in many common use-cases.</p><p><code>ret</code> is a function used to extract results from optimized JuMP models of the individual reactions. By default, it calls and returns the value of <code>JuMP.objective_value</code>. More information can be extracted e.g. by setting it to a function that returns a more elaborate data structure; such as <code>m -&gt; (JuMP.objective_value(m), JuMP.value.(m[:x]))</code>.</p><p>Returns a matrix of extracted <code>ret</code> values for minima and maxima, of total size (<code>length(reactions)</code>,2). The optimizer result status is checked with <a href="#COBREXA.is_solved-Tuple{Any}"><code>is_solved</code></a>; <code>nothing</code> is returned if the optimization failed for any reason.</p><p><strong>Example</strong></p><pre><code class="language-none">model = load_model(&quot;e_coli_core.json&quot;)
flux_variability_analysis(model, [1, 2, 3, 42], GLPK.optimizer)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/flux_variability_analysis.jl#L1-L56">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.flux_variability_analysis_dict-Tuple{MetabolicModel, Any}" href="#COBREXA.flux_variability_analysis_dict-Tuple{MetabolicModel, Any}"><code>COBREXA.flux_variability_analysis_dict</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">flux_variability_analysis_dict(
    model::MetabolicModel,
    optimizer;
    kwargs...
)</code></pre><p>A variant of <a href="#COBREXA.flux_variability_analysis-Tuple{MetabolicModel, Any}"><code>flux_variability_analysis</code></a> that returns the individual maximized and minimized fluxes of all reactions as two dictionaries (of dictionaries). All keyword arguments except <code>ret</code> are passed through.</p><p><strong>Example</strong></p><pre><code class="language-none">mins, maxs = flux_variability_analysis_dict(
    model,
    Tulip.Optimizer;
    bounds = objective_bounds(0.99),
    modifications = [
        change_optimizer_attribute(&quot;IPM_IterationsLimit&quot;, 500),
        change_constraint(&quot;EX_glc__D_e&quot;; lb = -10, ub = -10),
        change_constraint(&quot;EX_o2_e&quot;; lb = 0, ub = 0),
    ],
)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/flux_variability_analysis.jl#L117-L141">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.parsimonious_flux_balance_analysis-Tuple{MetabolicModel, Any}" href="#COBREXA.parsimonious_flux_balance_analysis-Tuple{MetabolicModel, Any}"><code>COBREXA.parsimonious_flux_balance_analysis</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">parsimonious_flux_balance_analysis(
    model::MetabolicModel,
    optimizer;
    modifications = [],
    qp_modifications = [],
    relax_bounds=[1.0, 0.999999, 0.99999, 0.9999, 0.999, 0.99],
)</code></pre><p>Run parsimonious flux balance analysis (pFBA) on the <code>model</code>. In short, pFBA runs two consecutive optimization problems. The first is traditional FBA:</p><pre><code class="language-none">max cᵀx = μ
s.t. S x = b
     xₗ ≤ x ≤ xᵤ</code></pre><p>And the second is a quadratic optimization problem:</p><pre><code class="language-none">min Σᵢ xᵢ²
s.t. S x = b
     xₗ ≤ x ≤ xᵤ
     μ = μ⁰</code></pre><p>Where the optimal solution of the FBA problem, μ⁰, has been added as an additional constraint. See &quot;Lewis, Nathan E, Hixson, Kim K, Conrad, Tom M, Lerman, Joshua A, Charusanti, Pep, Polpitiya, Ashoka D, Adkins, Joshua N, Schramm, Gunnar, Purvine, Samuel O, Lopez-Ferrer, Daniel, Weitz, Karl K, Eils, Roland, König, Rainer, Smith, Richard D, Palsson, Bernhard Ø, (2010) Omic data from evolved E. coli are consistent with computed optimal growth from genome-scale models. Molecular Systems Biology, 6. 390. doi: accession:10.1038/msb.2010.47&quot; for more details.</p><p>pFBA gets the model optimum by standard FBA (using <a href="#COBREXA.flux_balance_analysis-Union{Tuple{M}, Tuple{M, Any}} where M&lt;:MetabolicModel"><code>flux_balance_analysis</code></a> with <code>optimizer</code> and <code>modifications</code>), then finds a minimal total flux through the model that still satisfies the (slightly relaxed) optimum. This is done using a quadratic problem optimizer. If the original optimizer does not support quadratic optimization, it can be changed using the callback in <code>qp_modifications</code>, which are applied after the FBA. See the documentation of <a href="#COBREXA.flux_balance_analysis-Union{Tuple{M}, Tuple{M, Any}} where M&lt;:MetabolicModel"><code>flux_balance_analysis</code></a> for usage examples of modifications.</p><p>Thhe optimum relaxation sequence can be specified in <code>relax</code> parameter, it defaults to multiplicative range of <code>[1.0, 0.999999, ..., 0.99]</code> of the original bound.</p><p>Returns an optimized model that contains the pFBA solution; or <code>nothing</code> if the optimization failed.</p><p><strong>Example</strong></p><pre><code class="language-none">model = load_model(&quot;e_coli_core.json&quot;)
optmodel = parsimonious_flux_balance_analysis(model, biomass, Gurobi.Optimizer)
value.(solution[:x])  # extract the flux from the optimizer</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/parsimonious_flux_balance_analysis.jl#L1-L54">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.parsimonious_flux_balance_analysis_dict-Tuple{MetabolicModel, Vararg{Any, N} where N}" href="#COBREXA.parsimonious_flux_balance_analysis_dict-Tuple{MetabolicModel, Vararg{Any, N} where N}"><code>COBREXA.parsimonious_flux_balance_analysis_dict</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">parsimonious_flux_balance_analysis_dict(model::MetabolicModel, args...; kwargs...)</code></pre><p>Perform parsimonious flux balance analysis on <code>model</code> using <code>optimizer</code>. Returns a dictionary mapping the reaction IDs to fluxes. Arguments are forwarded to <a href="#COBREXA.parsimonious_flux_balance_analysis-Tuple{MetabolicModel, Any}"><code>parsimonious_flux_balance_analysis</code></a> internally.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/parsimonious_flux_balance_analysis.jl#L112-L118">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.parsimonious_flux_balance_analysis_vec-Tuple" href="#COBREXA.parsimonious_flux_balance_analysis_vec-Tuple"><code>COBREXA.parsimonious_flux_balance_analysis_vec</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">parsimonious_flux_balance_analysis_vec(args...; kwargs...)</code></pre><p>Perform parsimonious flux balance analysis on <code>model</code> using <code>optimizer</code>. Returns a vector of fluxes in the same order as the reactions in <code>model</code>. Arguments are forwarded to <a href="#COBREXA.parsimonious_flux_balance_analysis-Tuple{MetabolicModel, Any}"><code>parsimonious_flux_balance_analysis</code></a> internally.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/parsimonious_flux_balance_analysis.jl#L96-L103">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA._screen_args-Tuple{Any, Any, Any}" href="#COBREXA._screen_args-Tuple{Any, Any, Any}"><code>COBREXA._screen_args</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">_screen_args(argtuple, kwargtuple, modsname)</code></pre><p>Internal helper to check the presence and shape of modification and argument arrays in <a href="#COBREXA.screen-Tuple"><code>screen</code></a> and pals.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/screening.jl#L2-L7">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA._screen_impl-Union{Tuple{MetabolicModel}, Tuple{N}, Tuple{A}, Tuple{V}} where {V&lt;:(AbstractVector{T} where T), A, N}" href="#COBREXA._screen_impl-Union{Tuple{MetabolicModel}, Tuple{N}, Tuple{A}, Tuple{V}} where {V&lt;:(AbstractVector{T} where T), A, N}"><code>COBREXA._screen_impl</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">_screen_impl(
    model::MetabolicModel;
    variants::Array{V,N},
    analysis,
    args::Array{A,N},
    workers = [myid()],
)::Array where {V&lt;:AbstractVector,A,N}</code></pre><p>The actual implementation of <a href="#COBREXA.screen-Tuple"><code>screen</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/screening.jl#L113-L123">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA._screen_optmodel_item-Tuple{Any}" href="#COBREXA._screen_optmodel_item-Tuple{Any}"><code>COBREXA._screen_optmodel_item</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">_screen_optmodel_item((mods, args))</code></pre><p>Internal helper for <a href="#COBREXA.screen_optmodel_modifications-Tuple"><code>screen_optmodel_modifications</code></a> that computes one item of the screening task.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/screening.jl#L200-L205">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA._screen_optmodel_modifications_impl-Union{Tuple{N}, Tuple{A}, Tuple{VF}, Tuple{V}, Tuple{MetabolicModel, Any}} where {V&lt;:(AbstractVector{T} where T), VF&lt;:(AbstractVector{T} where T), A, N}" href="#COBREXA._screen_optmodel_modifications_impl-Union{Tuple{N}, Tuple{A}, Tuple{VF}, Tuple{V}, Tuple{MetabolicModel, Any}} where {V&lt;:(AbstractVector{T} where T), VF&lt;:(AbstractVector{T} where T), A, N}"><code>COBREXA._screen_optmodel_modifications_impl</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">_screen_optmodel_modifications_impl(
    model::MetabolicModel,
    optimizer;
    common_modifications::VF = [],
    modifications::Array{V,N},
    args::Array{A,N},
    analysis::Function,
    workers = [myid()],
)::Array where {V&lt;:AbstractVector,VF&lt;:AbstractVector,A,N}</code></pre><p>The actual implementation of <a href="#COBREXA.screen_optmodel_modifications-Tuple"><code>screen_optmodel_modifications</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/screening.jl#L256-L268">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA._screen_optmodel_prepare-Tuple{Any, Any, Any}" href="#COBREXA._screen_optmodel_prepare-Tuple{Any, Any, Any}"><code>COBREXA._screen_optmodel_prepare</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">_screen_optmodel_prepare(model, optimizer, common_modifications)</code></pre><p>Internal helper for <a href="#COBREXA.screen_optmodel_modifications-Tuple"><code>screen_optmodel_modifications</code></a> that creates the model and applies the modifications.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/screening.jl#L185-L190">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.screen-Tuple" href="#COBREXA.screen-Tuple"><code>COBREXA.screen</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">screen(
    model::MetabolicModel;
    variants::Array{V,N}, # defaults to an array of identities
    analysis,
    args::Array{A,N}, # defaults to an array of empty argument lists
    workers = [myid()],
)::Array where {V&lt;:AbstractVector,A,N}</code></pre><p>Take an array of model-modifying function vectors in <code>variants</code>, and execute the function <code>analysis</code> on all variants of the <code>model</code> specified by <code>variants</code>. The computation is distributed over worker IDs in <code>workers</code>. If <code>args</code> are supplied (as an array of the same size as the <code>variants</code>), they are forwarded as arguments to the corresponding analysis function calls.</p><p>The array of variants must contain vectors of single-parameter functions, these are applied to model in order. The functions must <em>not</em> modify the model, but rather return a modified copy. The copy should be made as shallow as possible, to increase memory efficiency of the process. Variant generators that modify the argument model in-place will cause unpredictable results. Refer to the definition of <a href="#COBREXA.screen_variant"><code>screen_variant</code></a> for details.</p><p>The function <code>analysis</code> will receive a single argument (the modified model), together with arguments from <code>args</code> expanded by <code>...</code>. Supply an array of tuples or vectors to pass in multiple arguments to each function. If the argument values should be left intact (not expanded to multiple arguments), they must be wrapped in single-item tuple or vector.</p><p>The modification and analysis functions are transferred to <code>workers</code> as-is; all packages required to run them (e.g. the optimization solvers) must be loaded there. Typically, you want to use the macro <code>@everywhere using MyFavoriteSolver</code> from <code>Distributed</code> package for loading the solvers.</p><p><strong>Return value</strong></p><p>The results of running <code>analysis</code> are collected in to the resulting array, in a way that preserves the shape of the <code>variants</code>, similarly as with <code>pmap</code>.</p><p>The results of <code>analysis</code> function must be serializable, preferably made only from pure Julia structures, because they may be transferred over the network between the computation nodes. For that reason, functions that return whole JuMP models that contain pointers to allocated C structures (such as <a href="#COBREXA.flux_balance_analysis-Union{Tuple{M}, Tuple{M, Any}} where M&lt;:MetabolicModel"><code>flux_balance_analysis</code></a> used with <code>GLPK</code> or <code>Gurobi</code> otimizers) will generally not in this context.</p><p>Note: this function is a thin argument-handling wrapper around <a href="#COBREXA._screen_impl-Union{Tuple{MetabolicModel}, Tuple{N}, Tuple{A}, Tuple{V}} where {V&lt;:(AbstractVector{T} where T), A, N}"><code>_screen_impl</code></a>.</p><p><strong>Example</strong></p><pre><code class="language-none">function reverse_reaction(i::Int)
    (model::CoreModel) -&gt; begin
        mod = copy(model)
        mod.S[:,i] .*= -1   # this is unrealistic but sufficient for demonstration
        mod
    end
end

m = load_model(CoreModel, &quot;e_coli_core.xml&quot;)

screen(m,
    variants = [
        [reverse_reaction(5)],
        [reverse_reaction(3), reverse_reaction(6)]
    ],
    analysis = mod -&gt; mod.S[:,3])  # observe the changes in S

screen(m,
    variants = [
        [reverse_reaction(5)],
        [reverse_reaction(3), reverse_reaction(6)]
    ],
    analysis = mod -&gt; flux_balance_analysis_vec(mod, GLPK.Optimizer))</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/screening.jl#L35-L109">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.screen_optimize_objective-Tuple{Any, Any}" href="#COBREXA.screen_optimize_objective-Tuple{Any, Any}"><code>COBREXA.screen_optimize_objective</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">screen_optimize_objective(_, optmodel)::Union{Float64,Nothing}</code></pre><p>A variant of <a href="#COBREXA.optimize_objective-Tuple{Any}"><code>optimize_objective</code></a> directly usable in <a href="#COBREXA.screen_optmodel_modifications-Tuple"><code>screen_optmodel_modifications</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/screening.jl#L177-L182">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.screen_optmodel_modifications-Tuple" href="#COBREXA.screen_optmodel_modifications-Tuple"><code>COBREXA.screen_optmodel_modifications</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">screen_optmodel_modifications(
    model::MetabolicModel,
    optimizer;
    common_modifications::VF = [],
    modifications::Array{V,N}, # defaults to an array with no modifications
    args::Array{A,N}, # defaults to an array of empty argument lists
    analysis::Function,
    workers = [myid()],
)::Array where {V&lt;:AbstractVector,VF&lt;:AbstractVector,A,N}</code></pre><p>Screen multiple modifications of the same optimization model.</p><p>This function is potentially more efficient than <a href="#COBREXA.screen-Tuple"><code>screen</code></a> because it avoids making variants of the model structure and remaking of the optimization model. On the other hand, modification functions need to keep the optimization model in a recoverable state (one that leaves the model usable for the next modification), which limits the possible spectrum of modifications applied.</p><p>Internally, <code>model</code> is distributed to <code>workers</code> and transformed into the optimization model using <a href="#COBREXA.make_optimization_model-Tuple{MetabolicModel, Any}"><code>make_optimization_model</code></a>. <code>common_modifications</code> are applied to the models at that point. Next, vectors of functions in <code>modifications</code> are consecutively applied, and the result of <code>analysis</code> function called on model are collected to an array of the same extent as <code>modifications</code>. Calls of <code>analysis</code> are optionally supplied with extra arguments from <code>args</code> expanded with <code>...</code>, just like in <a href="#COBREXA.screen-Tuple"><code>screen</code></a>.</p><p>Both the modification functions (in vectors) and the analysis function here have 2 base parameters (as opposed to 1 with <a href="#COBREXA.screen-Tuple"><code>screen</code></a>): first is the <code>model</code> (carried through as-is), second is the prepared JuMP optimization model that may be modified and acted upon. As an example, you can use modification <a href="#COBREXA.change_constraint-Tuple{String}"><code>change_constraint</code></a> and analysis <a href="#COBREXA.screen_optimize_objective-Tuple{Any, Any}"><code>screen_optimize_objective</code></a>.</p><p>Note: This function is a thin argument-handling wrapper around <a href="#COBREXA._screen_optmodel_modifications_impl-Union{Tuple{N}, Tuple{A}, Tuple{VF}, Tuple{V}, Tuple{MetabolicModel, Any}} where {V&lt;:(AbstractVector{T} where T), VF&lt;:(AbstractVector{T} where T), A, N}"><code>_screen_optmodel_modifications_impl</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/screening.jl#L214-L249">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.screen_variant" href="#COBREXA.screen_variant"><code>COBREXA.screen_variant</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">screen_variant(model::MetabolicModel, variant::Vector, analysis, args = ())</code></pre><p>Helper function for <a href="#COBREXA.screen-Tuple"><code>screen</code></a> that applies all single-argument functions in <code>variant</code> to the <code>model</code> (in order from &quot;first&quot; to &quot;last&quot;), and executes <code>analysis</code> on the result.</p><p>Can be used to test model variants locally.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/screening.jl#L153-L161">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.screen_variants-Tuple{Any, Any, Any}" href="#COBREXA.screen_variants-Tuple{Any, Any, Any}"><code>COBREXA.screen_variants</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">screen_variants(model, variants, analysis; workers=[myid()])</code></pre><p>A shortcut for <a href="#COBREXA.screen-Tuple"><code>screen</code></a> that only works with model variants.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/screening.jl#L169-L173">source</a></section></article><h3 id="Analysis-modifications"><a class="docs-heading-anchor" href="#Analysis-modifications">Analysis modifications</a><a id="Analysis-modifications-1"></a><a class="docs-heading-anchor-permalink" href="#Analysis-modifications" title="Permalink"></a></h3><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_crowding_constraint-Tuple{Dict{Int64, Float64}}" href="#COBREXA.add_crowding_constraint-Tuple{Dict{Int64, Float64}}"><code>COBREXA.add_crowding_constraint</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_crowding_constraint(weights::Dict{Int64, Float64})</code></pre><p>Adds a molecular crowding constraint to the optimization problem: <code>∑ wᵢ × vᵢ ≤ 1</code> where <code>wᵢ</code> is a weight and <code>vᵢ</code> is a flux index in the model&#39;s reactions specified in <code>weights</code> as <code>vᵢ =&gt; wᵢ</code> pairs.</p><p>See Beg, Qasim K., et al. &quot;Intracellular crowding defines the mode and sequence of substrate uptake by Escherichia coli and constrains its metabolic activity.&quot; Proceedings of the National Academy of Sciences 104.31 (2007) for more details.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/modifications/crowding.jl#L2-L12">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_crowding_constraint-Tuple{Dict{String, Float64}}" href="#COBREXA.add_crowding_constraint-Tuple{Dict{String, Float64}}"><code>COBREXA.add_crowding_constraint</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_crowding_constraint(weights::Dict{String, Float64})</code></pre><p>Variant of <a href="#COBREXA.add_crowding_constraint-Tuple{Dict{Int64, Float64}}"><code>add_crowding_constraint</code></a> that takes a dictinary of reactions <code>ids</code> instead of reaction indices mapped to weights.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/modifications/crowding.jl#L41-L46">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_crowding_constraint-Tuple{Float64}" href="#COBREXA.add_crowding_constraint-Tuple{Float64}"><code>COBREXA.add_crowding_constraint</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_crowding_constraint(weight::Float64; kwargs)</code></pre><p>Variant of <a href="#COBREXA.add_crowding_constraint-Tuple{Dict{Int64, Float64}}"><code>add_crowding_constraint</code></a> that takes a single weight and assigns it to each internal reaction flux, where internal reactions are identified with <a href="#COBREXA.find_internal_reactions-Tuple{MetabolicModel}"><code>find_internal_reactions</code></a> and <code>kwargs</code> are passed to this function.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/modifications/crowding.jl#L25-L31">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_constraint-Tuple{String}" href="#COBREXA.change_constraint-Tuple{String}"><code>COBREXA.change_constraint</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_constraint(id::String; lb=nothing, ub=nothing)</code></pre><p>Change the lower and upper bounds (<code>lb</code> and <code>ub</code> respectively) of reaction <code>id</code> if supplied.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/modifications/generic.jl#L15-L19">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_objective-Tuple{Union{String, Vector{String}}}" href="#COBREXA.change_objective-Tuple{Union{String, Vector{String}}}"><code>COBREXA.change_objective</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_objective(new_objective::Union{String,Vector{String}}; weights=[], sense=MAX_SENSE)</code></pre><p>Modification that changes the objective function used in a constraint based analysis function.  <code>new_objective</code> can be a single reaction identifier, or an array of reactions identifiers.</p><p>Optionally, the objective can be weighted by a vector of <code>weights</code>, and a optimization <code>sense</code> can be set to either <code>MAX_SENSE</code> or <code>MIN_SENSE</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/modifications/generic.jl#L27-L36">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.constrain_objective_value-Tuple{Any}" href="#COBREXA.constrain_objective_value-Tuple{Any}"><code>COBREXA.constrain_objective_value</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">constrain_objective_value(tolerance)</code></pre><p>Limit the objective value to <code>tolerance</code>-times the current objective value, as with <a href="#COBREXA.objective_bounds-Tuple{Any}"><code>objective_bounds</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/modifications/generic.jl#L2-L7">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA._do_knockout-Tuple{MetabolicModel, Any, Vector{String}}" href="#COBREXA._do_knockout-Tuple{MetabolicModel, Any, Vector{String}}"><code>COBREXA._do_knockout</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">_do_knockout(model::MetabolicModel, opt_model)</code></pre><p>Internal helper for knockouts on generic MetabolicModels. This can be overloaded so that the knockouts may work differently (more efficiently) with other models.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/modifications/knockout.jl#L17-L23">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.knockout-Tuple{String}" href="#COBREXA.knockout-Tuple{String}"><code>COBREXA.knockout</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">knockout(gene_id::String)</code></pre><p>A helper variant of <a href="#COBREXA.knockout-Tuple{String}"><code>knockout</code></a> for a single gene.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/modifications/knockout.jl#L10-L14">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.knockout-Tuple{Vector{String}}" href="#COBREXA.knockout-Tuple{Vector{String}}"><code>COBREXA.knockout</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">knockout(gene_ids::Vector{String})</code></pre><p>A modification that zeroes the bounds of all reactions that would be knocked out by the specified genes (effectively disables the reactions).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/modifications/knockout.jl#L1-L6">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.add_moment_constraints-Tuple{Dict{String, Float64}, Float64}" href="#COBREXA.add_moment_constraints-Tuple{Dict{String, Float64}, Float64}"><code>COBREXA.add_moment_constraints</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">add_moment_constraints(
    ksas::Dict{String,Float64},
    protein_mass_fraction::Float64
)</code></pre><p>A modification that adds enzyme capacity constraints to the problem using a <em>modified</em> version of the MOMENT algorithm. Requires specific activities, <code>ksas</code> [mmol product/g enzyme/h], for each reaction. Proteins are identified by their associated gene IDs. Adds a variable vector <code>y</code> to the problem corresponding to the protein concentration [g enzyme/gDW cell] of each gene product in the order of <code>genes(model)</code>. The total protein concentration [g protein/gDW cell] is constrained to be less than or equal to the <code>protein_mass_fraction</code>. Reaction flux constraints are changed to the MOMENT constraints (see below) for all reactions that have a gene reaction rule, otherwise the flux bounds are left unaltered.</p><p>See Adadi, Roi, et al. &quot;Prediction of microbial growth rate versus biomass yield by a metabolic network with kinetic parameters.&quot; PLoS computational biology (2012) for more details of the original algorithm.</p><p>Here, a streamlined version of the algorithm is implemented to ensure that the correct units are used. Specifically, this implementation uses specific activities instead of <code>kcats</code>. Thus, for a reaction that can only proceed forward and is catalyzed by protein <code>a</code>, the flux <code>x[i]</code> is bounded by <code>x[i] &lt;= ksas[i] * y[a]</code>. If isozymes <code>a</code> or <code>b</code> catalyse the reaction, then <code>x[i] &lt;= ksas[i] * (y[a] + y[b])</code>. If a reaction is catalyzed by subunits <code>a</code> and <code>b</code> then x[i] &lt;= ksas[i] * min(y[a], y[b]). These rules are applied recursively in the model like in the original algorithm. The enzyme capacity constraint is then implemented by <code>sum(y) ≤ protein_mass_fraction</code>. The major benefit of using <code>ksas</code> instead of <code>kcats</code> is that active site number and unit issues are prevented.</p><p><strong>Example</strong></p><pre><code class="language-none">sol = flux_balance_analysis_dict(
    model,
    Tulip.Optimizer;
    modifications = [
        add_moment_constraints(
            ksas,
            protein_mass_fraction
        ),
        change_constraint(&quot;EX_glc__D_e&quot;, lb = -1000),
    ],
)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/modifications/moment.jl#L1-L44">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_optimizer-Tuple{Any}" href="#COBREXA.change_optimizer-Tuple{Any}"><code>COBREXA.change_optimizer</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_optimizer(optimizer)</code></pre><p>Change the JuMP optimizer used to run the optimization.</p><p>This may be used to try different approaches for reaching the optimum, and in problems that may require different optimizers for different parts, such as the <a href="#COBREXA.parsimonious_flux_balance_analysis-Tuple{MetabolicModel, Any}"><code>parsimonious_flux_balance_analysis</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/modifications/optimizer.jl#L14-L22">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_optimizer_attribute-Tuple{Any, Any}" href="#COBREXA.change_optimizer_attribute-Tuple{Any, Any}"><code>COBREXA.change_optimizer_attribute</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_optimizer_attribute(attribute_key, value)</code></pre><p>Change a JuMP optimizer attribute. The attributes are optimizer-specific, refer to the JuMP documentation and the documentation of the specific optimizer for usable keys and values.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/modifications/optimizer.jl#L25-L31">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_sense-Tuple{Any}" href="#COBREXA.change_sense-Tuple{Any}"><code>COBREXA.change_sense</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_sense(objective_sense)</code></pre><p>Change the objective sense of optimization. Possible arguments are <code>MOI.MAX_SENSE</code> and <code>MOI.MIN_SENSE</code>.</p><p>If you want to change the objective and sense at the same time, use <a href="#COBREXA.change_objective-Tuple{Union{String, Vector{String}}}"><code>change_objective</code></a> instead to do both at once.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/modifications/optimizer.jl#L2-L10">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.silence" href="#COBREXA.silence"><code>COBREXA.silence</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">silence</code></pre><p>Modification that disable all output from the JuMP optimizer (shortcut for <code>set_silent</code> from JuMP).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/modifications/optimizer.jl#L35-L40">source</a></section></article><h3 id="Flux-sampling"><a class="docs-heading-anchor" href="#Flux-sampling">Flux sampling</a><a id="Flux-sampling-1"></a><a class="docs-heading-anchor-permalink" href="#Flux-sampling" title="Permalink"></a></h3><article class="docstring"><header><a class="docstring-binding" id="COBREXA._affine_hit_and_run_chain-NTuple{5, Any}" href="#COBREXA._affine_hit_and_run_chain-NTuple{5, Any}"><code>COBREXA._affine_hit_and_run_chain</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">_affine_hit_and_run_chain(warmup, lbs, ubs, iters, chain)</code></pre><p>Internal helper function for computing a single affine hit-and-run chain. The number of the chain is passed for possible future initialization of stable RNGs.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/sampling/affine_hit_and_run.jl#L73-L79">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.affine_hit_and_run-Tuple{Matrix{Float64}, Vector{Float64}, Vector{Float64}}" href="#COBREXA.affine_hit_and_run-Tuple{Matrix{Float64}, Vector{Float64}, Vector{Float64}}"><code>COBREXA.affine_hit_and_run</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">function affine_hit_and_run(
    warmup_points::Matrix{Float64},
    lbs::Vector{Float64},
    ubs::Vector{Float64};
    sample_iters = 100 .* (1:5),
    workers = [myid()],
    chains = length(workers),
)</code></pre><p>Run a hit-and-run style sampling that starts from <code>warmup_points</code> and uses their affine combinations for generating the run directions to sample the space delimited by <code>lbs</code> and <code>ubs</code>.  The points that represent fluxes in <code>warmup_points</code> should be organized in columns, i.e. <code>warmup_points[:,1]</code> is the first warmup flux.</p><p>There are total <code>chains</code> of hit-and-run runs, each on a batch of <code>size(warmup_points, 2)</code> points. The runs are scheduled on <code>workers</code>, for good load balancing <code>chains</code> should be ideally much greater than <code>length(workers)</code>.</p><p>Each run continues for <code>maximum(sample_iters)</code> iterations; the numbers in <code>sample_iters</code> represent the iterations at which the whole &quot;current&quot; batch of points is collected for output. For example, <code>sample_iters=[1,4,5]</code> causes the process run for 5 iterations, returning the sample batch that was produced by 1st, 4th and last (5th) iteration.</p><p>Returns a matrix of sampled fluxes (in columns), with all collected samples horizontally concatenated. The total number of samples (columns) will be <code>size(warmup_points,2) * chains * length(sample_iters)</code>.</p><p><strong>Example</strong></p><pre><code class="language-none">using COBREXA
using Tulip

model = load_model(StandardModel, model_path)

warmup, lbs, ubs = warmup_from_variability(model, Tulip.Optimizer, 100)
samples = affine_hit_and_run(warmup, lbs, ubs, sample_iters = 1:3)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/sampling/affine_hit_and_run.jl#L1-L41">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.warmup_from_variability" href="#COBREXA.warmup_from_variability"><code>COBREXA.warmup_from_variability</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">function warmup_from_variability(
    model::MetabolicModel,
    optimizer,
    min_reactions::Vector{Int}=1:n_reactions(model),
    max_reactions::Vector{Int}=1:n_reactions(model);
    modifications = [],
    workers::Vector{Int} = [myid()],
)::Tuple{Matrix{Float64}, Vector{Float64}, Vector{Float64}}</code></pre><p>Generate FVA-like warmup points for samplers, by minimizing and maximizing the specified reactions. The result is returned as a matrix, each point occupies as single column in the result.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/sampling/warmup_variability.jl#L33-L46">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.warmup_from_variability-Tuple{MetabolicModel, Any, Int64}" href="#COBREXA.warmup_from_variability-Tuple{MetabolicModel, Any, Int64}"><code>COBREXA.warmup_from_variability</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">warmup_from_variability(
    model::MetabolicModel,
    optimizer,
    n_points::Int;
    kwargs...
)</code></pre><p>Generates FVA-like warmup points for samplers, by selecting random points by minimizing and maximizing reactions. Can not return more than 2 times the number of reactions in the model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/analysis/sampling/warmup_variability.jl#L1-L12">source</a></section></article><h2 id="Miscellaneous-utilities"><a class="docs-heading-anchor" href="#Miscellaneous-utilities">Miscellaneous utilities</a><a id="Miscellaneous-utilities-1"></a><a class="docs-heading-anchor-permalink" href="#Miscellaneous-utilities" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="COBREXA.ambiguously_identified_items-Tuple{Dict{String, Dict{String, Set{String}}}}" href="#COBREXA.ambiguously_identified_items-Tuple{Dict{String, Dict{String, Set{String}}}}"><code>COBREXA.ambiguously_identified_items</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">ambiguously_identified_items(
    index::Dict{String,Dict{String,[String]}},
)::Vector{String}</code></pre><p>Find items (genes, metabolites, ...) from the annotation index that are identified non-uniquely by at least one of their annotations.</p><p>This often indicates that the items are duplicate or miscategorized.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/utils/Annotation.jl#L44-L53">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.annotation_index-Tuple{AbstractDict{String, V} where V}" href="#COBREXA.annotation_index-Tuple{AbstractDict{String, V} where V}"><code>COBREXA.annotation_index</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">annotation_index(
    xs::AbstractDict{String};
    annotations = _annotations,
)::Dict{String,Dict{String,[String]}}</code></pre><p>Extract annotations from a dictionary of items <code>xs</code> and build an index that maps annotation &quot;kinds&quot; (e.g. <code>&quot;PubChem&quot;</code>) to the mapping from the annotations (e.g.  <code>&quot;COMPOUND_12345&quot;</code>) to item IDs that carry the annotations.</p><p>Function <code>annotations</code> is used to access the <code>Annotations</code> object in the dictionary values.</p><p>This is extremely useful for finding items by annotation data.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/utils/Annotation.jl#L6-L20">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.check_duplicate_reaction-Tuple{Reaction, OrderedCollections.OrderedDict{String, Reaction}}" href="#COBREXA.check_duplicate_reaction-Tuple{Reaction, OrderedCollections.OrderedDict{String, Reaction}}"><code>COBREXA.check_duplicate_reaction</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">check_duplicate_reaction(rxn::Reaction, rxns::Dict{String, Reaction}; only_metabolites=true)</code></pre><p>Check if <code>rxn</code> already exists in <code>rxns</code> but has another <code>id</code>. If <code>only_metabolites</code> is <code>true</code> then only the metabolite <code>id</code>s are checked. Otherwise, compares metabolite <code>id</code>s and the absolute value of their stoichiometric coefficients to those of <code>rxn</code>. If <code>rxn</code> has the same reaction equation as another reaction in <code>rxns</code>, the return the <code>id</code>. Otherwise return <code>nothing</code>.</p><p>See also: <a href="#COBREXA.reaction_mass_balanced-Tuple{StandardModel, String}"><code>reaction_mass_balanced</code></a></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/utils/Reaction.jl#L1-L11">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.is_boundary-Tuple{Reaction}" href="#COBREXA.is_boundary-Tuple{Reaction}"><code>COBREXA.is_boundary</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">is_boundary(reaction::Reaction)</code></pre><p>Return true if reaction is a boundary reaction, otherwise return false. Checks if on boundary by inspecting the number of metabolites in reaction equation. Boundary reactions have only one metabolite, e.g. an exchange reaction, or a sink/demand reaction.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/utils/Reaction.jl#L41-L47">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_atom_balance-Tuple{StandardModel, Dict{String, Float64}}" href="#COBREXA.reaction_atom_balance-Tuple{StandardModel, Dict{String, Float64}}"><code>COBREXA.reaction_atom_balance</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_atom_balance(model::StandardModel, rxn)</code></pre><p>Returns a dictionary mapping the stoichiometry of atoms through a single reaction. Uses the metabolite information in <code>model</code> to determine the mass balance. Accepts a reaction dictionary, a reaction string id or a <code>Reaction</code> as an argument for <code>rxn</code>.</p><p>See also: <a href="#COBREXA.reaction_mass_balanced-Tuple{StandardModel, String}"><code>reaction_mass_balanced</code></a></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/utils/Reaction.jl#L56-L64">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.reaction_mass_balanced-Tuple{StandardModel, String}" href="#COBREXA.reaction_mass_balanced-Tuple{StandardModel, String}"><code>COBREXA.reaction_mass_balanced</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">reaction_mass_balanced(model::StandardModel, rxn)</code></pre><p>Checks if <code>rxn</code> is atom balanced. Returns a boolean for whether the reaction is balanced, and the associated balance of atoms for convenience (useful if not balanced). Calls <code>reaction_atom_balance</code> internally.</p><p>See also: <a href="#COBREXA.check_duplicate_reaction-Tuple{Reaction, OrderedCollections.OrderedDict{String, Reaction}}"><code>check_duplicate_reaction</code></a>, <a href="#COBREXA.reaction_atom_balance-Tuple{StandardModel, Dict{String, Float64}}"><code>reaction_atom_balance</code></a></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/utils/Reaction.jl#L86-L94">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.stoichiometry_string-Tuple{Any}" href="#COBREXA.stoichiometry_string-Tuple{Any}"><code>COBREXA.stoichiometry_string</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">stoichiometry_string(rxn_dict::Dict{String, Float64}; format_id = x -&gt; x)</code></pre><p>Return the reaction equation as a string. The metabolite strings can be manipulated by setting <code>format_id</code>.</p><p><strong>Example</strong></p><pre><code class="language-none">julia&gt; req = Dict(&quot;coa_c&quot; =&gt; -1, &quot;for_c&quot; =&gt; 1, &quot;accoa_c&quot; =&gt; 1, &quot;pyr_c&quot; =&gt; -1)
julia&gt; stoichiometry_string(req)
&quot;coa_c + pyr_c = for_c + accoa_c&quot;

julia&gt; stoichiometry_string(req; format_id = x -&gt; x[1:end-2])
&quot;coa + pyr = for + accoa&quot;</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/utils/Reaction.jl#L104-L119">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.stoichiometry_string-Tuple{Reaction}" href="#COBREXA.stoichiometry_string-Tuple{Reaction}"><code>COBREXA.stoichiometry_string</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">stoichiometry_string(rxn::Reaction; kwargs)</code></pre><p>Alternative of <a href="#COBREXA.stoichiometry_string-Tuple{Any}"><code>stoichiometry_string</code></a> take takes a <code>Reaction</code> as an argument.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/utils/Reaction.jl#L129-L133">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.serialize_model-Tuple{Serialized, String}" href="#COBREXA.serialize_model-Tuple{Serialized, String}"><code>COBREXA.serialize_model</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">serialize_model(model::Serialized, filename::String)::Serialized</code></pre><p>Specialization of <a href="#COBREXA.serialize_model-Tuple{Serialized, String}"><code>serialize_model</code></a> that prevents nested serialization of already-serialized models.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/utils/Serialized.jl#L17-L22">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.serialize_model-Union{Tuple{MM}, Tuple{MM, String}} where MM&lt;:MetabolicModel" href="#COBREXA.serialize_model-Union{Tuple{MM}, Tuple{MM, String}} where MM&lt;:MetabolicModel"><code>COBREXA.serialize_model</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">serialize_model(model::MM, filename::String)::Serialized{MM} where {MM&lt;:MetabolicModel}</code></pre><p>Serialize the <code>model</code> to file <code>filename</code>, returning a <a href="#COBREXA.Serialized"><code>Serialized</code></a> model that is able to load itself back automatically upon precaching by <a href="#COBREXA.precache!-Tuple{MetabolicModel}"><code>precache!</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/utils/Serialized.jl#L2-L8">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Base.copy-Tuple{Gene}" href="#Base.copy-Tuple{Gene}"><code>Base.copy</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">Base.copy(g::Gene)</code></pre><p>Shallow copy of a <a href="#COBREXA.Gene"><code>Gene</code></a></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/utils/StandardModel.jl#L45-L49">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Base.copy-Tuple{Metabolite}" href="#Base.copy-Tuple{Metabolite}"><code>Base.copy</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">Base.copy(m::Metabolite)</code></pre><p>Shallow copy of a <a href="#COBREXA.Metabolite"><code>Metabolite</code></a></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/utils/StandardModel.jl#L31-L35">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Base.copy-Tuple{Reaction}" href="#Base.copy-Tuple{Reaction}"><code>Base.copy</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">Base.copy(r::Reaction)</code></pre><p>Shallow copy of a <a href="#COBREXA.Reaction"><code>Reaction</code></a></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/utils/StandardModel.jl#L14-L18">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Base.copy-Tuple{StandardModel}" href="#Base.copy-Tuple{StandardModel}"><code>Base.copy</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">Base.copy(m::StandardModel)</code></pre><p>Shallow copy of a <a href="#COBREXA.StandardModel"><code>StandardModel</code></a></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/utils/StandardModel.jl#L2-L6">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.gamma_bounds-Tuple{Any}" href="#COBREXA.gamma_bounds-Tuple{Any}"><code>COBREXA.gamma_bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">gamma_bounds(gamma)</code></pre><p>A bounds-generating function for <a href="#COBREXA.flux_variability_analysis-Tuple{MetabolicModel, Any}"><code>flux_variability_analysis</code></a> that limits the objective value to be at least <code>gamma*Z₀</code>, as usual in COBRA packages. Use as the <code>bounds</code> argument:</p><pre><code class="language-none">flux_variability_analysis(model, some_optimizer; bounds = gamma_bounds(0.9))</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/utils/bounds.jl#L1-L10">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.objective_bounds-Tuple{Any}" href="#COBREXA.objective_bounds-Tuple{Any}"><code>COBREXA.objective_bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">objective_bounds(tolerance)</code></pre><p>A bounds-generating function for <a href="#COBREXA.flux_variability_analysis-Tuple{MetabolicModel, Any}"><code>flux_variability_analysis</code></a> that limits the objective value to a small multiple of Z₀. Use as <code>bounds</code> argument, similarly to <a href="#COBREXA.gamma_bounds-Tuple{Any}"><code>gamma_bounds</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/utils/bounds.jl#L13-L19">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA._parse_formula-Tuple{String}" href="#COBREXA._parse_formula-Tuple{String}"><code>COBREXA._parse_formula</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">_parse_formula(f::String)::MetaboliteFormula</code></pre><p>Parse a formula in format <code>C2H6O</code> into a <a href="#COBREXA.MetaboliteFormula"><code>MetaboliteFormula</code></a>, which is basically a dictionary of atom counts in the molecule.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/utils/chemical_formulas.jl#L2-L7">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA._unparse_formula-Tuple{Dict{String, Int64}}" href="#COBREXA._unparse_formula-Tuple{Dict{String, Int64}}"><code>COBREXA._unparse_formula</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">_unparse_formula(f::MetaboliteFormula)::String</code></pre><p>Format <a href="#COBREXA.MetaboliteFormula"><code>MetaboliteFormula</code></a> to <code>String</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/utils/chemical_formulas.jl#L19-L23">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.atom_fluxes-Tuple{MetabolicModel, Dict{String, Float64}}" href="#COBREXA.atom_fluxes-Tuple{MetabolicModel, Dict{String, Float64}}"><code>COBREXA.atom_fluxes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">atom_fluxes(model::MetabolicModel, reaction_fluxes::Dict{String, Float64})</code></pre><p>Return a dictionary mapping the flux of atoms across a flux solution given by <code>reaction_fluxes</code> using the reactions in <code>model</code> to determine the appropriate stoichiometry.</p><p>Note, this function ignores metabolites with no formula assigned to them, no error message will be displayed.</p><p>Note, if a model is mass balanced there should be not net flux of any atom. By removing reactions from the flux_solution you can investigate how that impacts the mass balances.</p><p><strong>Example</strong></p><pre><code class="language-none"># Find flux of Carbon through all metabolic reactions except the biomass reaction
delete!(fluxes, &quot;BIOMASS_Ecoli_core_w_GAM&quot;)
atom_fluxes(model, fluxes)[&quot;C&quot;]</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/utils/fluxes.jl#L35-L53">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.metabolite_fluxes-Tuple{MetabolicModel, Dict{String, Float64}}" href="#COBREXA.metabolite_fluxes-Tuple{MetabolicModel, Dict{String, Float64}}"><code>COBREXA.metabolite_fluxes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">metabolite_fluxes(model::MetabolicModel, flux_dict::Dict{String, Float64})</code></pre><p>Return two dictionaries of metabolite <code>id</code>s mapped to reactions that consume or produce them, given the flux distribution supplied in <code>flux_dict</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/utils/fluxes.jl#L1-L6">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA._parse_grr-Tuple{SBML.GeneProductAssociation}" href="#COBREXA._parse_grr-Tuple{SBML.GeneProductAssociation}"><code>COBREXA._parse_grr</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">_parse_grr(gpa::SBML.GeneProductAssociation)::GeneAssociation</code></pre><p>Parse <code>SBML.GeneProductAssociation</code> structure to the simpler GeneAssociation. The input must be (implicitly) in a positive DNF.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/utils/gene_associations.jl#L2-L7">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA._parse_grr-Tuple{String}" href="#COBREXA._parse_grr-Tuple{String}"><code>COBREXA._parse_grr</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">_parse_grr(s::String)::GeneAssociation</code></pre><p>Parse a DNF gene association rule in format <code>(YIL010W and YLR043C) or (YIL010W and YGR209C)</code> to <code>GeneAssociation. Also accepts</code>OR<code>,</code>|<code>,</code>||<code>,</code>AND<code>,</code>&amp;<code>, and</code>&amp;&amp;`.</p><p><strong>Example</strong></p><pre><code class="language-none">julia&gt; _parse_grr(&quot;(YIL010W and YLR043C) or (YIL010W and YGR209C)&quot;)
2-element Array{Array{String,1},1}:
 [&quot;YIL010W&quot;, &quot;YLR043C&quot;]
 [&quot;YIL010W&quot;, &quot;YGR209C&quot;]</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/utils/gene_associations.jl#L37-L51">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA._unparse_grr-Tuple{Type{SBML.GeneProductAssociation}, Vector{Vector{String}}}" href="#COBREXA._unparse_grr-Tuple{Type{SBML.GeneProductAssociation}, Vector{Vector{String}}}"><code>COBREXA._unparse_grr</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">_unparse_grr(
    ::Type{SBML.GeneProductAssociation},
    x::GeneAssociation,
)::SBML.GeneAssociation</code></pre><p>Convert a GeneAssociation to the corresponding <code>SBML.jl</code> structure.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/utils/gene_associations.jl#L22-L29">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA._unparse_grr-Tuple{Type{String}, Vector{Vector{String}}}" href="#COBREXA._unparse_grr-Tuple{Type{String}, Vector{Vector{String}}}"><code>COBREXA._unparse_grr</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">unparse_grr(grr::Vector{Vector{Gene}}</code></pre><p>Converts a nested string gene reaction array  back into a gene reaction rule string.</p><p><strong>Example</strong></p><pre><code class="language-none">julia&gt; _unparse_grr(String, [[&quot;YIL010W&quot;, &quot;YLR043C&quot;], [&quot;YIL010W&quot;, &quot;YGR209C&quot;]])
&quot;(YIL010W and YLR043C) or (YIL010W and YGR209C)&quot;</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/utils/gene_associations.jl#L63-L74">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA._guesskey-Tuple{Any, Any}" href="#COBREXA._guesskey-Tuple{Any, Any}"><code>COBREXA._guesskey</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">_guesskey(ks, possibilities)</code></pre><p>Unfortunately, many model types that contain dictionares do not have standardized field names, so we need to try a few possibilities and guess the best one. The keys used to look for valid field names should be ideally specified as constants in <code>src/base/constants.jl</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/utils/guesskey.jl#L2-L9">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.find_biomass_reaction_ids-Tuple{MetabolicModel}" href="#COBREXA.find_biomass_reaction_ids-Tuple{MetabolicModel}"><code>COBREXA.find_biomass_reaction_ids</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">find_biomass_reaction_ids(m::MetabolicModel; kwargs...)</code></pre><p>Shortcut for finding biomass reaction identifiers in a model; arguments are forwarded to <a href="#COBREXA.looks_like_biomass_reaction-Tuple{String}"><code>looks_like_biomass_reaction</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/utils/looks_like.jl#L92-L97">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.find_biomass_reactions-Tuple{MetabolicModel}" href="#COBREXA.find_biomass_reactions-Tuple{MetabolicModel}"><code>COBREXA.find_biomass_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">find_biomass_reactions(m::MetabolicModel; kwargs...)</code></pre><p>Shortcut for finding biomass reaction indexes in a model; arguments are forwarded to <a href="#COBREXA.looks_like_biomass_reaction-Tuple{String}"><code>looks_like_biomass_reaction</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/utils/looks_like.jl#L83-L88">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.find_exchange_metabolite_ids-Tuple{MetabolicModel}" href="#COBREXA.find_exchange_metabolite_ids-Tuple{MetabolicModel}"><code>COBREXA.find_exchange_metabolite_ids</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">find_exchange_metabolite_ids(m::MetabolicModel; kwargs...)</code></pre><p>Shortcut for finding exchange metabolite identifiers in a model; arguments are forwarded to <a href="#COBREXA.looks_like_exchange_metabolite-Tuple{String}"><code>looks_like_exchange_metabolite</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/utils/looks_like.jl#L192-L197">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.find_exchange_metabolites-Tuple{MetabolicModel}" href="#COBREXA.find_exchange_metabolites-Tuple{MetabolicModel}"><code>COBREXA.find_exchange_metabolites</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">find_exchange_metabolites(m::MetabolicModel; kwargs...)</code></pre><p>Shortcut for finding exchange metabolite indexes in a model; arguments are forwarded to <a href="#COBREXA.looks_like_exchange_metabolite-Tuple{String}"><code>looks_like_exchange_metabolite</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/utils/looks_like.jl#L184-L189">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.find_exchange_reaction_ids-Tuple{MetabolicModel}" href="#COBREXA.find_exchange_reaction_ids-Tuple{MetabolicModel}"><code>COBREXA.find_exchange_reaction_ids</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">find_exchange_reaction_ids(m::MetabolicModel; kwargs...)</code></pre><p>Shortcut for finding exchange reaction identifiers in a model; arguments are forwarded to <a href="#COBREXA.looks_like_exchange_reaction-Tuple{String}"><code>looks_like_exchange_reaction</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/utils/looks_like.jl#L46-L51">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.find_exchange_reactions-Tuple{MetabolicModel}" href="#COBREXA.find_exchange_reactions-Tuple{MetabolicModel}"><code>COBREXA.find_exchange_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">find_exchange_reactions(m::MetabolicModel; kwargs...)</code></pre><p>Shortcut for finding exchange reaction indexes in a model; arguments are forwarded to <a href="#COBREXA.looks_like_exchange_reaction-Tuple{String}"><code>looks_like_exchange_reaction</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/utils/looks_like.jl#L37-L42">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.find_internal_reaction_ids-Tuple{MetabolicModel}" href="#COBREXA.find_internal_reaction_ids-Tuple{MetabolicModel}"><code>COBREXA.find_internal_reaction_ids</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">find_internal_reaction_ids(m::MetabolicModel; kwargs...)</code></pre><p>Shortcut for finding internal reaction identifiers in a model; arguments are forwarded to <a href="#COBREXA.looks_like_internal_reaction-Tuple{String}"><code>looks_like_internal_reaction</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/utils/looks_like.jl#L153-L158">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.find_internal_reactions-Tuple{MetabolicModel}" href="#COBREXA.find_internal_reactions-Tuple{MetabolicModel}"><code>COBREXA.find_internal_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">find_internal_reactions(m::MetabolicModel; kwargs...)</code></pre><p>Shortcut for finding internal reaction indices in a model; arguments are forwarded to <a href="#COBREXA.looks_like_internal_reaction-Tuple{String}"><code>looks_like_internal_reaction</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/utils/looks_like.jl#L144-L149">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.looks_like_biomass_reaction-Tuple{String}" href="#COBREXA.looks_like_biomass_reaction-Tuple{String}"><code>COBREXA.looks_like_biomass_reaction</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">looks_like_biomass_reaction(rxn_id::String;
    exclude_exchanges = false,
    exchange_prefixes = _constants.exchange_prefixes,
    biomass_strings = _constants.biomass_strings,
)::Bool</code></pre><p>A predicate that matches reaction identifiers that look like biomass reactions. Biomass reactions are identified by looking for occurences of <code>biomass_strings</code> in the reaction id. If <code>exclude_exchanges</code> is set, the strings that look like exchanges (from <a href="#COBREXA.looks_like_exchange_reaction-Tuple{String}"><code>looks_like_exchange_reaction</code></a>) will not match.</p><p><strong>Example</strong></p><pre><code class="language-none">filter(looks_like_biomass_reaction, reactions(model)) # returns strings
findall(looks_like_biomass_reaction, reactions(model)) # returns indices</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/utils/looks_like.jl#L55-L72">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.looks_like_exchange_metabolite-Tuple{String}" href="#COBREXA.looks_like_exchange_metabolite-Tuple{String}"><code>COBREXA.looks_like_exchange_metabolite</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">looks_like_exchange_metabolite(rxn_id::String;
    exchange_suffixes = _constants.exchange_suffixes,
    )::Bool</code></pre><p>A predicate that matches metabolite identifiers that look like involved in exchange reactions. Exchange metabolites are identified by <code>exchange_suffixes</code> at the end of the metabolite id.</p><p><strong>Example</strong></p><pre><code class="language-none">filter(looks_like_exchange_metabolite, metabolites(model)) # returns strings
findall(looks_like_exchange_metabolite, metabolites(model)) # returns indices</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/utils/looks_like.jl#L162-L176">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.looks_like_exchange_reaction-Tuple{String}" href="#COBREXA.looks_like_exchange_reaction-Tuple{String}"><code>COBREXA.looks_like_exchange_reaction</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">looks_like_exchange_reaction(rxn_id::String;
    exclude_biomass = false,
    biomass_strings = _constants.biomass_strings,
    exchange_prefixes = _constants.exchange_prefixes,
)</code></pre><p>A predicate that matches reaction identifiers that look like exchange or biomass reactions, given the usual naming schemes in common model repositories. Exchange reactions are identified based on matching prefixes in the set <code>exchange_prefixes</code> and biomass reactions are identified by looking for occurences of <code>biomass_strings</code> in the reaction id.</p><p>Also see <a href="#COBREXA.find_exchange_reactions-Tuple{MetabolicModel}"><code>find_exchange_reactions</code></a>.</p><p><strong>Example</strong></p><pre><code class="language-none">findall(looks_like_exchange_reaction, reactions(model)) # returns indices
filter(looks_like_exchange_reaction, reactions(model)) # returns Strings

# to use the optional arguments you need to expand the function&#39;s arguments
# using an anonymous function
findall(x -&gt; looks_like_exchange_reaction(x; exclude_biomass=true), reactions(model)) # returns indices
filter(x -&gt; looks_like_exchange_reaction(x; exclude_biomass=true), reactions(model)) # returns Strings</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/utils/looks_like.jl#L1-L26">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.looks_like_internal_reaction-Tuple{String}" href="#COBREXA.looks_like_internal_reaction-Tuple{String}"><code>COBREXA.looks_like_internal_reaction</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">looks_like_internal_reaction(
    rxn_id::String;
    exchange_prefixes = _constants.exchange_prefixes,
    biomass_strings = _constants.biomass_strings,
)::Bool</code></pre><p>A predicate that matches reaction identifiers that look like internal reactions, i.e. reactions that are neither exchange nor biomass reactions. Exchange reactions are identified based on matching prefixes in the set <code>exchange_prefixes</code> and biomass reactions are identified by looking for occurences of <code>biomass_strings</code> in the reaction id.</p><p>Also see <a href="#COBREXA.find_internal_reactions-Tuple{MetabolicModel}"><code>find_internal_reactions</code></a>.</p><p><strong>Example</strong></p><pre><code class="language-none">findall(looks_like_internal_reaction, reactions(model)) # returns indices
filter(looks_like_internal_reaction, reactions(model)) # returns Strings

# to use the optional arguments you need to expand the function&#39;s arguments
# using an anonymous function
findall(x -&gt; looks_like_internal_reaction(x; exchange_prefixes=[&quot;EX_&quot;, &quot;R_EX_&quot;]), reactions(model)) # returns indices
filter(x -&gt; looks_like_internal_reaction(x; exchange_prefixes=[&quot;EX_&quot;, &quot;R_EX_&quot;]), reactions(model)) # returns Strings</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/utils/looks_like.jl#L101-L126">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bound!-Tuple{CoreModel, Int64}" href="#COBREXA.change_bound!-Tuple{CoreModel, Int64}"><code>COBREXA.change_bound!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bound!(
    model::CoreModel,
    rxn_idx::Int;
    lower = nothing,
    upper = nothing,
)</code></pre><p>Change the specified reaction flux bound in the model in-place.</p><p><strong>Example</strong></p><pre><code class="language-none">new_model = change_bound!(model, 123, lower=-21.15, upper=42.3)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/change_bounds.jl#L44-L59">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bound!-Tuple{CoreModel, String}" href="#COBREXA.change_bound!-Tuple{CoreModel, String}"><code>COBREXA.change_bound!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bound!(
    model::CoreModel,
    rxn_id::String;
    lower = nothing,
    upper = nothing,
)</code></pre><p>Change the specified reaction flux bound in the model in-place.</p><p><strong>Example</strong></p><pre><code class="language-none">new_model = change_bound!(model, &quot;ReactionB&quot;, lower=-21.15, upper=42.3)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/change_bounds.jl#L44-L59">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bound!-Tuple{CoreModelCoupled, Int64}" href="#COBREXA.change_bound!-Tuple{CoreModelCoupled, Int64}"><code>COBREXA.change_bound!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bound!(
    model::CoreModelCoupled,
    rxn_idx::Int;
    lower = nothing,
    upper = nothing,
)</code></pre><p>Change the specified reaction flux bound in the model in-place.</p><p><strong>Example</strong></p><pre><code class="language-none">new_model = change_bound!(model, 123, lower=-21.15, upper=42.3)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/change_bounds.jl#L44-L59">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bound!-Tuple{CoreModelCoupled, String}" href="#COBREXA.change_bound!-Tuple{CoreModelCoupled, String}"><code>COBREXA.change_bound!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bound!(
    model::CoreModelCoupled,
    rxn_id::String;
    lower = nothing,
    upper = nothing,
)</code></pre><p>Change the specified reaction flux bound in the model in-place.</p><p><strong>Example</strong></p><pre><code class="language-none">new_model = change_bound!(model, &quot;ReactionB&quot;, lower=-21.15, upper=42.3)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/change_bounds.jl#L44-L59">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bound!-Tuple{StandardModel, String}" href="#COBREXA.change_bound!-Tuple{StandardModel, String}"><code>COBREXA.change_bound!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bound!(
    model::StandardModel,
    rxn_id::String;
    lower = nothing,
    upper = nothing,
)</code></pre><p>Change the specified reaction flux bound in the model in-place.</p><p><strong>Example</strong></p><pre><code class="language-none">new_model = change_bound!(model, &quot;ReactionB&quot;, lower=-21.15, upper=42.3)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/change_bounds.jl#L44-L59">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bound-Tuple{CoreModel, Int64}" href="#COBREXA.change_bound-Tuple{CoreModel, Int64}"><code>COBREXA.change_bound</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bound(
    model::CoreModel,
    rxn_idx::Int;
    lower = nothing,
    upper = nothing,
)</code></pre><p>Change the specified reaction flux bound in the model and return the modified model.</p><p><strong>Example</strong></p><pre><code class="language-none">change_bound(model, 123, lower=-21.15, upper=42.3)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/change_bounds.jl#L44-L59">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bound-Tuple{CoreModel, String}" href="#COBREXA.change_bound-Tuple{CoreModel, String}"><code>COBREXA.change_bound</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bound(
    model::CoreModel,
    rxn_id::String;
    lower = nothing,
    upper = nothing,
)</code></pre><p>Change the specified reaction flux bound in the model and return the modified model.</p><p><strong>Example</strong></p><pre><code class="language-none">change_bound(model, &quot;ReactionB&quot;, lower=-21.15, upper=42.3)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/change_bounds.jl#L44-L59">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bound-Tuple{CoreModelCoupled, Int64}" href="#COBREXA.change_bound-Tuple{CoreModelCoupled, Int64}"><code>COBREXA.change_bound</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bound(
    model::CoreModelCoupled,
    rxn_idx::Int;
    lower = nothing,
    upper = nothing,
)</code></pre><p>Change the specified reaction flux bound in the model and return the modified model.</p><p><strong>Example</strong></p><pre><code class="language-none">change_bound(model, 123, lower=-21.15, upper=42.3)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/change_bounds.jl#L44-L59">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bound-Tuple{CoreModelCoupled, String}" href="#COBREXA.change_bound-Tuple{CoreModelCoupled, String}"><code>COBREXA.change_bound</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bound(
    model::CoreModelCoupled,
    rxn_id::String;
    lower = nothing,
    upper = nothing,
)</code></pre><p>Change the specified reaction flux bound in the model and return the modified model.</p><p><strong>Example</strong></p><pre><code class="language-none">change_bound(model, &quot;ReactionB&quot;, lower=-21.15, upper=42.3)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/change_bounds.jl#L44-L59">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bound-Tuple{StandardModel, String}" href="#COBREXA.change_bound-Tuple{StandardModel, String}"><code>COBREXA.change_bound</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bound(
    model::StandardModel,
    rxn_id::String;
    lower = nothing,
    upper = nothing,
)</code></pre><p>Change the specified reaction flux bound in the model and return the modified model.</p><p><strong>Example</strong></p><pre><code class="language-none">change_bound(model, &quot;ReactionB&quot;, lower=-21.15, upper=42.3)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/change_bounds.jl#L44-L59">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bounds!-Tuple{CoreModel, AbstractVector{Int64}}" href="#COBREXA.change_bounds!-Tuple{CoreModel, AbstractVector{Int64}}"><code>COBREXA.change_bounds!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bounds!(
    model::CoreModel,
    rxn_idxs::AbstractVector{Int64};
    lower = (nothing for _ = rxn_idxs),
    upper = (nothing for _ = rxn_idxs),
)</code></pre><p>Change the specified reaction flux bounds in the model in-place.</p><p><strong>Example</strong></p><pre><code class="language-none">new_model = change_bounds!(model, [123, 234], lower=[-2.1, -50.05], upper=[4.2, 100.1])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/change_bounds.jl#L44-L59">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bounds!-Tuple{CoreModel, AbstractVector{String}}" href="#COBREXA.change_bounds!-Tuple{CoreModel, AbstractVector{String}}"><code>COBREXA.change_bounds!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bounds!(
    model::CoreModel,
    rxn_ids::AbstractVector{String};
    lower = (nothing for _ = rxn_ids),
    upper = (nothing for _ = rxn_ids),
)</code></pre><p>Change the specified reaction flux bounds in the model in-place.</p><p><strong>Example</strong></p><pre><code class="language-none">new_model = change_bounds!(model, [&quot;ReactionA&quot;, &quot;ReactionC&quot;], lower=[-2.1, -50.05], upper=[4.2, 100.1])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/change_bounds.jl#L44-L59">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bounds!-Tuple{CoreModelCoupled, AbstractVector{Int64}}" href="#COBREXA.change_bounds!-Tuple{CoreModelCoupled, AbstractVector{Int64}}"><code>COBREXA.change_bounds!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bounds!(
    model::CoreModelCoupled,
    rxn_idxs::AbstractVector{Int64};
    lower = (nothing for _ = rxn_idxs),
    upper = (nothing for _ = rxn_idxs),
)</code></pre><p>Change the specified reaction flux bounds in the model in-place.</p><p><strong>Example</strong></p><pre><code class="language-none">new_model = change_bounds!(model, [123, 234], lower=[-2.1, -50.05], upper=[4.2, 100.1])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/change_bounds.jl#L44-L59">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bounds!-Tuple{CoreModelCoupled, AbstractVector{String}}" href="#COBREXA.change_bounds!-Tuple{CoreModelCoupled, AbstractVector{String}}"><code>COBREXA.change_bounds!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bounds!(
    model::CoreModelCoupled,
    rxn_ids::AbstractVector{String};
    lower = (nothing for _ = rxn_ids),
    upper = (nothing for _ = rxn_ids),
)</code></pre><p>Change the specified reaction flux bounds in the model in-place.</p><p><strong>Example</strong></p><pre><code class="language-none">new_model = change_bounds!(model, [&quot;ReactionA&quot;, &quot;ReactionC&quot;], lower=[-2.1, -50.05], upper=[4.2, 100.1])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/change_bounds.jl#L44-L59">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bounds!-Tuple{StandardModel, AbstractVector{String}}" href="#COBREXA.change_bounds!-Tuple{StandardModel, AbstractVector{String}}"><code>COBREXA.change_bounds!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bounds!(
    model::StandardModel,
    rxn_ids::AbstractVector{String};
    lower = (nothing for _ = rxn_ids),
    upper = (nothing for _ = rxn_ids),
)</code></pre><p>Change the specified reaction flux bounds in the model in-place.</p><p><strong>Example</strong></p><pre><code class="language-none">new_model = change_bounds!(model, [&quot;ReactionA&quot;, &quot;ReactionC&quot;], lower=[-2.1, -50.05], upper=[4.2, 100.1])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/change_bounds.jl#L44-L59">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bounds-Tuple{CoreModel, AbstractVector{Int64}}" href="#COBREXA.change_bounds-Tuple{CoreModel, AbstractVector{Int64}}"><code>COBREXA.change_bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bounds(
    model::CoreModel,
    rxn_idxs::AbstractVector{Int64};
    lower = (nothing for _ = rxn_idxs),
    upper = (nothing for _ = rxn_idxs),
)</code></pre><p>Change the specified reaction flux bounds in the model and return the modified model.</p><p><strong>Example</strong></p><pre><code class="language-none">change_bounds(model, [123, 234], lower=[-2.1, -50.05], upper=[4.2, 100.1])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/change_bounds.jl#L44-L59">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bounds-Tuple{CoreModel, AbstractVector{String}}" href="#COBREXA.change_bounds-Tuple{CoreModel, AbstractVector{String}}"><code>COBREXA.change_bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bounds(
    model::CoreModel,
    rxn_ids::AbstractVector{String};
    lower = (nothing for _ = rxn_ids),
    upper = (nothing for _ = rxn_ids),
)</code></pre><p>Change the specified reaction flux bounds in the model and return the modified model.</p><p><strong>Example</strong></p><pre><code class="language-none">change_bounds(model, [&quot;ReactionA&quot;, &quot;ReactionC&quot;], lower=[-2.1, -50.05], upper=[4.2, 100.1])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/change_bounds.jl#L44-L59">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bounds-Tuple{CoreModelCoupled, AbstractVector{Int64}}" href="#COBREXA.change_bounds-Tuple{CoreModelCoupled, AbstractVector{Int64}}"><code>COBREXA.change_bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bounds(
    model::CoreModelCoupled,
    rxn_idxs::AbstractVector{Int64};
    lower = (nothing for _ = rxn_idxs),
    upper = (nothing for _ = rxn_idxs),
)</code></pre><p>Change the specified reaction flux bounds in the model and return the modified model.</p><p><strong>Example</strong></p><pre><code class="language-none">change_bounds(model, [123, 234], lower=[-2.1, -50.05], upper=[4.2, 100.1])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/change_bounds.jl#L44-L59">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bounds-Tuple{CoreModelCoupled, AbstractVector{String}}" href="#COBREXA.change_bounds-Tuple{CoreModelCoupled, AbstractVector{String}}"><code>COBREXA.change_bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bounds(
    model::CoreModelCoupled,
    rxn_ids::AbstractVector{String};
    lower = (nothing for _ = rxn_ids),
    upper = (nothing for _ = rxn_ids),
)</code></pre><p>Change the specified reaction flux bounds in the model and return the modified model.</p><p><strong>Example</strong></p><pre><code class="language-none">change_bounds(model, [&quot;ReactionA&quot;, &quot;ReactionC&quot;], lower=[-2.1, -50.05], upper=[4.2, 100.1])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/change_bounds.jl#L44-L59">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.change_bounds-Tuple{StandardModel, AbstractVector{String}}" href="#COBREXA.change_bounds-Tuple{StandardModel, AbstractVector{String}}"><code>COBREXA.change_bounds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">change_bounds(
    model::StandardModel,
    rxn_ids::AbstractVector{String};
    lower = (nothing for _ = rxn_ids),
    upper = (nothing for _ = rxn_ids),
)</code></pre><p>Change the specified reaction flux bounds in the model and return the modified model.</p><p><strong>Example</strong></p><pre><code class="language-none">change_bounds(model, [&quot;ReactionA&quot;, &quot;ReactionC&quot;], lower=[-2.1, -50.05], upper=[4.2, 100.1])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/change_bounds.jl#L44-L59">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolite!-Tuple{CoreModel, Int64}" href="#COBREXA.remove_metabolite!-Tuple{CoreModel, Int64}"><code>COBREXA.remove_metabolite!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolite!(model::CoreModel, metabolite_idx::Int)</code></pre><p>Remove metabolite from the model of type <code>CoreModel</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolite!-Tuple{CoreModel, String}" href="#COBREXA.remove_metabolite!-Tuple{CoreModel, String}"><code>COBREXA.remove_metabolite!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolite!(model::CoreModel, metabolite_id::String)</code></pre><p>Remove metabolite from the model of type <code>CoreModel</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolite!-Tuple{CoreModelCoupled, Int64}" href="#COBREXA.remove_metabolite!-Tuple{CoreModelCoupled, Int64}"><code>COBREXA.remove_metabolite!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolite!(model::CoreModelCoupled, metabolite_idx::Int)</code></pre><p>Remove metabolite from the model of type <code>CoreModelCoupled</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolite!-Tuple{CoreModelCoupled, String}" href="#COBREXA.remove_metabolite!-Tuple{CoreModelCoupled, String}"><code>COBREXA.remove_metabolite!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolite!(model::CoreModelCoupled, metabolite_id::String)</code></pre><p>Remove metabolite from the model of type <code>CoreModelCoupled</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolite!-Tuple{StandardModel, String}" href="#COBREXA.remove_metabolite!-Tuple{StandardModel, String}"><code>COBREXA.remove_metabolite!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolite!(model::StandardModel, metabolite_id::String)</code></pre><p>Remove metabolite from the model of type <code>StandardModel</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolite-Tuple{CoreModel, Int64}" href="#COBREXA.remove_metabolite-Tuple{CoreModel, Int64}"><code>COBREXA.remove_metabolite</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolite(model::CoreModel, metabolite_idx::Int)</code></pre><p>Remove metabolite from the model of type <code>CoreModel</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolite-Tuple{CoreModel, String}" href="#COBREXA.remove_metabolite-Tuple{CoreModel, String}"><code>COBREXA.remove_metabolite</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolite(model::CoreModel, metabolite_id::String)</code></pre><p>Remove metabolite from the model of type <code>CoreModel</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolite-Tuple{CoreModelCoupled, Int64}" href="#COBREXA.remove_metabolite-Tuple{CoreModelCoupled, Int64}"><code>COBREXA.remove_metabolite</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolite(model::CoreModelCoupled, metabolite_idx::Int)</code></pre><p>Remove metabolite from the model of type <code>CoreModelCoupled</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolite-Tuple{CoreModelCoupled, String}" href="#COBREXA.remove_metabolite-Tuple{CoreModelCoupled, String}"><code>COBREXA.remove_metabolite</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolite(model::CoreModelCoupled, metabolite_id::String)</code></pre><p>Remove metabolite from the model of type <code>CoreModelCoupled</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolite-Tuple{StandardModel, String}" href="#COBREXA.remove_metabolite-Tuple{StandardModel, String}"><code>COBREXA.remove_metabolite</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolite(model::StandardModel, metabolite_id::String)</code></pre><p>Remove metabolite from the model of type <code>StandardModel</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolites!-Tuple{CoreModel, AbstractVector{Int64}}" href="#COBREXA.remove_metabolites!-Tuple{CoreModel, AbstractVector{Int64}}"><code>COBREXA.remove_metabolites!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolites!(model::CoreModel, metabolite_idxs::AbstractVector{Int64})</code></pre><p>Remove metabolites from the model of type <code>CoreModel</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolites!-Tuple{CoreModel, AbstractVector{String}}" href="#COBREXA.remove_metabolites!-Tuple{CoreModel, AbstractVector{String}}"><code>COBREXA.remove_metabolites!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolites!(model::CoreModel, metabolite_ids::AbstractVector{String})</code></pre><p>Remove metabolites from the model of type <code>CoreModel</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolites!-Tuple{CoreModelCoupled, AbstractVector{Int64}}" href="#COBREXA.remove_metabolites!-Tuple{CoreModelCoupled, AbstractVector{Int64}}"><code>COBREXA.remove_metabolites!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolites!(model::CoreModelCoupled, metabolite_idxs::AbstractVector{Int64})</code></pre><p>Remove metabolites from the model of type <code>CoreModelCoupled</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolites!-Tuple{CoreModelCoupled, AbstractVector{String}}" href="#COBREXA.remove_metabolites!-Tuple{CoreModelCoupled, AbstractVector{String}}"><code>COBREXA.remove_metabolites!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolites!(model::CoreModelCoupled, metabolite_ids::AbstractVector{String})</code></pre><p>Remove metabolites from the model of type <code>CoreModelCoupled</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolites!-Tuple{StandardModel, AbstractVector{String}}" href="#COBREXA.remove_metabolites!-Tuple{StandardModel, AbstractVector{String}}"><code>COBREXA.remove_metabolites!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolites!(model::StandardModel, metabolite_ids::AbstractVector{String})</code></pre><p>Remove metabolites from the model of type <code>StandardModel</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolites-Tuple{CoreModel, AbstractVector{Int64}}" href="#COBREXA.remove_metabolites-Tuple{CoreModel, AbstractVector{Int64}}"><code>COBREXA.remove_metabolites</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolites(model::CoreModel, metabolite_idxs::AbstractVector{Int64})</code></pre><p>Remove metabolites from the model of type <code>CoreModel</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolites-Tuple{CoreModel, AbstractVector{String}}" href="#COBREXA.remove_metabolites-Tuple{CoreModel, AbstractVector{String}}"><code>COBREXA.remove_metabolites</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolites(model::CoreModel, metabolite_ids::AbstractVector{String})</code></pre><p>Remove metabolites from the model of type <code>CoreModel</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolites-Tuple{CoreModelCoupled, AbstractVector{Int64}}" href="#COBREXA.remove_metabolites-Tuple{CoreModelCoupled, AbstractVector{Int64}}"><code>COBREXA.remove_metabolites</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolites(model::CoreModelCoupled, metabolite_idxs::AbstractVector{Int64})</code></pre><p>Remove metabolites from the model of type <code>CoreModelCoupled</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolites-Tuple{CoreModelCoupled, AbstractVector{String}}" href="#COBREXA.remove_metabolites-Tuple{CoreModelCoupled, AbstractVector{String}}"><code>COBREXA.remove_metabolites</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolites(model::CoreModelCoupled, metabolite_ids::AbstractVector{String})</code></pre><p>Remove metabolites from the model of type <code>CoreModelCoupled</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_metabolites-Tuple{StandardModel, AbstractVector{String}}" href="#COBREXA.remove_metabolites-Tuple{StandardModel, AbstractVector{String}}"><code>COBREXA.remove_metabolites</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_metabolites(model::StandardModel, metabolite_ids::AbstractVector{String})</code></pre><p>Remove metabolites from the model of type <code>StandardModel</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reaction!-Tuple{CoreModel, Int64}" href="#COBREXA.remove_reaction!-Tuple{CoreModel, Int64}"><code>COBREXA.remove_reaction!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reaction!(model::CoreModel, reaction_idx::Int)</code></pre><p>Remove reaction from the model of type <code>CoreModel</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reaction!-Tuple{CoreModel, String}" href="#COBREXA.remove_reaction!-Tuple{CoreModel, String}"><code>COBREXA.remove_reaction!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reaction!(model::CoreModel, reaction_id::String)</code></pre><p>Remove reaction from the model of type <code>CoreModel</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reaction!-Tuple{CoreModelCoupled, Int64}" href="#COBREXA.remove_reaction!-Tuple{CoreModelCoupled, Int64}"><code>COBREXA.remove_reaction!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reaction!(model::CoreModelCoupled, reaction_idx::Int)</code></pre><p>Remove reaction from the model of type <code>CoreModelCoupled</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reaction!-Tuple{CoreModelCoupled, String}" href="#COBREXA.remove_reaction!-Tuple{CoreModelCoupled, String}"><code>COBREXA.remove_reaction!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reaction!(model::CoreModelCoupled, reaction_id::String)</code></pre><p>Remove reaction from the model of type <code>CoreModelCoupled</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reaction!-Tuple{StandardModel, String}" href="#COBREXA.remove_reaction!-Tuple{StandardModel, String}"><code>COBREXA.remove_reaction!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reaction!(model::StandardModel, reaction_id::String)</code></pre><p>Remove reaction from the model of type <code>StandardModel</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reaction-Tuple{CoreModel, Int64}" href="#COBREXA.remove_reaction-Tuple{CoreModel, Int64}"><code>COBREXA.remove_reaction</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reaction(model::CoreModel, reaction_idx::Int)</code></pre><p>Remove reaction from the model of type <code>CoreModel</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reaction-Tuple{CoreModel, String}" href="#COBREXA.remove_reaction-Tuple{CoreModel, String}"><code>COBREXA.remove_reaction</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reaction(model::CoreModel, reaction_id::String)</code></pre><p>Remove reaction from the model of type <code>CoreModel</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reaction-Tuple{CoreModelCoupled, Int64}" href="#COBREXA.remove_reaction-Tuple{CoreModelCoupled, Int64}"><code>COBREXA.remove_reaction</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reaction(model::CoreModelCoupled, reaction_idx::Int)</code></pre><p>Remove reaction from the model of type <code>CoreModelCoupled</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reaction-Tuple{CoreModelCoupled, String}" href="#COBREXA.remove_reaction-Tuple{CoreModelCoupled, String}"><code>COBREXA.remove_reaction</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reaction(model::CoreModelCoupled, reaction_id::String)</code></pre><p>Remove reaction from the model of type <code>CoreModelCoupled</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reaction-Tuple{StandardModel, String}" href="#COBREXA.remove_reaction-Tuple{StandardModel, String}"><code>COBREXA.remove_reaction</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reaction(model::StandardModel, reaction_id::String)</code></pre><p>Remove reaction from the model of type <code>StandardModel</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reactions!-Tuple{CoreModel, AbstractVector{Int64}}" href="#COBREXA.remove_reactions!-Tuple{CoreModel, AbstractVector{Int64}}"><code>COBREXA.remove_reactions!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reactions!(model::CoreModel, reaction_idxs::AbstractVector{Int64})</code></pre><p>Remove reactions from the model of type <code>CoreModel</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reactions!-Tuple{CoreModel, AbstractVector{String}}" href="#COBREXA.remove_reactions!-Tuple{CoreModel, AbstractVector{String}}"><code>COBREXA.remove_reactions!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reactions!(model::CoreModel, reaction_ids::AbstractVector{String})</code></pre><p>Remove reactions from the model of type <code>CoreModel</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reactions!-Tuple{CoreModelCoupled, AbstractVector{Int64}}" href="#COBREXA.remove_reactions!-Tuple{CoreModelCoupled, AbstractVector{Int64}}"><code>COBREXA.remove_reactions!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reactions!(model::CoreModelCoupled, reaction_idxs::AbstractVector{Int64})</code></pre><p>Remove reactions from the model of type <code>CoreModelCoupled</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reactions!-Tuple{CoreModelCoupled, AbstractVector{String}}" href="#COBREXA.remove_reactions!-Tuple{CoreModelCoupled, AbstractVector{String}}"><code>COBREXA.remove_reactions!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reactions!(model::CoreModelCoupled, reaction_ids::AbstractVector{String})</code></pre><p>Remove reactions from the model of type <code>CoreModelCoupled</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reactions!-Tuple{StandardModel, AbstractVector{String}}" href="#COBREXA.remove_reactions!-Tuple{StandardModel, AbstractVector{String}}"><code>COBREXA.remove_reactions!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reactions!(model::StandardModel, reaction_ids::AbstractVector{String})</code></pre><p>Remove reactions from the model of type <code>StandardModel</code> in-place.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reactions-Tuple{CoreModel, AbstractVector{Int64}}" href="#COBREXA.remove_reactions-Tuple{CoreModel, AbstractVector{Int64}}"><code>COBREXA.remove_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reactions(model::CoreModel, reaction_idxs::AbstractVector{Int64})</code></pre><p>Remove reactions from the model of type <code>CoreModel</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reactions-Tuple{CoreModel, AbstractVector{String}}" href="#COBREXA.remove_reactions-Tuple{CoreModel, AbstractVector{String}}"><code>COBREXA.remove_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reactions(model::CoreModel, reaction_ids::AbstractVector{String})</code></pre><p>Remove reactions from the model of type <code>CoreModel</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reactions-Tuple{CoreModelCoupled, AbstractVector{Int64}}" href="#COBREXA.remove_reactions-Tuple{CoreModelCoupled, AbstractVector{Int64}}"><code>COBREXA.remove_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reactions(model::CoreModelCoupled, reaction_idxs::AbstractVector{Int64})</code></pre><p>Remove reactions from the model of type <code>CoreModelCoupled</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reactions-Tuple{CoreModelCoupled, AbstractVector{String}}" href="#COBREXA.remove_reactions-Tuple{CoreModelCoupled, AbstractVector{String}}"><code>COBREXA.remove_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reactions(model::CoreModelCoupled, reaction_ids::AbstractVector{String})</code></pre><p>Remove reactions from the model of type <code>CoreModelCoupled</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.remove_reactions-Tuple{StandardModel, AbstractVector{String}}" href="#COBREXA.remove_reactions-Tuple{StandardModel, AbstractVector{String}}"><code>COBREXA.remove_reactions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">remove_reactions(model::StandardModel, reaction_ids::AbstractVector{String})</code></pre><p>Remove reactions from the model of type <code>StandardModel</code> and return the modified model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/macros/remove_item.jl#L27-L32">source</a></section></article><h3 id="Logging-and-debugging-helpers"><a class="docs-heading-anchor" href="#Logging-and-debugging-helpers">Logging and debugging helpers</a><a id="Logging-and-debugging-helpers-1"></a><a class="docs-heading-anchor-permalink" href="#Logging-and-debugging-helpers" title="Permalink"></a></h3><article class="docstring"><header><a class="docstring-binding" id="COBREXA.log_io" href="#COBREXA.log_io"><code>COBREXA.log_io</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">log_io(enable::Bool=true)</code></pre><p>Enable (default) or disable (by passing <code>false</code>) output of messages and warnings from model input/output.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/logging/log.jl#L40-L44">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.log_models" href="#COBREXA.log_models"><code>COBREXA.log_models</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">log_models(enable::Bool=true)</code></pre><p>Enable (default) or disable (by passing <code>false</code>) output of model-related messages.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/logging/log.jl#L40-L44">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.log_perf" href="#COBREXA.log_perf"><code>COBREXA.log_perf</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">log_perf(enable::Bool=true)</code></pre><p>Enable (default) or disable (by passing <code>false</code>) output of performance-related tracing information.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/logging/log.jl#L40-L44">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="COBREXA.@_make_logging_tag-Tuple{Symbol, String}" href="#COBREXA.@_make_logging_tag-Tuple{Symbol, String}"><code>COBREXA.@_make_logging_tag</code></a> — <span class="docstring-category">Macro</span></header><section><div><pre><code class="language-julia">macro _make_logging_group(sym::Symbol, doc::String)</code></pre><p>This creates a group of functions that allow masking out topic-related logging actions. A call that goes as follows:</p><pre><code class="language-none">@_make_logging_tag XYZ</code></pre><p>creates the following tools:</p><ul><li>global variable <code>_XYZ_log_enabled</code> defaulted to false</li><li>function <code>log_XYZ</code> that can be called to turn the logging on/off</li><li>a masking macro <code>@_XYZ_log</code> that can be prepended to commands that should only happen if the logging of tag XYZ is enabled.</li></ul><p>The masking macro is then used as follows:</p><pre><code class="language-none">@_XYZ_log @info &quot;This is the extra verbose information you wanted!&quot; a b c</code></pre><p>The user can direct logging with these:</p><pre><code class="language-none">log_XYZ()
log_XYZ(false)</code></pre><p><code>doc</code> should be a name of the stuff that is being printed if the corresponding log_XYZ() is enabled – it is used to create a friendly documentation for the logging switch. In this case it could say <code>&quot;X, Y and Z-related messages&quot;</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/LCSB-BioCore/COBREXA.jl/blob/25d5eb530ba0e93c16ea72b7f2d3e0d91a08c64d/src/base/logging/log.jl#L2-L29">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../notebooks/">« Examples and notebooks</a><a class="docs-footer-nextpage" href="../howToContribute/">How to contribute »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Monday 18 October 2021 11:43">Monday 18 October 2021</span>. Using Julia version 1.6.0.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
